<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">
<meta name="generator" content="pdoc3 0.11.1">
<title>kitchen.ingredients API documentation</title>
<meta name="description" content="Resources and utility functions">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/13.0.0/sanitize.min.css" integrity="sha512-y1dtMcuvtTMJc1yPgEqF0ZjQbhnc/bFhyvIyVNb9Zk5mIGtqVaAB1Ttl28su8AvFMOY0EwRbAe+HCLqj6W7/KA==" crossorigin>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/13.0.0/typography.min.css" integrity="sha512-Y1DYSb995BAfxobCkKepB1BqJJTPrOp3zPL74AWFugHHmmdcvO+C48WLrUOlhGMc0QG7AE3f7gmvvcrmX2fDoA==" crossorigin>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/styles/default.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:1.5em;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:2em 0 .50em 0}h3{font-size:1.4em;margin:1.6em 0 .7em 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .2s ease-in-out}a:visited{color:#503}a:hover{color:#b62}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900;font-weight:bold}pre code{font-size:.8em;line-height:1.4em;padding:1em;display:block}code{background:#f3f3f3;font-family:"DejaVu Sans Mono",monospace;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em 1em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul ul{padding-left:1em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/highlight.min.js" integrity="sha512-D9gUyxqja7hBtkWpPWGt9wfbfaMGVt9gnyCvYa+jojwwPHLCzUm5i8rpk7vD7wNee9bA35eYIjobYPaQuKS1MQ==" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => {
hljs.configure({languages: ['bash', 'css', 'diff', 'graphql', 'ini', 'javascript', 'json', 'plaintext', 'python', 'python-repl', 'rust', 'shell', 'sql', 'typescript', 'xml', 'yaml']});
hljs.highlightAll();
})</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>kitchen.ingredients</code></h1>
</header>
<section id="section-intro">
<p>Resources and utility functions</p>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="kitchen.ingredients.RPKM_to_TPM"><code class="name flex">
<span>def <span class="ident">RPKM_to_TPM</span></span>(<span>RPKM_mat)</span>
</code></dt>
<dd>
<div class="desc"><p>Convert <code>RPKM_mat</code> (reads per KB per million) to TPM (transcripts per million)</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>RPKM_mat</code></strong> :&ensp;<code>np.array</code></dt>
<dd>Matrix of RPKM values in samples x genes format (genes as columns)</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>TPM_mat</code></strong> :&ensp;<code>np.array</code></dt>
<dd>Matrix in same shape as <code>RPKM_mat</code> containing TPM values</dd>
</dl></div>
</dd>
<dt id="kitchen.ingredients.check_dir_exists"><code class="name flex">
<span>def <span class="ident">check_dir_exists</span></span>(<span>path)</span>
</code></dt>
<dd>
<div class="desc"><p>Checks if directory already exists or not and creates it if it doesn't</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>path</code></strong> :&ensp;<code>str</code></dt>
<dd>path to directory</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>tries to make directory at <code>path</code>, unless it already exists</p></div>
</dd>
<dt id="kitchen.ingredients.counts_to_RPKM"><code class="name flex">
<span>def <span class="ident">counts_to_RPKM</span></span>(<span>counts_mat, gene_lengths, mapped_reads=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Convert <code>counts_mat</code> to RPKM (reads per kilobase per million)</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>counts_mat</code></strong> :&ensp;<code>np.array</code></dt>
<dd>Matrix of counts in samples x genes format (genes as columns)</dd>
<dt><strong><code>gene_lengths</code></strong> :&ensp;<code>np.array</code></dt>
<dd>1D array of length <code>counts_mat.shape[1]</code> containing lengths for each gene in
base pairs</dd>
<dt><strong><code>mapped_reads</code></strong> :&ensp;<code>np.array</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>1D array of length <code>counts_mat.shape[0]</code> containing total mapped reads for each
sample. If <code>None</code>, calculate sums manually from columns of <code>counts_mat</code>.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>RPKM_mat</code></strong> :&ensp;<code>np.array</code></dt>
<dd>Matrix in same shape as <code>counts_mat</code> containing RPKM values</dd>
</dl></div>
</dd>
<dt id="kitchen.ingredients.counts_to_TPM"><code class="name flex">
<span>def <span class="ident">counts_to_TPM</span></span>(<span>counts_mat, gene_lengths, mapped_reads=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Convert <code>counts_mat</code> to TPM (transcripts per million)</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>counts_mat</code></strong> :&ensp;<code>np.array</code></dt>
<dd>Matrix of counts in samples x genes format (genes as columns)</dd>
<dt><strong><code>gene_lengths</code></strong> :&ensp;<code>np.array</code></dt>
<dd>1D array of length <code>counts_mat.shape[1]</code> containing lengths for each gene in
base pairs</dd>
<dt><strong><code>mapped_reads</code></strong> :&ensp;<code>np.array (default=</code>None<code>)</code></dt>
<dd>1D array of length <code>counts_mat.shape[0]</code> containing total mapped reads for each
sample. If <code>None</code>, calculate sums manually from columns of <code>counts_mat</code>.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>TPM_mat</code></strong> :&ensp;<code>np.array</code></dt>
<dd>Matrix in same shape as <code>counts_mat</code> containing TPM values</dd>
</dl></div>
</dd>
<dt id="kitchen.ingredients.fetch_decoupler_resources"><code class="name flex">
<span>def <span class="ident">fetch_decoupler_resources</span></span>(<span>resources=['msigdb', 'panglaodb', 'progeny', 'collectri', 'liana'], genome='human')</span>
</code></dt>
<dd>
<div class="desc"><p>Retrieve prior-knowledge networks from OmniPath for use with <code>decoupler</code> pathway
analysis methods
* MSigDB: biological pathways from HALLMARK (ORA)
* PanglaoDB: cell-type and cell-state for scRNA labeling (ORA)
* PROGENy: canonical signaling pathways (MLM)
* CollecTRI: transcription factor regulon networks (ULM)
* LIANA: ligand-receptor interactions (ULM)</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>resources</code></strong> :&ensp;<code>list</code>, optional <code>(default=["msigdb","panglaodb","progeny","collectri","liana"])</code></dt>
<dd>List of resources to fetch. Default all; remove networks not desired.</dd>
<dt><strong><code>genome</code></strong> :&ensp;<code>str literal</code>, optional <code>(default="human")</code></dt>
<dd>One of "human" or "mouse" to determine which gene symbols to return in
<code>genesymbol</code> or <code>target</code> columns of network dataframes</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>nets</code></strong> :&ensp;<code>dict</code></dt>
<dd>Dictionary containing names of OmniPath networks (keys) and the corresponding
dataframes containing gene-pathway information (values).</dd>
</dl></div>
</dd>
<dt id="kitchen.ingredients.filter_signatures_with_var_names"><code class="name flex">
<span>def <span class="ident">filter_signatures_with_var_names</span></span>(<span>signatures_dict, adata)</span>
</code></dt>
<dd>
<div class="desc"><p>Filter lists of genes in <code>signatures_dict</code> to include genes in <code>adata.var_names</code></p></div>
</dd>
<dt id="kitchen.ingredients.flip_signature_dict"><code class="name flex">
<span>def <span class="ident">flip_signature_dict</span></span>(<span>signatures_dict)</span>
</code></dt>
<dd>
<div class="desc"><p>"Flip" dictionary of signatures where keys are signature names and values are lists
of features, returning a dictionary where keys are individual features and values
are signature names</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>signatures_dict</code></strong> :&ensp;<code>dict</code></dt>
<dd>dictionary where keys are signature names and values are lists of features</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>signatures_dict_flipped</code></strong> :&ensp;<code>dict</code></dt>
<dd>dictionary where keys are features and values are signature names</dd>
</dl></div>
</dd>
<dt id="kitchen.ingredients.human_to_mouse_simple"><code class="name flex">
<span>def <span class="ident">human_to_mouse_simple</span></span>(<span>symbol)</span>
</code></dt>
<dd>
<div class="desc"><p>Convert human to mouse symbols by simple case-conversion</p></div>
</dd>
<dt id="kitchen.ingredients.ingest_gene_signatures"><code class="name flex">
<span>def <span class="ident">ingest_gene_signatures</span></span>(<span>sig_files, form='short', sig_col='signature', gene_col='gene')</span>
</code></dt>
<dd>
<div class="desc"><p>Read in gene signatures from one or more flat files</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>sig_files</code></strong> :&ensp;<code>Union[list, str]</code></dt>
<dd>Path to single signature file or list of paths to signature files</dd>
<dt><strong><code>form</code></strong> :&ensp;<code>Literal('short','long'), Optional (default='short')</code></dt>
<dd>Format of the data in <code>sig_files</code>. If 'short', columns of <code>sig_files</code> are
assumed to contain separate signatures, with first row of column headers as
signature names. If 'long', expect <code>sig_col</code> and <code>gene_col</code> headers to describe
signature names and constituent genes in long-form, respectively.</dd>
<dt><strong><code>sig_col</code></strong> :&ensp;<code>str, Optional (default='signature')</code></dt>
<dd>Column in <code>sig_files</code> containing signature names. Ignored if <code>form</code>=='short'.</dd>
<dt><strong><code>gene_col</code></strong> :&ensp;<code>str, Optional (default='gene')</code></dt>
<dd>Column in <code>sig_files</code> containing gene names. Ignored if <code>form</code>=='short'.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>genes</code></strong> :&ensp;<code>dict</code></dt>
<dd>Dictionary of gene signatures with signature names as keys and lists of genes
as values.</dd>
</dl></div>
</dd>
<dt id="kitchen.ingredients.signature_dict_from_rank_genes_groups"><code class="name flex">
<span>def <span class="ident">signature_dict_from_rank_genes_groups</span></span>(<span>adata, uns_key='rank_genes_groups', groups=None, n_genes=5, ambient=False)</span>
</code></dt>
<dd>
<div class="desc"><p>Extract DEGs from AnnData into signature dictionary</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>AnnData object containing DEG results in <code>.uns</code></dd>
<dt><strong><code>uns_key</code></strong> :&ensp;<code>str</code>, optional <code>(default='rank_genes_groups')</code></dt>
<dd>Key from <code>adata.uns</code> containing DEG results. Should reflect categories in
<code>adata.obs[groupby]</code>.</dd>
<dt><strong><code>groups</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of groups within <code>adata.uns[uns_key]</code> to extract DEGs for. If <code>None</code>,
retrieve all groups.</dd>
<dt><strong><code>n_genes</code></strong> :&ensp;<code>int</code>, optional <code>(default=5)</code></dt>
<dd>Number of top genes per group to show</dd>
<dt><strong><code>ambient</code></strong> :&ensp;<code>bool</code>, optional <code>(default=False)</code></dt>
<dd>Include ambient genes as a group in the plot/output dictionary. If <code>True</code>,
<code>adata.var</code> must have a boolean column called 'ambient' labeling ambient genes.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>markers</code></strong> :&ensp;<code>dict</code></dt>
<dd>Dictionary of DEGs group names as keys and lists of genes as values</dd>
</dl></div>
</dd>
<dt id="kitchen.ingredients.signature_dict_values"><code class="name flex">
<span>def <span class="ident">signature_dict_values</span></span>(<span>signatures_dict, unique=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Extract features from dictionary of signatures where keys are signature names and
values are lists of features, returning a single list of features</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>signatures_dict</code></strong> :&ensp;<code>dict</code></dt>
<dd>dictionary where keys are signature names and values are lists of features</dd>
<dt><strong><code>unique</code></strong> :&ensp;<code>bool</code>, optional <code>(default=</code>True<code>)</code></dt>
<dd>get only unique features across all <code>signatures_dict.values()</code></dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>dict_values</code></strong> :&ensp;<code>list</code></dt>
<dd>&nbsp;</dd>
</dl></div>
</dd>
<dt id="kitchen.ingredients.signatures_to_long_form"><code class="name flex">
<span>def <span class="ident">signatures_to_long_form</span></span>(<span>sig_short, sig_col='signature', gene_col='gene')</span>
</code></dt>
<dd>
<div class="desc"><p>Convert gene signatures dict or dataframe from short to long form</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>sig_short</code></strong> :&ensp;<code>Union[dict,pd.DataFrame]</code></dt>
<dd>Gene signatures in dict or short form dataframe, where columns are assumed to
contain separate signatures, with first row of column headers as signature
names.</dd>
<dt><strong><code>sig_col</code></strong> :&ensp;<code>str, Optional (default='signature')</code></dt>
<dd>Column in <code>sig_long</code> to contain signature names.</dd>
<dt><strong><code>gene_col</code></strong> :&ensp;<code>str, Optional (default='gene')</code></dt>
<dd>Column in <code>sig_long</code> to contain gene names.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>sig_long</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>Gene signatures in long form, with signature names in <code>sig_col</code> and gene names
in <code>gene_col</code>.</dd>
</dl></div>
</dd>
<dt id="kitchen.ingredients.signatures_to_short_form"><code class="name flex">
<span>def <span class="ident">signatures_to_short_form</span></span>(<span>sig_long, sig_col='signature', gene_col='gene')</span>
</code></dt>
<dd>
<div class="desc"><p>Convert gene signatures dict or dataframe from long to short form</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>sig_long</code></strong> :&ensp;<code>Union[dict,pd.DataFrame]</code></dt>
<dd>Gene signatures in dict or long form dataframe, where <code>sig_col</code> and <code>gene_col</code>
headers describe signature names and constituent genes, respectively.</dd>
<dt><strong><code>sig_col</code></strong> :&ensp;<code>str, Optional (default='signature')</code></dt>
<dd>Column in <code>sig_long</code> containing signature names.</dd>
<dt><strong><code>gene_col</code></strong> :&ensp;<code>str, Optional (default='gene')</code></dt>
<dd>Column in <code>sig_long</code> containing gene names.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>sig_short</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>Gene signatures in short form, where columns are assumed to
contain separate signatures, with first row of column headers as signature
names.</dd>
</dl></div>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="kitchen" href="index.html">kitchen</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="kitchen.ingredients.RPKM_to_TPM" href="#kitchen.ingredients.RPKM_to_TPM">RPKM_to_TPM</a></code></li>
<li><code><a title="kitchen.ingredients.check_dir_exists" href="#kitchen.ingredients.check_dir_exists">check_dir_exists</a></code></li>
<li><code><a title="kitchen.ingredients.counts_to_RPKM" href="#kitchen.ingredients.counts_to_RPKM">counts_to_RPKM</a></code></li>
<li><code><a title="kitchen.ingredients.counts_to_TPM" href="#kitchen.ingredients.counts_to_TPM">counts_to_TPM</a></code></li>
<li><code><a title="kitchen.ingredients.fetch_decoupler_resources" href="#kitchen.ingredients.fetch_decoupler_resources">fetch_decoupler_resources</a></code></li>
<li><code><a title="kitchen.ingredients.filter_signatures_with_var_names" href="#kitchen.ingredients.filter_signatures_with_var_names">filter_signatures_with_var_names</a></code></li>
<li><code><a title="kitchen.ingredients.flip_signature_dict" href="#kitchen.ingredients.flip_signature_dict">flip_signature_dict</a></code></li>
<li><code><a title="kitchen.ingredients.human_to_mouse_simple" href="#kitchen.ingredients.human_to_mouse_simple">human_to_mouse_simple</a></code></li>
<li><code><a title="kitchen.ingredients.ingest_gene_signatures" href="#kitchen.ingredients.ingest_gene_signatures">ingest_gene_signatures</a></code></li>
<li><code><a title="kitchen.ingredients.signature_dict_from_rank_genes_groups" href="#kitchen.ingredients.signature_dict_from_rank_genes_groups">signature_dict_from_rank_genes_groups</a></code></li>
<li><code><a title="kitchen.ingredients.signature_dict_values" href="#kitchen.ingredients.signature_dict_values">signature_dict_values</a></code></li>
<li><code><a title="kitchen.ingredients.signatures_to_long_form" href="#kitchen.ingredients.signatures_to_long_form">signatures_to_long_form</a></code></li>
<li><code><a title="kitchen.ingredients.signatures_to_short_form" href="#kitchen.ingredients.signatures_to_short_form">signatures_to_short_form</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.11.1</a>.</p>
</footer>
</body>
</html>
