<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.2" />
<title>kitchen.ingredients API documentation</title>
<meta name="description" content="Functions for manipulating .h5ad objects and automated processing of scRNA-seq data" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>kitchen.ingredients</code></h1>
</header>
<section id="section-intro">
<p>Functions for manipulating .h5ad objects and automated processing of scRNA-seq data</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python"># -*- coding: utf-8 -*-
&#34;&#34;&#34;
Functions for manipulating .h5ad objects and automated processing of scRNA-seq data
&#34;&#34;&#34;
import os, errno, argparse
import numpy as np
import scanpy as sc
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
from math import ceil
from matplotlib import rcParams
from emptydrops import find_nonambient_barcodes
from emptydrops.matrix import CountMatrix

sc.set_figure_params(frameon=False, dpi=100, dpi_save=200, format=&#34;png&#34;)

# define cell cycle phase genes
#  human genes (&#39;_h&#39;) from satija lab list
#  mouse genes (&#39;_m&#39;) converted from human list using biomaRt package in R
s_genes_h = [
    &#34;MCM5&#34;,
    &#34;PCNA&#34;,
    &#34;TYMS&#34;,
    &#34;FEN1&#34;,
    &#34;MCM2&#34;,
    &#34;MCM4&#34;,
    &#34;RRM1&#34;,
    &#34;UNG&#34;,
    &#34;GINS2&#34;,
    &#34;MCM6&#34;,
    &#34;CDCA7&#34;,
    &#34;DTL&#34;,
    &#34;PRIM1&#34;,
    &#34;UHRF1&#34;,
    &#34;MLF1IP&#34;,
    &#34;HELLS&#34;,
    &#34;RFC2&#34;,
    &#34;RPA2&#34;,
    &#34;NASP&#34;,
    &#34;RAD51AP1&#34;,
    &#34;GMNN&#34;,
    &#34;WDR76&#34;,
    &#34;SLBP&#34;,
    &#34;CCNE2&#34;,
    &#34;UBR7&#34;,
    &#34;POLD3&#34;,
    &#34;MSH2&#34;,
    &#34;ATAD2&#34;,
    &#34;RAD51&#34;,
    &#34;RRM2&#34;,
    &#34;CDC45&#34;,
    &#34;CDC6&#34;,
    &#34;EXO1&#34;,
    &#34;TIPIN&#34;,
    &#34;DSCC1&#34;,
    &#34;BLM&#34;,
    &#34;CASP8AP2&#34;,
    &#34;USP1&#34;,
    &#34;CLSPN&#34;,
    &#34;POLA1&#34;,
    &#34;CHAF1B&#34;,
    &#34;BRIP1&#34;,
    &#34;E2F8&#34;,
]
s_genes_m = [
    &#34;Gmnn&#34;,
    &#34;Rad51&#34;,
    &#34;Cdca7&#34;,
    &#34;Pold3&#34;,
    &#34;Slbp&#34;,
    &#34;Prim1&#34;,
    &#34;Dscc1&#34;,
    &#34;Rad51ap1&#34;,
    &#34;Fen1&#34;,
    &#34;Mcm4&#34;,
    &#34;Ccne2&#34;,
    &#34;Tyms&#34;,
    &#34;Rrm2&#34;,
    &#34;Usp1&#34;,
    &#34;Wdr76&#34;,
    &#34;Mcm2&#34;,
    &#34;Ung&#34;,
    &#34;E2f8&#34;,
    &#34;Exo1&#34;,
    &#34;Chaf1b&#34;,
    &#34;Blm&#34;,
    &#34;Clspn&#34;,
    &#34;Cdc45&#34;,
    &#34;Cdc6&#34;,
    &#34;Hells&#34;,
    &#34;Nasp&#34;,
    &#34;Ubr7&#34;,
    &#34;Casp8ap2&#34;,
    &#34;Mcm5&#34;,
    &#34;Uhrf1&#34;,
    &#34;Pcna&#34;,
    &#34;Rrm1&#34;,
    &#34;Rfc2&#34;,
    &#34;Tipin&#34;,
    &#34;Brip1&#34;,
    &#34;Gins2&#34;,
    &#34;Dtl&#34;,
    &#34;Pola1&#34;,
    &#34;Rpa2&#34;,
    &#34;Mcm6&#34;,
    &#34;Msh2&#34;,
]
g2m_genes_h = [
    &#34;HMGB2&#34;,
    &#34;CDK1&#34;,
    &#34;NUSAP1&#34;,
    &#34;UBE2C&#34;,
    &#34;BIRC5&#34;,
    &#34;TPX2&#34;,
    &#34;TOP2A&#34;,
    &#34;NDC80&#34;,
    &#34;CKS2&#34;,
    &#34;NUF2&#34;,
    &#34;CKS1B&#34;,
    &#34;MKI67&#34;,
    &#34;TMPO&#34;,
    &#34;CENPF&#34;,
    &#34;TACC3&#34;,
    &#34;FAM64A&#34;,
    &#34;SMC4&#34;,
    &#34;CCNB2&#34;,
    &#34;CKAP2L&#34;,
    &#34;CKAP2&#34;,
    &#34;AURKB&#34;,
    &#34;BUB1&#34;,
    &#34;KIF11&#34;,
    &#34;ANP32E&#34;,
    &#34;TUBB4B&#34;,
    &#34;GTSE1&#34;,
    &#34;KIF20B&#34;,
    &#34;HJURP&#34;,
    &#34;CDCA3&#34;,
    &#34;HN1&#34;,
    &#34;CDC20&#34;,
    &#34;TTK&#34;,
    &#34;CDC25C&#34;,
    &#34;KIF2C&#34;,
    &#34;RANGAP1&#34;,
    &#34;NCAPD2&#34;,
    &#34;DLGAP5&#34;,
    &#34;CDCA2&#34;,
    &#34;CDCA8&#34;,
    &#34;ECT2&#34;,
    &#34;KIF23&#34;,
    &#34;HMMR&#34;,
    &#34;AURKA&#34;,
    &#34;PSRC1&#34;,
    &#34;ANLN&#34;,
    &#34;LBR&#34;,
    &#34;CKAP5&#34;,
    &#34;CENPE&#34;,
    &#34;CTCF&#34;,
    &#34;NEK2&#34;,
    &#34;G2E3&#34;,
    &#34;GAS2L3&#34;,
    &#34;CBX5&#34;,
    &#34;CENPA&#34;,
]
g2m_genes_m = [
    &#34;Tmpo&#34;,
    &#34;Smc4&#34;,
    &#34;Tacc3&#34;,
    &#34;Cdk1&#34;,
    &#34;Ckap2l&#34;,
    &#34;Cks2&#34;,
    &#34;Mki67&#34;,
    &#34;Ckap5&#34;,
    &#34;Nusap1&#34;,
    &#34;Top2a&#34;,
    &#34;Ect2&#34;,
    &#34;Cdca3&#34;,
    &#34;Cdc25c&#34;,
    &#34;Cks1b&#34;,
    &#34;Kif11&#34;,
    &#34;Psrc1&#34;,
    &#34;Dlgap5&#34;,
    &#34;Ckap2&#34;,
    &#34;Aurkb&#34;,
    &#34;Ttk&#34;,
    &#34;Cdca8&#34;,
    &#34;Cdca2&#34;,
    &#34;Cdc20&#34;,
    &#34;Lbr&#34;,
    &#34;Birc5&#34;,
    &#34;Kif20b&#34;,
    &#34;Nuf2&#34;,
    &#34;Anp32e&#34;,
    &#34;Cenpf&#34;,
    &#34;Kif2c&#34;,
    &#34;Ube2c&#34;,
    &#34;Cenpa&#34;,
    &#34;Rangap1&#34;,
    &#34;Hjurp&#34;,
    &#34;Ndc80&#34;,
    &#34;Ncapd2&#34;,
    &#34;Anln&#34;,
    &#34;Cenpe&#34;,
    &#34;Cbx5&#34;,
    &#34;Hmgb2&#34;,
    &#34;Gas2l3&#34;,
    &#34;Cks1brt&#34;,
    &#34;Bub1&#34;,
    &#34;Gtse1&#34;,
    &#34;Nek2&#34;,
    &#34;G2e3&#34;,
    &#34;Tpx2&#34;,
    &#34;Hmmr&#34;,
    &#34;Aurka&#34;,
    &#34;Ccnb2&#34;,
    &#34;Ctcf&#34;,
    &#34;Tubb4b&#34;,
    &#34;Kif23&#34;,
]


def check_dir_exists(path):
    &#34;&#34;&#34;
    Checks if directory already exists or not and creates it if it doesn&#39;t

    Parameters
    ----------

    path : str
        path to directory

    Returns
    -------

    tries to make directory at `path`, unless it already exists
    &#34;&#34;&#34;
    try:
        os.makedirs(path)
    except OSError as exception:
        if exception.errno != errno.EEXIST:
            raise


def list_union(lst1, lst2):
    &#34;&#34;&#34;
    Combines two lists by the union of their values

    Parameters
    ----------

    lst1, lst2 : list
        lists to combine

    Returns
    -------

    final_list : list
        union of values in lst1 and lst2
    &#34;&#34;&#34;
    final_list = set(lst1).union(set(lst2))
    return final_list


def cellranger2(
    adata,
    expected=1500,
    upper_quant=0.99,
    lower_prop=0.1,
    label=&#34;CellRanger_2&#34;,
    verbose=True,
):
    &#34;&#34;&#34;
    Labels cells using &#34;knee point&#34; method from CellRanger 2.1

    Parameters
    ----------

    adata : anndata.AnnData
        object containing unfiltered counts
    expected : int, optional (default=1500)
        estimated number of real cells expected in dataset
    upper_quant : float, optional (default=0.99)
        upper quantile of real cells to test
    lower_prop : float, optional (default=0.1)
        percentage of expected quantile to calculate total counts threshold for
    label : str, optional (default=&#34;CellRanger_2&#34;)
        how to name .obs column containing output
    verbose : bool, optional (default=True)
        print updates to console

    Returns
    -------

    adata edited in place to add .obs[label] binary label
    &#34;&#34;&#34;
    if &#34;total_counts&#34; not in adata.obs.columns:
        adata.obs[&#34;total_counts&#34;] = adata.X.sum(axis=1)
    tmp = np.sort(np.array(adata.obs[&#34;total_counts&#34;]))[::-1]
    thresh = np.quantile(tmp[0:expected], upper_quant) * lower_prop
    adata.uns[&#34;{}_knee_thresh&#34;.format(label)] = thresh
    adata.obs[label] = &#34;False&#34;
    adata.obs.loc[adata.obs.total_counts &gt; thresh, label] = &#34;True&#34;
    adata.obs[label] = adata.obs[label].astype(&#34;category&#34;)
    if verbose:
        print(&#34;Detected knee point: {}&#34;.format(round(thresh, 3)))
        print(adata.obs[label].value_counts())


def cellranger3(
    adata,
    init_counts=15000,
    min_umi_frac_of_median=0.01,
    min_umis_nonambient=500,
    max_adj_pvalue=0.01,
):
    &#34;&#34;&#34;
    Labels cells using &#34;emptydrops&#34; method from CellRanger 3.0

    Parameters
    ----------

    adata : anndata.AnnData
        object containing unfiltered counts
    init_counts : int, optional (default=15000)
        initial total counts threshold for calling cells
    min_umi_frac_of_median : float, optional (default=0.01)
        minimum total counts for testing barcodes as fraction of median counts for
        initially labeled cells
    min_umis_nonambient : float, optional (default=500)
        minimum total counts for testing barcodes
    max_adj_pvalue : float, optional (default=0.01)
        maximum p-value for cell calling after B-H correction

    Returns
    -------

    adata edited in place to add .obs[&#34;CellRanger_3&#34;] binary label
    and .obs[&#34;CellRanger_3_ll&#34;] log-likelihoods for tested barcodes
    &#34;&#34;&#34;
    m = CountMatrix.from_anndata(adata)  # create emptydrops object from adata
    if &#34;total_counts&#34; not in adata.obs.columns:
        adata.obs[&#34;total_counts&#34;] = adata.X.sum(axis=1)
    # label initial cell calls above total counts threshold
    adata.obs[&#34;CellRanger_3&#34;] = False
    adata.obs.loc[adata.obs.total_counts &gt; init_counts, &#34;CellRanger_3&#34;] = True
    # call emptydrops to test for nonambient barcodes
    out = find_nonambient_barcodes(
        m,
        np.array(
            adata.obs.loc[
                adata.obs.CellRanger_3 == True,
            ].index,
            dtype=m.bcs.dtype,
        ),
        min_umi_frac_of_median=min_umi_frac_of_median,
        min_umis_nonambient=min_umis_nonambient,
        max_adj_pvalue=max_adj_pvalue,
    )
    # assign binary labels from emptydrops
    adata.obs.CellRanger_3.iloc[out[0]] = out[-1]
    adata.obs.CellRanger_3 = adata.obs.CellRanger_3.astype(str).astype(
        &#34;category&#34;
    )  # convert to category
    # assign log-likelihoods from emptydrops to .obs
    adata.obs[&#34;CellRanger_3_ll&#34;] = 0
    adata.obs.CellRanger_3_ll.iloc[out[0]] = out[1]


def subset_adata(adata, subset, verbose=True):
    &#34;&#34;&#34;
    Subsets AnnData object on one or more .obs columns

    Columns should contain 0/False for cells to throw out, and 1/True for cells to
    keep. Keeps union of all labels provided in subset.

    Parameters
    ----------

    adata : anndata.AnnData
        the data
    subset : str or list of str
        adata.obs labels to use for subsetting. Labels must be binary (0, &#34;0&#34;, False,
        &#34;False&#34; to toss - 1, &#34;1&#34;, True, &#34;True&#34; to keep). Multiple labels will keep
        intersection.
    verbose : bool, optional (default=True)
        print updates to console

    Returns
    -------

    adata : anndata.AnnData
        new anndata object as subset of `adata`
    &#34;&#34;&#34;
    if verbose:
        print(&#34;Subsetting AnnData on {}&#34;.format(subset), end=&#34;&#34;)
    if isinstance(subset, str):
        subset = [subset]
    # initialize .obs column for choosing cells
    adata.obs[&#34;adata_subset_combined&#34;] = 0
    # create label as union of given subset args
    for i in range(len(subset)):
        adata.obs.loc[
            adata.obs[subset[i]].isin([&#34;True&#34;, True, 1.0, 1]), &#34;adata_subset_combined&#34;
        ] = 1
    adata = adata[adata.obs[&#34;adata_subset_combined&#34;] == 1, :].copy()
    adata.obs.drop(columns=&#34;adata_subset_combined&#34;, inplace=True)
    if verbose:
        print(&#34; - now {} cells and {} genes&#34;.format(adata.n_obs, adata.n_vars))
    return adata


def cc_score(adata, layer=None, seed=18, verbose=True):
    &#34;&#34;&#34;
    Calculates cell cycle scores and implied phase for each observation

    Parameters
    ----------

    adata : anndata.AnnData
        object containing transformed and normalized (arcsinh or log1p) counts in
        &#39;layer&#39;.
    layer : str, optional (default=None)
        key from adata.layers to use for cc phase calculation. Default None to
        use .X
    seed : int, optional (default=18)
        random state for PCA, neighbors graph and clustering
    verbose : bool, optional (default=True)
        print updates to console

    Returns
    -------

    adata is edited in place to add &#39;G2M_score&#39;, &#39;S_score&#39;, and &#39;phase&#39; to .obs
    &#34;&#34;&#34;
    if layer is not None:
        adata.layers[&#34;temp&#34;] = adata.X.copy()
        adata.X = adata.layers[layer].copy()
        if verbose:
            print(&#34;Calculating cell cycle scores using layer: {}&#34;.format(layer))
    else:
        if verbose:
            print(&#34;Calculating cell cycle scores&#34;)
    # determine if sample is mouse or human based on gene names
    if any(item in adata.var_names for item in s_genes_h + g2m_genes_h):
        s_genes, g2m_genes = s_genes_h, g2m_genes_h
    elif any(item in adata.var_names for item in s_genes_m + g2m_genes_m):
        s_genes, g2m_genes = s_genes_m, g2m_genes_m
    # score cell cycle using scanpy function
    sc.tl.score_genes_cell_cycle(
        adata,
        s_genes=s_genes,  # defined at top of script
        g2m_genes=g2m_genes,  # defined at top of script
        random_state=seed,
    )
    if layer is not None:
        adata.X = adata.layers[&#34;temp&#34;].copy()
        del adata.layers[&#34;temp&#34;]


def dim_reduce(
    adata,
    layer=None,
    use_rep=None,
    clust_resolution=1.0,
    paga=True,
    seed=18,
    verbose=True,
):
    &#34;&#34;&#34;
    Reduces dimensions of single-cell dataset using standard methods

    Parameters
    ----------

    adata : anndata.AnnData
        object containing preprocessed counts matrix
    layer : str, optional (default=None)
        layer to use; default None for .X
    use_rep : str, optional (default=None)
        .obsm key to use for neighbors graph instead of PCA;
        default None, generate new PCA from layer
    clust_resolution : float, optional (default=1.0)
        resolution as fraction on [0.0, 1.0] for leiden
        clustering. default 1.0
    paga : bool, optional (default=True)
        run PAGA to seed UMAP embedding
    seed : int, optional (default=18)
        random state for PCA, neighbors graph and clustering
    verbose : bool, optional (default=True)
        print updates to console

    Returns
    -------

    adata is edited in place, adding PCA, neighbors graph, PAGA, and UMAP
    &#34;&#34;&#34;
    if use_rep is None:
        if layer is not None:
            if verbose:
                print(&#34;Using layer {} for dimension reduction&#34;.format(layer))
            adata.X = adata.layers[layer].copy()
        if verbose:
            print(
                &#34;Performing {}-component PCA and building kNN graph with {} neighbors&#34;.format(
                    &#34;50&#34; if adata.n_obs &gt;= 50 else adata.n_obs,
                    int(np.sqrt(adata.n_obs)),
                )
            )
        sc.pp.pca(
            adata,
            n_comps=50 if adata.n_obs &gt;= 50 else adata.n_obs - 1,
            random_state=seed,
        )
        sc.pp.neighbors(
            adata,
            n_neighbors=int(np.sqrt(adata.n_obs)),
            n_pcs=20 if adata.n_obs &gt;= 50 else int(0.4 * (adata.n_obs - 1)),
            random_state=seed,
        )
    else:
        if verbose:
            print(
                &#34;Building kNN graph with {} nearest neighbors using {}&#34;.format(
                    int(np.sqrt(adata.n_obs)), use_rep
                )
            )
        sc.pp.neighbors(
            adata,
            n_neighbors=int(np.sqrt(adata.n_obs)),
            use_rep=use_rep,
            random_state=seed,
        )
    if verbose:
        print(&#34;Clustering cells using Leiden algorithm &#34;, end=&#34;&#34;)
    sc.tl.leiden(adata, random_state=seed, resolution=clust_resolution)
    if verbose:
        print(
            &#34;- {} clusters identified&#34;.format(
                (len(adata.obs.leiden.cat.categories) + 1)
            )
        )
    if paga:
        if verbose:
            print(&#34;Building PAGA graph and UMAP from coordinates&#34;)
        sc.tl.paga(adata)
        sc.pl.paga(adata, show=False)
        sc.tl.umap(adata, init_pos=&#34;paga&#34;, random_state=seed)
    else:
        sc.tl.umap(adata, random_state=seed)


def plot_embedding(
    adata,
    basis=&#34;X_umap&#34;,
    colors=None,
    show_clustering=True,
    ncols=5,
    n_cnmf_markers=7,
    figsize_scale=1.0,
    cmap=&#34;viridis&#34;,
    seed=18,
    save_to=None,
    verbose=True,
    **kwargs,
):
    &#34;&#34;&#34;
    Plots reduced-dimension embeddings of single-cell dataset

    Parameters
    ----------
    adata : anndata.AnnData
        object containing preprocessed and dimension-reduced counts matrix
    basis : str, optional (default=&#34;X_umap&#34;)
        key from `adata.obsm` containing embedding coordinates
    colors : list of str, optional (default=None)
        colors to plot; can be genes or .obs columns
    show_clustering : bool, optional (default=True)
        plot PAGA graph and leiden clusters on first two axes
    basis : str, optional (default=&#34;X_umap&#34;)
        embedding to plot - key from `adata.obsm`
    ncols : int, optional (default=5)
        number of columns in gridspec
    n_cnmf_markers : int, optional (default=7)
        number of top genes to print on cNMF plots
    figsize_scale : float, optional (default=1.0)
        scaler for figure size. calculated using ncols to keep each panel square.
        values &lt; 1.0 will compress figure, &gt; 1.0 will expand.
    cmap : str, optional (default=&#34;viridis&#34;)
        valid color map for the plot
    seed : int, optional (default=18)
        random state for plotting PAGA
    save_to : str, optional (default=None)
        path to .png file for saving figure; default is plt.show()
    verbose : bool, optional (default=True)
        print updates to console
    **kwargs : optional
        args to pass to `sc.pl.embedding` (e.g. &#34;size&#34;, &#34;add_outline&#34;, etc.)

    Returns
    -------
    plot of embedding with overlays from &#34;colors&#34; as matplotlib gridspec object,
    unless `save_to` is not None.
    &#34;&#34;&#34;
    if isinstance(colors, str):  # force colors into list if single string
        colors = [colors]
    if &#34;paga&#34; in adata.uns:
        cluster_colors = [&#34;paga&#34;, &#34;leiden&#34;]
    else:
        cluster_colors = [&#34;leiden&#34;]
    if colors is not None:
        # get full list of things to plot on embedding
        if show_clustering:
            colors = cluster_colors + colors
            unique_colors = list_union(cluster_colors, colors)
        else:
            unique_colors = set(colors)
    else:
        # with no colors provided, plot PAGA graph and leiden clusters
        colors = cluster_colors
        unique_colors = set(colors)
    if verbose:
        print(&#34;Plotting embedding with overlays: {}&#34;.format(list(unique_colors)))
    # set up figure size based on number of plots
    n_plots = len(unique_colors)
    if n_plots &lt;= ncols:
        n_rows, n_cols = 1, n_plots
    else:
        n_rows, n_cols = ceil(n_plots / ncols), ncols
    fig = plt.figure(
        figsize=(ncols * n_cols * figsize_scale, ncols * n_rows * figsize_scale)
    )
    # arrange axes as subplots
    gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
    # add plots to axes
    i = 0
    for color in colors:
        if color in unique_colors:
            ax = plt.subplot(gs[i])
            if color.lower() == &#34;paga&#34;:
                sc.pl.paga(
                    adata,
                    ax=ax,
                    frameon=False,
                    show=False,
                    fontsize=&#34;large&#34;,
                    fontoutline=2.0,
                    node_size_scale=2.5,
                    random_state=seed,
                )
                ax.set_title(label=&#34;PAGA&#34;, loc=&#34;left&#34;, fontweight=&#34;bold&#34;, fontsize=16)
            else:
                if color in [&#34;leiden&#34;, &#34;louvain&#34;, &#34;cluster&#34;, &#34;group&#34;, &#34;cell_type&#34;]:
                    leg_loc, leg_fontsize, leg_fontoutline = &#34;on data&#34;, &#34;large&#34;, 2.0
                else:
                    leg_loc, leg_fontsize, leg_fontoutline = (
                        &#34;right margin&#34;,
                        12,
                        None,
                    )
                sc.pl.embedding(
                    adata,
                    basis=basis,
                    color=color,
                    ax=ax,
                    frameon=False,
                    show=False,
                    legend_loc=leg_loc,
                    legend_fontsize=leg_fontsize,
                    legend_fontoutline=leg_fontoutline,
                    title=&#34;&#34;,
                    color_map=cmap,
                    **kwargs,
                )
                # add top three gene loadings if cNMF
                if color.startswith(&#34;usage_&#34;):
                    y_range = (
                        adata.obsm[basis][:, 1].max() - adata.obsm[basis][:, 1].min()
                    )
                    [
                        ax.text(
                            x=adata.obsm[basis][:, 0].max(),
                            y=adata.obsm[basis][:, 1].max() - (0.06 * y_range * x),
                            s=&#34;&#34;
                            + adata.uns[&#34;cnmf_markers&#34;].loc[x, color.split(&#34;_&#34;)[1]],
                            fontsize=12,
                            fontstyle=&#34;italic&#34;,
                            color=&#34;k&#34;,
                            ha=&#34;right&#34;,
                        )
                        for x in range(n_cnmf_markers)
                    ]
            if color in adata.var_names:  # italicize title if plotting a gene
                ax.set_title(
                    label=color,
                    loc=&#34;left&#34;,
                    fontweight=&#34;bold&#34;,
                    fontsize=16,
                    fontstyle=&#34;italic&#34;,
                )
            else:
                ax.set_title(label=color, loc=&#34;left&#34;, fontweight=&#34;bold&#34;, fontsize=16)
            unique_colors.remove(color)
            i = i + 1
    fig.tight_layout()
    if save_to is not None:
        if verbose:
            print(&#34;Saving embeddings to {}&#34;.format(save_to))
        plt.savefig(save_to)
    else:
        return fig


def plot_genes(
    adata,
    de_method=&#34;t-test_overestim_var&#34;,
    plot_type=&#34;heatmap&#34;,
    groupby=&#34;leiden&#34;,
    n_genes=5,
    dendrogram=True,
    ambient=False,
    cmap=&#34;Reds&#34;,
    save_to=&#34;de.png&#34;,
    verbose=True,
    **kwargs,
):
    &#34;&#34;&#34;
    Calculates and plot `rank_genes_groups` results

    Parameters
    ----------

    adata : anndata.AnnData
        object containing preprocessed and dimension-reduced counts matrix
    de_method : str, optional (default=&#34;t-test_overestim_var&#34;)
        one of &#34;t-test&#34;, &#34;t-test_overestim_var&#34;, &#34;wilcoxon&#34;
    plot_type : str, optional (default=&#34;heatmap&#34;)
        one of &#34;heatmap&#34;, &#34;dotplot&#34;, &#34;matrixplot&#34;
    groupby : str, optional (default=&#34;leiden&#34;)
        .obs key to group cells by
    n_genes : int, optional (default=5)
        number of top genes per group to show
    dendrogram : bool, optional (default=True)
        show dendrogram of cluster similarity
    ambient : bool, optional (default=False)
        include ambient genes as a group in the plot
    cmap : str, optional (default=&#34;Reds&#34;)
        valid color map for the plot
    save_to : str, optional (default=&#34;de.png&#34;)
        string to add to plot name using scanpy plot defaults
    verbose : bool, optional (default=True)
        print updates to console
    **kwargs : optional
        keyword args to add to sc.pl.matrixplot, sc.pl.dotplot, or sc.pl.heatmap

    Returns
    -------

    matplotlib figure
    &#34;&#34;&#34;
    if verbose:
        print(&#34;Performing differential expression analysis...&#34;)
    if de_method in [&#34;t-test&#34;, &#34;t-test_overestim_var&#34;]:
        # rank genes with t-test and B-H correction
        sc.tl.rank_genes_groups(
            adata, groupby=groupby, layer=&#34;log1p_norm&#34;, use_raw=False, method=de_method
        )
    elif de_method == &#34;wilcoxon&#34;:
        # rank genes with wilcoxon rank sum test
        sc.tl.rank_genes_groups(
            adata, groupby=groupby, layer=&#34;raw_counts&#34;, use_raw=False, method=de_method
        )
    else:
        print(
            &#34;Invalid de_method. Must be one of [&#39;t-test&#39;,&#39;t-test_overestim_var&#39;,&#39;wilcoxon&#39;].&#34;
        )
        return

    # calculate arcsinh counts for visualization
    adata.X = adata.layers[&#34;raw_counts&#34;].copy()
    sc.pp.normalize_total(adata)
    adata.X = np.arcsinh(adata.X)
    adata.layers[&#34;arcsinh&#34;] = adata.X.copy()
    adata.X = adata.layers[&#34;raw_counts&#34;].copy()  # return raw counts to .X

    # adjust rcParams
    rcParams[&#34;figure.figsize&#34;] = (4, 4)
    rcParams[&#34;figure.subplot.left&#34;] = 0.18
    rcParams[&#34;figure.subplot.right&#34;] = 0.96
    rcParams[&#34;figure.subplot.bottom&#34;] = 0.15
    rcParams[&#34;figure.subplot.top&#34;] = 0.91

    if ambient:
        # get markers manually and append ambient genes
        markers = {}
        for clu in adata.obs[groupby].unique().tolist():
            markers[clu] = [
                adata.uns[&#34;rank_genes_groups&#34;][&#34;names&#34;][x][clu] for x in range(n_genes)
            ]
        markers[&#34;ambient&#34;] = adata.var_names[adata.var.ambient].tolist()

        if plot_type == &#34;heatmap&#34;:
            myplot = sc.pl.heatmap(
                adata,
                markers,
                dendrogram=dendrogram,
                groupby=groupby,
                swap_axes=True,
                show_gene_labels=True,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                var_group_rotation=0,
                cmap=cmap,
                show=False,
                **kwargs,
            )
            myplot[&#34;heatmap_ax&#34;].set_yticklabels(
                myplot[&#34;heatmap_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
            )
        if plot_type == &#34;dotplot&#34;:
            myplot = sc.pl.dotplot(
                adata,
                markers,
                dendrogram=dendrogram,
                groupby=groupby,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                swap_axes=True,
                var_group_rotation=90,
                show=False,
                return_fig=True,
                **kwargs,
            )
            myplot.style(
                cmap=cmap, color_on=&#34;square&#34;, dot_edge_color=None, dot_edge_lw=1
            )
            myplot.get_axes()[&#34;mainplot_ax&#34;].set_yticklabels(
                myplot.get_axes()[&#34;mainplot_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
            )
        if plot_type == &#34;matrixplot&#34;:
            myplot = sc.pl.matrixplot(
                adata,
                markers,
                dendrogram=dendrogram,
                groupby=groupby,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                var_group_rotation=0,
                cmap=cmap,
                show=False,
                return_fig=True,
                **kwargs,
            )
            myplot.get_axes()[&#34;mainplot_ax&#34;].set_xticklabels(
                myplot.get_axes()[&#34;mainplot_ax&#34;].get_xticklabels(), fontstyle=&#34;italic&#34;
            )

    else:
        if plot_type == &#34;heatmap&#34;:
            myplot = sc.pl.rank_genes_groups_heatmap(
                adata,
                dendrogram=dendrogram,
                groupby=groupby,
                n_genes=n_genes,
                swap_axes=True,
                show_gene_labels=True,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                var_group_rotation=0,
                cmap=cmap,
                show=False,
                **kwargs,
            )
            myplot[&#34;heatmap_ax&#34;].set_yticklabels(
                myplot[&#34;heatmap_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
            )
        if plot_type == &#34;dotplot&#34;:
            myplot = sc.pl.rank_genes_groups_dotplot(
                adata,
                dendrogram=dendrogram,
                groupby=groupby,
                n_genes=n_genes,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                swap_axes=True,
                var_group_rotation=90,
                show=False,
                return_fig=True,
                **kwargs,
            )
            myplot.style(
                cmap=cmap, color_on=&#34;square&#34;, dot_edge_color=None, dot_edge_lw=1
            )
            myplot.get_axes()[&#34;mainplot_ax&#34;].set_yticklabels(
                myplot.get_axes()[&#34;mainplot_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
            )
        if plot_type == &#34;matrixplot&#34;:
            myplot = sc.pl.rank_genes_groups_matrixplot(
                adata,
                dendrogram=dendrogram,
                groupby=groupby,
                n_genes=n_genes,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                var_group_rotation=0,
                cmap=cmap,
                show=False,
                return_fig=True,
                **kwargs,
            )
            myplot.get_axes()[&#34;mainplot_ax&#34;].set_xticklabels(
                myplot.get_axes()[&#34;mainplot_ax&#34;].get_xticklabels(), fontstyle=&#34;italic&#34;
            )

    if save_to is not None:
        plt.savefig(save_to, bbox_inches=&#34;tight&#34;)
    else:
        return myplot


def plot_genes_cnmf(
    adata,
    plot_type=&#34;heatmap&#34;,
    groupby=&#34;leiden&#34;,
    attr=&#34;varm&#34;,
    keys=&#34;cnmf_spectra&#34;,
    indices=None,
    n_genes=5,
    dendrogram=True,
    cmap=&#34;Reds&#34;,
    save_to=&#34;de_cnmf.png&#34;,
    **kwargs,
):
    &#34;&#34;&#34;
    Calculates and plots top cNMF gene loadings

    Parameters
    ----------

    adata : anndata.AnnData
        object containing preprocessed and dimension-reduced counts matrix
    plot_type : str, optional (default=&#34;heatmap&#34;)
        one of &#34;heatmap&#34;, &#34;dotplot&#34;, &#34;matrixplot&#34;
    groupby : str, optional (default=&#34;leiden&#34;)
        .obs key to group cells by
    attr : str {&#34;var&#34;, &#34;obs&#34;, &#34;uns&#34;, &#34;varm&#34;, &#34;obsm&#34;}
        attribute of adata that contains the score
    keys : str or list of str, optional (default=&#34;cnmf_spectra&#34;)
        scores to look up an array from the attribute of adata
    indices : list of int, optional (default=None)
        column indices of keys for which to plot (e.g. [0,1,2] for first three keys)
    n_genes : int, optional (default=5)
        number of top genes per group to show
    dendrogram : bool, optional (default=True)
        show dendrogram of cluster similarity
    cmap : str, optional (default=&#34;Reds&#34;)
        valid color map for the plot
    save_to : str, optional (default=&#34;de.png&#34;)
        string to add to plot name using scanpy plot defaults
    **kwargs : optional
        keyword args to add to sc.pl.matrixplot, sc.pl.dotplot, or sc.pl.heatmap

    Returns
    -------

    matplotlib figure
    &#34;&#34;&#34;
    # calculate arcsinh counts for visualization
    adata.X = adata.layers[&#34;raw_counts&#34;].copy()
    sc.pp.normalize_total(adata)
    adata.X = np.arcsinh(adata.X)
    adata.layers[&#34;arcsinh&#34;] = adata.X.copy()
    adata.X = adata.layers[&#34;raw_counts&#34;].copy()  # return raw counts to .X

    # default to all usages
    if indices is None:
        indices = [x for x in range(getattr(adata, attr)[keys].shape[1])]
    # get scores for each usage
    if isinstance(keys, str) and indices is not None:
        scores = np.array(getattr(adata, attr)[keys])[:, indices]
        keys = [&#34;{}_{}&#34;.format(keys, i + 1) for i in indices]
    labels = adata.var_names  # search all var_names for top genes based on spectra
    # get top n_genes for each spectra
    markers = {}
    for iscore, score in enumerate(scores.T):
        markers[keys[iscore]] = []
        indices = np.argsort(score)[::-1][:n_genes]
        for x in indices[::-1]:
            markers[keys[iscore]].append(labels[x])

    # adjust rcParams
    rcParams[&#34;figure.figsize&#34;] = (4, 4)
    rcParams[&#34;figure.subplot.left&#34;] = 0.18
    rcParams[&#34;figure.subplot.right&#34;] = 0.96
    rcParams[&#34;figure.subplot.bottom&#34;] = 0.15
    rcParams[&#34;figure.subplot.top&#34;] = 0.91

    if plot_type == &#34;heatmap&#34;:
        myplot = sc.pl.heatmap(
            adata,
            markers,
            dendrogram=dendrogram,
            groupby=groupby,
            swap_axes=True,
            show_gene_labels=True,
            layer=&#34;arcsinh&#34;,
            standard_scale=&#34;var&#34;,
            var_group_rotation=0,
            cmap=cmap,
            show=False,
            **kwargs,
        )
        myplot[&#34;heatmap_ax&#34;].set_yticklabels(
            myplot[&#34;heatmap_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
        )
    if plot_type == &#34;dotplot&#34;:
        myplot = sc.pl.dotplot(
            adata,
            markers,
            dendrogram=dendrogram,
            groupby=groupby,
            layer=&#34;arcsinh&#34;,
            standard_scale=&#34;var&#34;,
            swap_axes=True,
            var_group_rotation=90,
            show=False,
            return_fig=True,
            **kwargs,
        )
        myplot.style(cmap=cmap, color_on=&#34;square&#34;, dot_edge_color=None, dot_edge_lw=1)
        myplot.get_axes()[&#34;mainplot_ax&#34;].set_yticklabels(
            myplot.get_axes()[&#34;mainplot_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
        )
    if plot_type == &#34;matrixplot&#34;:
        myplot = sc.pl.matrixplot(
            adata,
            markers,
            dendrogram=dendrogram,
            groupby=groupby,
            layer=&#34;arcsinh&#34;,
            standard_scale=&#34;var&#34;,
            var_group_rotation=0,
            cmap=cmap,
            show=False,
            return_fig=True,
            **kwargs,
        )
        myplot.get_axes()[&#34;mainplot_ax&#34;].set_xticklabels(
            myplot.get_axes()[&#34;mainplot_ax&#34;].get_xticklabels(), fontstyle=&#34;italic&#34;
        )

    if save_to is not None:
        plt.savefig(save_to, bbox_inches=&#34;tight&#34;)
    else:
        return myplot


def rank_genes_cnmf(
    adata,
    attr=&#34;varm&#34;,
    keys=&#34;cnmf_spectra&#34;,
    indices=None,
    labels=None,
    titles=None,
    color=&#34;black&#34;,
    n_points=20,
    ncols=5,
    log=False,
    show=None,
    figsize=(5, 5),
):
    &#34;&#34;&#34;
    Plots rankings. [Adapted from `scanpy.plotting._anndata.ranking`]

    See, for example, how this is used in `pl.pca_ranking`.

    Parameters
    ----------

    adata : anndata.AnnData
        the data
    attr : str {&#39;var&#39;, &#39;obs&#39;, &#39;uns&#39;, &#39;varm&#39;, &#39;obsm&#39;}
        the attribute of adata that contains the score
    keys : str or list of str, optional (default=&#34;cnmf_spectra&#34;)
        scores to look up an array from the attribute of adata
    indices : list of int, optional (default=None)
        the column indices of keys for which to plot (e.g. [0,1,2] for first three
        keys)
    labels : list of str, optional (default=None)
        Labels to use for features displayed as plt.txt objects on the axes
    titles : list of str, optional (default=None)
        Labels for titles of each plot panel, in order
    ncols : int, optional (default=5)
        number of columns in gridspec
    show : bool, optional (default=None)
        show figure or just return axes
    figsize : tuple of float, optional (default=(5,5))
        size of matplotlib figure

    Returns
    -------

    matplotlib gridspec with access to the axes
    &#34;&#34;&#34;
    # default to all usages
    if indices is None:
        indices = [x for x in range(getattr(adata, attr)[keys].shape[1])]
    # get scores for each usage
    if isinstance(keys, str) and indices is not None:
        scores = np.array(getattr(adata, attr)[keys])[:, indices]
        keys = [&#34;{}_{}&#34;.format(keys, i + 1) for i in indices]
    n_panels = len(indices) if isinstance(indices, list) else 1
    if n_panels == 1:
        scores, keys = scores[:, None], [keys]
    if log:
        scores = np.log(scores)
    if labels is None:
        labels = (
            adata.var_names
            if attr in {&#34;var&#34;, &#34;varm&#34;}
            else np.arange(scores.shape[0]).astype(str)
        )
    if titles is not None:
        assert len(titles) == n_panels, &#34;Must provide {} titles&#34;.format(n_panels)
    if isinstance(labels, str):
        labels = [labels + str(i + 1) for i in range(scores.shape[0])]
    if n_panels &lt;= ncols:
        n_rows, n_cols = 1, n_panels
    else:
        n_rows, n_cols = ceil(n_panels / ncols), ncols
    fig = plt.figure(figsize=(n_cols * figsize[0], n_rows * figsize[1]))
    left, bottom = 0.1 / n_cols, 0.1 / n_rows
    gs = gridspec.GridSpec(
        nrows=n_rows,
        ncols=n_cols,
        wspace=0.1,
        left=left,
        bottom=bottom,
        right=1 - (n_cols - 1) * left - 0.01 / n_cols,
        top=1 - (n_rows - 1) * bottom - 0.1 / n_rows,
    )
    for iscore, score in enumerate(scores.T):
        plt.subplot(gs[iscore])
        indices = np.argsort(score)[::-1][: n_points + 1]
        for ig, g in enumerate(indices[::-1]):
            plt.text(
                x=score[g],
                y=ig,
                s=labels[g],
                color=color,
                verticalalignment=&#34;center&#34;,
                horizontalalignment=&#34;right&#34;,
                fontsize=&#34;medium&#34;,
                fontstyle=&#34;italic&#34;,
            )
        if titles is not None:
            plt.title(titles[iscore], fontsize=&#34;x-large&#34;)
        else:
            plt.title(keys[iscore].replace(&#34;_&#34;, &#34; &#34;), fontsize=&#34;x-large&#34;)
        plt.ylim(-0.9, ig + 0.9)
        score_min, score_max = np.min(score[indices]), np.max(score[indices])
        plt.xlim(
            (0.95 if score_min &gt; 0 else 1.05) * score_min,
            (1.05 if score_max &gt; 0 else 0.95) * score_max,
        )
        plt.xticks(rotation=45)
        plt.tick_params(labelsize=&#34;medium&#34;)
        plt.tick_params(
            axis=&#34;y&#34;,  # changes apply to the y-axis
            which=&#34;both&#34;,  # both major and minor ticks are affected
            left=False,
            right=False,
            labelleft=False,
        )
        plt.grid(False)
    gs.tight_layout(fig)
    if show == False:
        return gs


def cluster_pie(
    adata,
    pie_by=&#34;batch&#34;,
    groupby=&#34;leiden&#34;,
    ncols=5,
    show=None,
    figsize=(5, 5),
):
    &#34;&#34;&#34;
    Plots pie graphs showing makeup of cluster groups

    Parameters
    ----------

    adata : anndata.AnnData
        the data
    pie_by : str, optional (default=&#34;batch&#34;)
        adata.obs column to split pie charts by
    groupby : str, optional (default=&#34;leiden&#34;)
        adata.obs column to create pie charts for
    ncols : int, optional (default=5)
        number of columns in gridspec
    show : bool, optional (default=None)
        show figure or just return axes
    figsize : tuple of float, optional (default=(5,5))
        size of matplotlib figure

    Returns
    -------

    matplotlib gridspec with access to the axes
    &#34;&#34;&#34;
    if adata.obs[groupby].value_counts().min() == 0:
        print(
            &#34;Warning: unused categories detected in adata.obs[&#39;{}&#39;]; removing prior to building plots...&#34;.format(
                groupby
            )
        )
        adata.obs[groupby] = adata.obs[groupby].cat.remove_unused_categories()
    # get portions for each cluster
    pies = {}  # init empty dict
    for c in adata.obs[groupby].cat.categories:
        pies[c] = (
            adata.obs.loc[adata.obs[groupby] == c, pie_by].value_counts()
            / adata.obs[groupby].value_counts()[c]
        ).to_dict()
    n_panels = len(adata.obs[groupby].cat.categories)
    if n_panels &lt;= ncols:
        n_rows, n_cols = 1, n_panels
    else:
        n_rows, n_cols = ceil(n_panels / ncols), ncols
    fig = plt.figure(figsize=(n_cols * figsize[0], n_rows * figsize[1]))
    left, bottom = 0.1 / n_cols, 0.1 / n_rows
    gs = gridspec.GridSpec(
        nrows=n_rows,
        ncols=n_cols,
        wspace=0.1,
        left=left,
        bottom=bottom,
        right=1 - (n_cols - 1) * left - 0.01 / n_cols,
        top=1 - (n_rows - 1) * bottom - 0.1 / n_rows,
    )
    # get pie chart colors
    cdict = {}
    # use existing scanpy colors, if applicable
    if &#34;{}_colors&#34;.format(pie_by) in adata.uns:
        for ic, c in enumerate(adata.obs[pie_by].cat.categories):
            cdict[c] = adata.uns[&#34;{}_colors&#34;.format(pie_by)][ic]
    else:
        cmap = plt.get_cmap(&#34;tab10&#34;)
        for ic, c in enumerate(adata.obs[pie_by].cat.categories):
            cdict[c] = cmap(np.linspace(0, 1, len(adata.obs[pie_by].cat.categories)))[
                ic
            ]
    for ipie, pie in enumerate(pies.keys()):
        plt.subplot(gs[ipie])
        plt.pie(
            pies[pie].values(),
            labels=pies[pie].keys(),
            colors=[cdict[x] for x in pies[pie].keys()],
            radius=0.85,
            wedgeprops=dict(width=0.5),
            textprops={&#34;fontsize&#34;: 12},
        )
        plt.title(
            label=&#34;{}_{}&#34;.format(groupby, pie),
            loc=&#34;left&#34;,
            fontweight=&#34;bold&#34;,
            fontsize=16,
        )
    gs.tight_layout(fig)
    if show == False:
        return gs</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="kitchen.ingredients.cc_score"><code class="name flex">
<span>def <span class="ident">cc_score</span></span>(<span>adata, layer=None, seed=18, verbose=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Calculates cell cycle scores and implied phase for each observation</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing transformed and normalized (arcsinh or log1p) counts in
'layer'.</dd>
<dt><strong><code>layer</code></strong> :&ensp;<code>str</code>, optional <code>(default=None)</code></dt>
<dd>key from adata.layers to use for cc phase calculation. Default None to
use .X</dd>
<dt><strong><code>seed</code></strong> :&ensp;<code>int</code>, optional <code>(default=18)</code></dt>
<dd>random state for PCA, neighbors graph and clustering</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>print updates to console</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>adata is edited in place to add 'G2M_score', 'S_score', and 'phase' to .obs</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def cc_score(adata, layer=None, seed=18, verbose=True):
    &#34;&#34;&#34;
    Calculates cell cycle scores and implied phase for each observation

    Parameters
    ----------

    adata : anndata.AnnData
        object containing transformed and normalized (arcsinh or log1p) counts in
        &#39;layer&#39;.
    layer : str, optional (default=None)
        key from adata.layers to use for cc phase calculation. Default None to
        use .X
    seed : int, optional (default=18)
        random state for PCA, neighbors graph and clustering
    verbose : bool, optional (default=True)
        print updates to console

    Returns
    -------

    adata is edited in place to add &#39;G2M_score&#39;, &#39;S_score&#39;, and &#39;phase&#39; to .obs
    &#34;&#34;&#34;
    if layer is not None:
        adata.layers[&#34;temp&#34;] = adata.X.copy()
        adata.X = adata.layers[layer].copy()
        if verbose:
            print(&#34;Calculating cell cycle scores using layer: {}&#34;.format(layer))
    else:
        if verbose:
            print(&#34;Calculating cell cycle scores&#34;)
    # determine if sample is mouse or human based on gene names
    if any(item in adata.var_names for item in s_genes_h + g2m_genes_h):
        s_genes, g2m_genes = s_genes_h, g2m_genes_h
    elif any(item in adata.var_names for item in s_genes_m + g2m_genes_m):
        s_genes, g2m_genes = s_genes_m, g2m_genes_m
    # score cell cycle using scanpy function
    sc.tl.score_genes_cell_cycle(
        adata,
        s_genes=s_genes,  # defined at top of script
        g2m_genes=g2m_genes,  # defined at top of script
        random_state=seed,
    )
    if layer is not None:
        adata.X = adata.layers[&#34;temp&#34;].copy()
        del adata.layers[&#34;temp&#34;]</code></pre>
</details>
</dd>
<dt id="kitchen.ingredients.cellranger2"><code class="name flex">
<span>def <span class="ident">cellranger2</span></span>(<span>adata, expected=1500, upper_quant=0.99, lower_prop=0.1, label='CellRanger_2', verbose=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Labels cells using "knee point" method from CellRanger 2.1</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing unfiltered counts</dd>
<dt><strong><code>expected</code></strong> :&ensp;<code>int</code>, optional <code>(default=1500)</code></dt>
<dd>estimated number of real cells expected in dataset</dd>
<dt><strong><code>upper_quant</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.99)</code></dt>
<dd>upper quantile of real cells to test</dd>
<dt><strong><code>lower_prop</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.1)</code></dt>
<dd>percentage of expected quantile to calculate total counts threshold for</dd>
<dt><strong><code>label</code></strong> :&ensp;<code>str</code>, optional <code>(default="CellRanger_2")</code></dt>
<dd>how to name .obs column containing output</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>print updates to console</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>adata edited in place to add .obs[label] binary label</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def cellranger2(
    adata,
    expected=1500,
    upper_quant=0.99,
    lower_prop=0.1,
    label=&#34;CellRanger_2&#34;,
    verbose=True,
):
    &#34;&#34;&#34;
    Labels cells using &#34;knee point&#34; method from CellRanger 2.1

    Parameters
    ----------

    adata : anndata.AnnData
        object containing unfiltered counts
    expected : int, optional (default=1500)
        estimated number of real cells expected in dataset
    upper_quant : float, optional (default=0.99)
        upper quantile of real cells to test
    lower_prop : float, optional (default=0.1)
        percentage of expected quantile to calculate total counts threshold for
    label : str, optional (default=&#34;CellRanger_2&#34;)
        how to name .obs column containing output
    verbose : bool, optional (default=True)
        print updates to console

    Returns
    -------

    adata edited in place to add .obs[label] binary label
    &#34;&#34;&#34;
    if &#34;total_counts&#34; not in adata.obs.columns:
        adata.obs[&#34;total_counts&#34;] = adata.X.sum(axis=1)
    tmp = np.sort(np.array(adata.obs[&#34;total_counts&#34;]))[::-1]
    thresh = np.quantile(tmp[0:expected], upper_quant) * lower_prop
    adata.uns[&#34;{}_knee_thresh&#34;.format(label)] = thresh
    adata.obs[label] = &#34;False&#34;
    adata.obs.loc[adata.obs.total_counts &gt; thresh, label] = &#34;True&#34;
    adata.obs[label] = adata.obs[label].astype(&#34;category&#34;)
    if verbose:
        print(&#34;Detected knee point: {}&#34;.format(round(thresh, 3)))
        print(adata.obs[label].value_counts())</code></pre>
</details>
</dd>
<dt id="kitchen.ingredients.cellranger3"><code class="name flex">
<span>def <span class="ident">cellranger3</span></span>(<span>adata, init_counts=15000, min_umi_frac_of_median=0.01, min_umis_nonambient=500, max_adj_pvalue=0.01)</span>
</code></dt>
<dd>
<div class="desc"><p>Labels cells using "emptydrops" method from CellRanger 3.0</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing unfiltered counts</dd>
<dt><strong><code>init_counts</code></strong> :&ensp;<code>int</code>, optional <code>(default=15000)</code></dt>
<dd>initial total counts threshold for calling cells</dd>
<dt><strong><code>min_umi_frac_of_median</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.01)</code></dt>
<dd>minimum total counts for testing barcodes as fraction of median counts for
initially labeled cells</dd>
<dt><strong><code>min_umis_nonambient</code></strong> :&ensp;<code>float</code>, optional <code>(default=500)</code></dt>
<dd>minimum total counts for testing barcodes</dd>
<dt><strong><code>max_adj_pvalue</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.01)</code></dt>
<dd>maximum p-value for cell calling after B-H correction</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>adata edited in place to add .obs["CellRanger_3"] binary label</code></dt>
<dd>&nbsp;</dd>
<dt><code>and .obs["CellRanger_3_ll"] log-likelihoods for tested barcodes</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def cellranger3(
    adata,
    init_counts=15000,
    min_umi_frac_of_median=0.01,
    min_umis_nonambient=500,
    max_adj_pvalue=0.01,
):
    &#34;&#34;&#34;
    Labels cells using &#34;emptydrops&#34; method from CellRanger 3.0

    Parameters
    ----------

    adata : anndata.AnnData
        object containing unfiltered counts
    init_counts : int, optional (default=15000)
        initial total counts threshold for calling cells
    min_umi_frac_of_median : float, optional (default=0.01)
        minimum total counts for testing barcodes as fraction of median counts for
        initially labeled cells
    min_umis_nonambient : float, optional (default=500)
        minimum total counts for testing barcodes
    max_adj_pvalue : float, optional (default=0.01)
        maximum p-value for cell calling after B-H correction

    Returns
    -------

    adata edited in place to add .obs[&#34;CellRanger_3&#34;] binary label
    and .obs[&#34;CellRanger_3_ll&#34;] log-likelihoods for tested barcodes
    &#34;&#34;&#34;
    m = CountMatrix.from_anndata(adata)  # create emptydrops object from adata
    if &#34;total_counts&#34; not in adata.obs.columns:
        adata.obs[&#34;total_counts&#34;] = adata.X.sum(axis=1)
    # label initial cell calls above total counts threshold
    adata.obs[&#34;CellRanger_3&#34;] = False
    adata.obs.loc[adata.obs.total_counts &gt; init_counts, &#34;CellRanger_3&#34;] = True
    # call emptydrops to test for nonambient barcodes
    out = find_nonambient_barcodes(
        m,
        np.array(
            adata.obs.loc[
                adata.obs.CellRanger_3 == True,
            ].index,
            dtype=m.bcs.dtype,
        ),
        min_umi_frac_of_median=min_umi_frac_of_median,
        min_umis_nonambient=min_umis_nonambient,
        max_adj_pvalue=max_adj_pvalue,
    )
    # assign binary labels from emptydrops
    adata.obs.CellRanger_3.iloc[out[0]] = out[-1]
    adata.obs.CellRanger_3 = adata.obs.CellRanger_3.astype(str).astype(
        &#34;category&#34;
    )  # convert to category
    # assign log-likelihoods from emptydrops to .obs
    adata.obs[&#34;CellRanger_3_ll&#34;] = 0
    adata.obs.CellRanger_3_ll.iloc[out[0]] = out[1]</code></pre>
</details>
</dd>
<dt id="kitchen.ingredients.check_dir_exists"><code class="name flex">
<span>def <span class="ident">check_dir_exists</span></span>(<span>path)</span>
</code></dt>
<dd>
<div class="desc"><p>Checks if directory already exists or not and creates it if it doesn't</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>path</code></strong> :&ensp;<code>str</code></dt>
<dd>path to directory</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>tries to make directory at <code>path</code>, unless it already exists</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def check_dir_exists(path):
    &#34;&#34;&#34;
    Checks if directory already exists or not and creates it if it doesn&#39;t

    Parameters
    ----------

    path : str
        path to directory

    Returns
    -------

    tries to make directory at `path`, unless it already exists
    &#34;&#34;&#34;
    try:
        os.makedirs(path)
    except OSError as exception:
        if exception.errno != errno.EEXIST:
            raise</code></pre>
</details>
</dd>
<dt id="kitchen.ingredients.cluster_pie"><code class="name flex">
<span>def <span class="ident">cluster_pie</span></span>(<span>adata, pie_by='batch', groupby='leiden', ncols=5, show=None, figsize=(5, 5))</span>
</code></dt>
<dd>
<div class="desc"><p>Plots pie graphs showing makeup of cluster groups</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>the data</dd>
<dt><strong><code>pie_by</code></strong> :&ensp;<code>str</code>, optional <code>(default="batch")</code></dt>
<dd>adata.obs column to split pie charts by</dd>
<dt><strong><code>groupby</code></strong> :&ensp;<code>str</code>, optional <code>(default="leiden")</code></dt>
<dd>adata.obs column to create pie charts for</dd>
<dt><strong><code>ncols</code></strong> :&ensp;<code>int</code>, optional <code>(default=5)</code></dt>
<dd>number of columns in gridspec</dd>
<dt><strong><code>show</code></strong> :&ensp;<code>bool</code>, optional <code>(default=None)</code></dt>
<dd>show figure or just return axes</dd>
<dt><strong><code>figsize</code></strong> :&ensp;<code>tuple</code> of <code>float</code>, optional <code>(default=(5,5))</code></dt>
<dd>size of matplotlib figure</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>matplotlib gridspec with access to the axes</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def cluster_pie(
    adata,
    pie_by=&#34;batch&#34;,
    groupby=&#34;leiden&#34;,
    ncols=5,
    show=None,
    figsize=(5, 5),
):
    &#34;&#34;&#34;
    Plots pie graphs showing makeup of cluster groups

    Parameters
    ----------

    adata : anndata.AnnData
        the data
    pie_by : str, optional (default=&#34;batch&#34;)
        adata.obs column to split pie charts by
    groupby : str, optional (default=&#34;leiden&#34;)
        adata.obs column to create pie charts for
    ncols : int, optional (default=5)
        number of columns in gridspec
    show : bool, optional (default=None)
        show figure or just return axes
    figsize : tuple of float, optional (default=(5,5))
        size of matplotlib figure

    Returns
    -------

    matplotlib gridspec with access to the axes
    &#34;&#34;&#34;
    if adata.obs[groupby].value_counts().min() == 0:
        print(
            &#34;Warning: unused categories detected in adata.obs[&#39;{}&#39;]; removing prior to building plots...&#34;.format(
                groupby
            )
        )
        adata.obs[groupby] = adata.obs[groupby].cat.remove_unused_categories()
    # get portions for each cluster
    pies = {}  # init empty dict
    for c in adata.obs[groupby].cat.categories:
        pies[c] = (
            adata.obs.loc[adata.obs[groupby] == c, pie_by].value_counts()
            / adata.obs[groupby].value_counts()[c]
        ).to_dict()
    n_panels = len(adata.obs[groupby].cat.categories)
    if n_panels &lt;= ncols:
        n_rows, n_cols = 1, n_panels
    else:
        n_rows, n_cols = ceil(n_panels / ncols), ncols
    fig = plt.figure(figsize=(n_cols * figsize[0], n_rows * figsize[1]))
    left, bottom = 0.1 / n_cols, 0.1 / n_rows
    gs = gridspec.GridSpec(
        nrows=n_rows,
        ncols=n_cols,
        wspace=0.1,
        left=left,
        bottom=bottom,
        right=1 - (n_cols - 1) * left - 0.01 / n_cols,
        top=1 - (n_rows - 1) * bottom - 0.1 / n_rows,
    )
    # get pie chart colors
    cdict = {}
    # use existing scanpy colors, if applicable
    if &#34;{}_colors&#34;.format(pie_by) in adata.uns:
        for ic, c in enumerate(adata.obs[pie_by].cat.categories):
            cdict[c] = adata.uns[&#34;{}_colors&#34;.format(pie_by)][ic]
    else:
        cmap = plt.get_cmap(&#34;tab10&#34;)
        for ic, c in enumerate(adata.obs[pie_by].cat.categories):
            cdict[c] = cmap(np.linspace(0, 1, len(adata.obs[pie_by].cat.categories)))[
                ic
            ]
    for ipie, pie in enumerate(pies.keys()):
        plt.subplot(gs[ipie])
        plt.pie(
            pies[pie].values(),
            labels=pies[pie].keys(),
            colors=[cdict[x] for x in pies[pie].keys()],
            radius=0.85,
            wedgeprops=dict(width=0.5),
            textprops={&#34;fontsize&#34;: 12},
        )
        plt.title(
            label=&#34;{}_{}&#34;.format(groupby, pie),
            loc=&#34;left&#34;,
            fontweight=&#34;bold&#34;,
            fontsize=16,
        )
    gs.tight_layout(fig)
    if show == False:
        return gs</code></pre>
</details>
</dd>
<dt id="kitchen.ingredients.dim_reduce"><code class="name flex">
<span>def <span class="ident">dim_reduce</span></span>(<span>adata, layer=None, use_rep=None, clust_resolution=1.0, paga=True, seed=18, verbose=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Reduces dimensions of single-cell dataset using standard methods</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing preprocessed counts matrix</dd>
<dt><strong><code>layer</code></strong> :&ensp;<code>str</code>, optional <code>(default=None)</code></dt>
<dd>layer to use; default None for .X</dd>
<dt><strong><code>use_rep</code></strong> :&ensp;<code>str</code>, optional <code>(default=None)</code></dt>
<dd>.obsm key to use for neighbors graph instead of PCA;
default None, generate new PCA from layer</dd>
<dt><strong><code>clust_resolution</code></strong> :&ensp;<code>float</code>, optional <code>(default=1.0)</code></dt>
<dd>resolution as fraction on [0.0, 1.0] for leiden
clustering. default 1.0</dd>
<dt><strong><code>paga</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>run PAGA to seed UMAP embedding</dd>
<dt><strong><code>seed</code></strong> :&ensp;<code>int</code>, optional <code>(default=18)</code></dt>
<dd>random state for PCA, neighbors graph and clustering</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>print updates to console</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>adata is edited in place, adding PCA, neighbors graph, PAGA, and UMAP</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def dim_reduce(
    adata,
    layer=None,
    use_rep=None,
    clust_resolution=1.0,
    paga=True,
    seed=18,
    verbose=True,
):
    &#34;&#34;&#34;
    Reduces dimensions of single-cell dataset using standard methods

    Parameters
    ----------

    adata : anndata.AnnData
        object containing preprocessed counts matrix
    layer : str, optional (default=None)
        layer to use; default None for .X
    use_rep : str, optional (default=None)
        .obsm key to use for neighbors graph instead of PCA;
        default None, generate new PCA from layer
    clust_resolution : float, optional (default=1.0)
        resolution as fraction on [0.0, 1.0] for leiden
        clustering. default 1.0
    paga : bool, optional (default=True)
        run PAGA to seed UMAP embedding
    seed : int, optional (default=18)
        random state for PCA, neighbors graph and clustering
    verbose : bool, optional (default=True)
        print updates to console

    Returns
    -------

    adata is edited in place, adding PCA, neighbors graph, PAGA, and UMAP
    &#34;&#34;&#34;
    if use_rep is None:
        if layer is not None:
            if verbose:
                print(&#34;Using layer {} for dimension reduction&#34;.format(layer))
            adata.X = adata.layers[layer].copy()
        if verbose:
            print(
                &#34;Performing {}-component PCA and building kNN graph with {} neighbors&#34;.format(
                    &#34;50&#34; if adata.n_obs &gt;= 50 else adata.n_obs,
                    int(np.sqrt(adata.n_obs)),
                )
            )
        sc.pp.pca(
            adata,
            n_comps=50 if adata.n_obs &gt;= 50 else adata.n_obs - 1,
            random_state=seed,
        )
        sc.pp.neighbors(
            adata,
            n_neighbors=int(np.sqrt(adata.n_obs)),
            n_pcs=20 if adata.n_obs &gt;= 50 else int(0.4 * (adata.n_obs - 1)),
            random_state=seed,
        )
    else:
        if verbose:
            print(
                &#34;Building kNN graph with {} nearest neighbors using {}&#34;.format(
                    int(np.sqrt(adata.n_obs)), use_rep
                )
            )
        sc.pp.neighbors(
            adata,
            n_neighbors=int(np.sqrt(adata.n_obs)),
            use_rep=use_rep,
            random_state=seed,
        )
    if verbose:
        print(&#34;Clustering cells using Leiden algorithm &#34;, end=&#34;&#34;)
    sc.tl.leiden(adata, random_state=seed, resolution=clust_resolution)
    if verbose:
        print(
            &#34;- {} clusters identified&#34;.format(
                (len(adata.obs.leiden.cat.categories) + 1)
            )
        )
    if paga:
        if verbose:
            print(&#34;Building PAGA graph and UMAP from coordinates&#34;)
        sc.tl.paga(adata)
        sc.pl.paga(adata, show=False)
        sc.tl.umap(adata, init_pos=&#34;paga&#34;, random_state=seed)
    else:
        sc.tl.umap(adata, random_state=seed)</code></pre>
</details>
</dd>
<dt id="kitchen.ingredients.list_union"><code class="name flex">
<span>def <span class="ident">list_union</span></span>(<span>lst1, lst2)</span>
</code></dt>
<dd>
<div class="desc"><p>Combines two lists by the union of their values</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>lst1</code></strong>, <strong><code>lst2</code></strong> :&ensp;<code>list</code></dt>
<dd>lists to combine</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>final_list</code></strong> :&ensp;<code>list</code></dt>
<dd>union of values in lst1 and lst2</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def list_union(lst1, lst2):
    &#34;&#34;&#34;
    Combines two lists by the union of their values

    Parameters
    ----------

    lst1, lst2 : list
        lists to combine

    Returns
    -------

    final_list : list
        union of values in lst1 and lst2
    &#34;&#34;&#34;
    final_list = set(lst1).union(set(lst2))
    return final_list</code></pre>
</details>
</dd>
<dt id="kitchen.ingredients.plot_embedding"><code class="name flex">
<span>def <span class="ident">plot_embedding</span></span>(<span>adata, basis='X_umap', colors=None, show_clustering=True, ncols=5, n_cnmf_markers=7, figsize_scale=1.0, cmap='viridis', seed=18, save_to=None, verbose=True, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Plots reduced-dimension embeddings of single-cell dataset</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing preprocessed and dimension-reduced counts matrix</dd>
<dt><strong><code>basis</code></strong> :&ensp;<code>str</code>, optional <code>(default="X_umap")</code></dt>
<dd>key from <code>adata.obsm</code> containing embedding coordinates</dd>
<dt><strong><code>colors</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=None)</code></dt>
<dd>colors to plot; can be genes or .obs columns</dd>
<dt><strong><code>show_clustering</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>plot PAGA graph and leiden clusters on first two axes</dd>
<dt><strong><code>basis</code></strong> :&ensp;<code>str</code>, optional <code>(default="X_umap")</code></dt>
<dd>embedding to plot - key from <code>adata.obsm</code></dd>
<dt><strong><code>ncols</code></strong> :&ensp;<code>int</code>, optional <code>(default=5)</code></dt>
<dd>number of columns in gridspec</dd>
<dt><strong><code>n_cnmf_markers</code></strong> :&ensp;<code>int</code>, optional <code>(default=7)</code></dt>
<dd>number of top genes to print on cNMF plots</dd>
<dt><strong><code>figsize_scale</code></strong> :&ensp;<code>float</code>, optional <code>(default=1.0)</code></dt>
<dd>scaler for figure size. calculated using ncols to keep each panel square.
values &lt; 1.0 will compress figure, &gt; 1.0 will expand.</dd>
<dt><strong><code>cmap</code></strong> :&ensp;<code>str</code>, optional <code>(default="viridis")</code></dt>
<dd>valid color map for the plot</dd>
<dt><strong><code>seed</code></strong> :&ensp;<code>int</code>, optional <code>(default=18)</code></dt>
<dd>random state for plotting PAGA</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code>, optional <code>(default=None)</code></dt>
<dd>path to .png file for saving figure; default is plt.show()</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>print updates to console</dd>
<dt><strong><code>**kwargs</code></strong> :&ensp;<code>optional</code></dt>
<dd>args to pass to <code>sc.pl.embedding</code> (e.g. "size", "add_outline", etc.)</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>plot</code> of <code>embedding with overlays from "colors" as matplotlib gridspec object,</code></dt>
<dd>&nbsp;</dd>
</dl>
<p>unless <code>save_to</code> is not None.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_embedding(
    adata,
    basis=&#34;X_umap&#34;,
    colors=None,
    show_clustering=True,
    ncols=5,
    n_cnmf_markers=7,
    figsize_scale=1.0,
    cmap=&#34;viridis&#34;,
    seed=18,
    save_to=None,
    verbose=True,
    **kwargs,
):
    &#34;&#34;&#34;
    Plots reduced-dimension embeddings of single-cell dataset

    Parameters
    ----------
    adata : anndata.AnnData
        object containing preprocessed and dimension-reduced counts matrix
    basis : str, optional (default=&#34;X_umap&#34;)
        key from `adata.obsm` containing embedding coordinates
    colors : list of str, optional (default=None)
        colors to plot; can be genes or .obs columns
    show_clustering : bool, optional (default=True)
        plot PAGA graph and leiden clusters on first two axes
    basis : str, optional (default=&#34;X_umap&#34;)
        embedding to plot - key from `adata.obsm`
    ncols : int, optional (default=5)
        number of columns in gridspec
    n_cnmf_markers : int, optional (default=7)
        number of top genes to print on cNMF plots
    figsize_scale : float, optional (default=1.0)
        scaler for figure size. calculated using ncols to keep each panel square.
        values &lt; 1.0 will compress figure, &gt; 1.0 will expand.
    cmap : str, optional (default=&#34;viridis&#34;)
        valid color map for the plot
    seed : int, optional (default=18)
        random state for plotting PAGA
    save_to : str, optional (default=None)
        path to .png file for saving figure; default is plt.show()
    verbose : bool, optional (default=True)
        print updates to console
    **kwargs : optional
        args to pass to `sc.pl.embedding` (e.g. &#34;size&#34;, &#34;add_outline&#34;, etc.)

    Returns
    -------
    plot of embedding with overlays from &#34;colors&#34; as matplotlib gridspec object,
    unless `save_to` is not None.
    &#34;&#34;&#34;
    if isinstance(colors, str):  # force colors into list if single string
        colors = [colors]
    if &#34;paga&#34; in adata.uns:
        cluster_colors = [&#34;paga&#34;, &#34;leiden&#34;]
    else:
        cluster_colors = [&#34;leiden&#34;]
    if colors is not None:
        # get full list of things to plot on embedding
        if show_clustering:
            colors = cluster_colors + colors
            unique_colors = list_union(cluster_colors, colors)
        else:
            unique_colors = set(colors)
    else:
        # with no colors provided, plot PAGA graph and leiden clusters
        colors = cluster_colors
        unique_colors = set(colors)
    if verbose:
        print(&#34;Plotting embedding with overlays: {}&#34;.format(list(unique_colors)))
    # set up figure size based on number of plots
    n_plots = len(unique_colors)
    if n_plots &lt;= ncols:
        n_rows, n_cols = 1, n_plots
    else:
        n_rows, n_cols = ceil(n_plots / ncols), ncols
    fig = plt.figure(
        figsize=(ncols * n_cols * figsize_scale, ncols * n_rows * figsize_scale)
    )
    # arrange axes as subplots
    gs = gridspec.GridSpec(n_rows, n_cols, figure=fig)
    # add plots to axes
    i = 0
    for color in colors:
        if color in unique_colors:
            ax = plt.subplot(gs[i])
            if color.lower() == &#34;paga&#34;:
                sc.pl.paga(
                    adata,
                    ax=ax,
                    frameon=False,
                    show=False,
                    fontsize=&#34;large&#34;,
                    fontoutline=2.0,
                    node_size_scale=2.5,
                    random_state=seed,
                )
                ax.set_title(label=&#34;PAGA&#34;, loc=&#34;left&#34;, fontweight=&#34;bold&#34;, fontsize=16)
            else:
                if color in [&#34;leiden&#34;, &#34;louvain&#34;, &#34;cluster&#34;, &#34;group&#34;, &#34;cell_type&#34;]:
                    leg_loc, leg_fontsize, leg_fontoutline = &#34;on data&#34;, &#34;large&#34;, 2.0
                else:
                    leg_loc, leg_fontsize, leg_fontoutline = (
                        &#34;right margin&#34;,
                        12,
                        None,
                    )
                sc.pl.embedding(
                    adata,
                    basis=basis,
                    color=color,
                    ax=ax,
                    frameon=False,
                    show=False,
                    legend_loc=leg_loc,
                    legend_fontsize=leg_fontsize,
                    legend_fontoutline=leg_fontoutline,
                    title=&#34;&#34;,
                    color_map=cmap,
                    **kwargs,
                )
                # add top three gene loadings if cNMF
                if color.startswith(&#34;usage_&#34;):
                    y_range = (
                        adata.obsm[basis][:, 1].max() - adata.obsm[basis][:, 1].min()
                    )
                    [
                        ax.text(
                            x=adata.obsm[basis][:, 0].max(),
                            y=adata.obsm[basis][:, 1].max() - (0.06 * y_range * x),
                            s=&#34;&#34;
                            + adata.uns[&#34;cnmf_markers&#34;].loc[x, color.split(&#34;_&#34;)[1]],
                            fontsize=12,
                            fontstyle=&#34;italic&#34;,
                            color=&#34;k&#34;,
                            ha=&#34;right&#34;,
                        )
                        for x in range(n_cnmf_markers)
                    ]
            if color in adata.var_names:  # italicize title if plotting a gene
                ax.set_title(
                    label=color,
                    loc=&#34;left&#34;,
                    fontweight=&#34;bold&#34;,
                    fontsize=16,
                    fontstyle=&#34;italic&#34;,
                )
            else:
                ax.set_title(label=color, loc=&#34;left&#34;, fontweight=&#34;bold&#34;, fontsize=16)
            unique_colors.remove(color)
            i = i + 1
    fig.tight_layout()
    if save_to is not None:
        if verbose:
            print(&#34;Saving embeddings to {}&#34;.format(save_to))
        plt.savefig(save_to)
    else:
        return fig</code></pre>
</details>
</dd>
<dt id="kitchen.ingredients.plot_genes"><code class="name flex">
<span>def <span class="ident">plot_genes</span></span>(<span>adata, de_method='t-test_overestim_var', plot_type='heatmap', groupby='leiden', n_genes=5, dendrogram=True, ambient=False, cmap='Reds', save_to='de.png', verbose=True, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Calculates and plot <code>rank_genes_groups</code> results</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing preprocessed and dimension-reduced counts matrix</dd>
<dt><strong><code>de_method</code></strong> :&ensp;<code>str</code>, optional <code>(default="t-test_overestim_var")</code></dt>
<dd>one of "t-test", "t-test_overestim_var", "wilcoxon"</dd>
<dt><strong><code>plot_type</code></strong> :&ensp;<code>str</code>, optional <code>(default="heatmap")</code></dt>
<dd>one of "heatmap", "dotplot", "matrixplot"</dd>
<dt><strong><code>groupby</code></strong> :&ensp;<code>str</code>, optional <code>(default="leiden")</code></dt>
<dd>.obs key to group cells by</dd>
<dt><strong><code>n_genes</code></strong> :&ensp;<code>int</code>, optional <code>(default=5)</code></dt>
<dd>number of top genes per group to show</dd>
<dt><strong><code>dendrogram</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>show dendrogram of cluster similarity</dd>
<dt><strong><code>ambient</code></strong> :&ensp;<code>bool</code>, optional <code>(default=False)</code></dt>
<dd>include ambient genes as a group in the plot</dd>
<dt><strong><code>cmap</code></strong> :&ensp;<code>str</code>, optional <code>(default="Reds")</code></dt>
<dd>valid color map for the plot</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code>, optional <code>(default="de.png")</code></dt>
<dd>string to add to plot name using scanpy plot defaults</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>print updates to console</dd>
<dt><strong><code>**kwargs</code></strong> :&ensp;<code>optional</code></dt>
<dd>keyword args to add to sc.pl.matrixplot, sc.pl.dotplot, or sc.pl.heatmap</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>matplotlib figure</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_genes(
    adata,
    de_method=&#34;t-test_overestim_var&#34;,
    plot_type=&#34;heatmap&#34;,
    groupby=&#34;leiden&#34;,
    n_genes=5,
    dendrogram=True,
    ambient=False,
    cmap=&#34;Reds&#34;,
    save_to=&#34;de.png&#34;,
    verbose=True,
    **kwargs,
):
    &#34;&#34;&#34;
    Calculates and plot `rank_genes_groups` results

    Parameters
    ----------

    adata : anndata.AnnData
        object containing preprocessed and dimension-reduced counts matrix
    de_method : str, optional (default=&#34;t-test_overestim_var&#34;)
        one of &#34;t-test&#34;, &#34;t-test_overestim_var&#34;, &#34;wilcoxon&#34;
    plot_type : str, optional (default=&#34;heatmap&#34;)
        one of &#34;heatmap&#34;, &#34;dotplot&#34;, &#34;matrixplot&#34;
    groupby : str, optional (default=&#34;leiden&#34;)
        .obs key to group cells by
    n_genes : int, optional (default=5)
        number of top genes per group to show
    dendrogram : bool, optional (default=True)
        show dendrogram of cluster similarity
    ambient : bool, optional (default=False)
        include ambient genes as a group in the plot
    cmap : str, optional (default=&#34;Reds&#34;)
        valid color map for the plot
    save_to : str, optional (default=&#34;de.png&#34;)
        string to add to plot name using scanpy plot defaults
    verbose : bool, optional (default=True)
        print updates to console
    **kwargs : optional
        keyword args to add to sc.pl.matrixplot, sc.pl.dotplot, or sc.pl.heatmap

    Returns
    -------

    matplotlib figure
    &#34;&#34;&#34;
    if verbose:
        print(&#34;Performing differential expression analysis...&#34;)
    if de_method in [&#34;t-test&#34;, &#34;t-test_overestim_var&#34;]:
        # rank genes with t-test and B-H correction
        sc.tl.rank_genes_groups(
            adata, groupby=groupby, layer=&#34;log1p_norm&#34;, use_raw=False, method=de_method
        )
    elif de_method == &#34;wilcoxon&#34;:
        # rank genes with wilcoxon rank sum test
        sc.tl.rank_genes_groups(
            adata, groupby=groupby, layer=&#34;raw_counts&#34;, use_raw=False, method=de_method
        )
    else:
        print(
            &#34;Invalid de_method. Must be one of [&#39;t-test&#39;,&#39;t-test_overestim_var&#39;,&#39;wilcoxon&#39;].&#34;
        )
        return

    # calculate arcsinh counts for visualization
    adata.X = adata.layers[&#34;raw_counts&#34;].copy()
    sc.pp.normalize_total(adata)
    adata.X = np.arcsinh(adata.X)
    adata.layers[&#34;arcsinh&#34;] = adata.X.copy()
    adata.X = adata.layers[&#34;raw_counts&#34;].copy()  # return raw counts to .X

    # adjust rcParams
    rcParams[&#34;figure.figsize&#34;] = (4, 4)
    rcParams[&#34;figure.subplot.left&#34;] = 0.18
    rcParams[&#34;figure.subplot.right&#34;] = 0.96
    rcParams[&#34;figure.subplot.bottom&#34;] = 0.15
    rcParams[&#34;figure.subplot.top&#34;] = 0.91

    if ambient:
        # get markers manually and append ambient genes
        markers = {}
        for clu in adata.obs[groupby].unique().tolist():
            markers[clu] = [
                adata.uns[&#34;rank_genes_groups&#34;][&#34;names&#34;][x][clu] for x in range(n_genes)
            ]
        markers[&#34;ambient&#34;] = adata.var_names[adata.var.ambient].tolist()

        if plot_type == &#34;heatmap&#34;:
            myplot = sc.pl.heatmap(
                adata,
                markers,
                dendrogram=dendrogram,
                groupby=groupby,
                swap_axes=True,
                show_gene_labels=True,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                var_group_rotation=0,
                cmap=cmap,
                show=False,
                **kwargs,
            )
            myplot[&#34;heatmap_ax&#34;].set_yticklabels(
                myplot[&#34;heatmap_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
            )
        if plot_type == &#34;dotplot&#34;:
            myplot = sc.pl.dotplot(
                adata,
                markers,
                dendrogram=dendrogram,
                groupby=groupby,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                swap_axes=True,
                var_group_rotation=90,
                show=False,
                return_fig=True,
                **kwargs,
            )
            myplot.style(
                cmap=cmap, color_on=&#34;square&#34;, dot_edge_color=None, dot_edge_lw=1
            )
            myplot.get_axes()[&#34;mainplot_ax&#34;].set_yticklabels(
                myplot.get_axes()[&#34;mainplot_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
            )
        if plot_type == &#34;matrixplot&#34;:
            myplot = sc.pl.matrixplot(
                adata,
                markers,
                dendrogram=dendrogram,
                groupby=groupby,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                var_group_rotation=0,
                cmap=cmap,
                show=False,
                return_fig=True,
                **kwargs,
            )
            myplot.get_axes()[&#34;mainplot_ax&#34;].set_xticklabels(
                myplot.get_axes()[&#34;mainplot_ax&#34;].get_xticklabels(), fontstyle=&#34;italic&#34;
            )

    else:
        if plot_type == &#34;heatmap&#34;:
            myplot = sc.pl.rank_genes_groups_heatmap(
                adata,
                dendrogram=dendrogram,
                groupby=groupby,
                n_genes=n_genes,
                swap_axes=True,
                show_gene_labels=True,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                var_group_rotation=0,
                cmap=cmap,
                show=False,
                **kwargs,
            )
            myplot[&#34;heatmap_ax&#34;].set_yticklabels(
                myplot[&#34;heatmap_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
            )
        if plot_type == &#34;dotplot&#34;:
            myplot = sc.pl.rank_genes_groups_dotplot(
                adata,
                dendrogram=dendrogram,
                groupby=groupby,
                n_genes=n_genes,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                swap_axes=True,
                var_group_rotation=90,
                show=False,
                return_fig=True,
                **kwargs,
            )
            myplot.style(
                cmap=cmap, color_on=&#34;square&#34;, dot_edge_color=None, dot_edge_lw=1
            )
            myplot.get_axes()[&#34;mainplot_ax&#34;].set_yticklabels(
                myplot.get_axes()[&#34;mainplot_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
            )
        if plot_type == &#34;matrixplot&#34;:
            myplot = sc.pl.rank_genes_groups_matrixplot(
                adata,
                dendrogram=dendrogram,
                groupby=groupby,
                n_genes=n_genes,
                layer=&#34;arcsinh&#34;,
                standard_scale=&#34;var&#34;,
                var_group_rotation=0,
                cmap=cmap,
                show=False,
                return_fig=True,
                **kwargs,
            )
            myplot.get_axes()[&#34;mainplot_ax&#34;].set_xticklabels(
                myplot.get_axes()[&#34;mainplot_ax&#34;].get_xticklabels(), fontstyle=&#34;italic&#34;
            )

    if save_to is not None:
        plt.savefig(save_to, bbox_inches=&#34;tight&#34;)
    else:
        return myplot</code></pre>
</details>
</dd>
<dt id="kitchen.ingredients.plot_genes_cnmf"><code class="name flex">
<span>def <span class="ident">plot_genes_cnmf</span></span>(<span>adata, plot_type='heatmap', groupby='leiden', attr='varm', keys='cnmf_spectra', indices=None, n_genes=5, dendrogram=True, cmap='Reds', save_to='de_cnmf.png', **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Calculates and plots top cNMF gene loadings</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing preprocessed and dimension-reduced counts matrix</dd>
<dt><strong><code>plot_type</code></strong> :&ensp;<code>str</code>, optional <code>(default="heatmap")</code></dt>
<dd>one of "heatmap", "dotplot", "matrixplot"</dd>
<dt><strong><code>groupby</code></strong> :&ensp;<code>str</code>, optional <code>(default="leiden")</code></dt>
<dd>.obs key to group cells by</dd>
<dt><strong><code>attr</code></strong> :&ensp;<code>str {"var", "obs", "uns", "varm", "obsm"}</code></dt>
<dd>attribute of adata that contains the score</dd>
<dt><strong><code>keys</code></strong> :&ensp;<code>str</code> or <code>list</code> of <code>str</code>, optional <code>(default="cnmf_spectra")</code></dt>
<dd>scores to look up an array from the attribute of adata</dd>
<dt><strong><code>indices</code></strong> :&ensp;<code>list</code> of <code>int</code>, optional <code>(default=None)</code></dt>
<dd>column indices of keys for which to plot (e.g. [0,1,2] for first three keys)</dd>
<dt><strong><code>n_genes</code></strong> :&ensp;<code>int</code>, optional <code>(default=5)</code></dt>
<dd>number of top genes per group to show</dd>
<dt><strong><code>dendrogram</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>show dendrogram of cluster similarity</dd>
<dt><strong><code>cmap</code></strong> :&ensp;<code>str</code>, optional <code>(default="Reds")</code></dt>
<dd>valid color map for the plot</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code>, optional <code>(default="de.png")</code></dt>
<dd>string to add to plot name using scanpy plot defaults</dd>
<dt><strong><code>**kwargs</code></strong> :&ensp;<code>optional</code></dt>
<dd>keyword args to add to sc.pl.matrixplot, sc.pl.dotplot, or sc.pl.heatmap</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>matplotlib figure</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_genes_cnmf(
    adata,
    plot_type=&#34;heatmap&#34;,
    groupby=&#34;leiden&#34;,
    attr=&#34;varm&#34;,
    keys=&#34;cnmf_spectra&#34;,
    indices=None,
    n_genes=5,
    dendrogram=True,
    cmap=&#34;Reds&#34;,
    save_to=&#34;de_cnmf.png&#34;,
    **kwargs,
):
    &#34;&#34;&#34;
    Calculates and plots top cNMF gene loadings

    Parameters
    ----------

    adata : anndata.AnnData
        object containing preprocessed and dimension-reduced counts matrix
    plot_type : str, optional (default=&#34;heatmap&#34;)
        one of &#34;heatmap&#34;, &#34;dotplot&#34;, &#34;matrixplot&#34;
    groupby : str, optional (default=&#34;leiden&#34;)
        .obs key to group cells by
    attr : str {&#34;var&#34;, &#34;obs&#34;, &#34;uns&#34;, &#34;varm&#34;, &#34;obsm&#34;}
        attribute of adata that contains the score
    keys : str or list of str, optional (default=&#34;cnmf_spectra&#34;)
        scores to look up an array from the attribute of adata
    indices : list of int, optional (default=None)
        column indices of keys for which to plot (e.g. [0,1,2] for first three keys)
    n_genes : int, optional (default=5)
        number of top genes per group to show
    dendrogram : bool, optional (default=True)
        show dendrogram of cluster similarity
    cmap : str, optional (default=&#34;Reds&#34;)
        valid color map for the plot
    save_to : str, optional (default=&#34;de.png&#34;)
        string to add to plot name using scanpy plot defaults
    **kwargs : optional
        keyword args to add to sc.pl.matrixplot, sc.pl.dotplot, or sc.pl.heatmap

    Returns
    -------

    matplotlib figure
    &#34;&#34;&#34;
    # calculate arcsinh counts for visualization
    adata.X = adata.layers[&#34;raw_counts&#34;].copy()
    sc.pp.normalize_total(adata)
    adata.X = np.arcsinh(adata.X)
    adata.layers[&#34;arcsinh&#34;] = adata.X.copy()
    adata.X = adata.layers[&#34;raw_counts&#34;].copy()  # return raw counts to .X

    # default to all usages
    if indices is None:
        indices = [x for x in range(getattr(adata, attr)[keys].shape[1])]
    # get scores for each usage
    if isinstance(keys, str) and indices is not None:
        scores = np.array(getattr(adata, attr)[keys])[:, indices]
        keys = [&#34;{}_{}&#34;.format(keys, i + 1) for i in indices]
    labels = adata.var_names  # search all var_names for top genes based on spectra
    # get top n_genes for each spectra
    markers = {}
    for iscore, score in enumerate(scores.T):
        markers[keys[iscore]] = []
        indices = np.argsort(score)[::-1][:n_genes]
        for x in indices[::-1]:
            markers[keys[iscore]].append(labels[x])

    # adjust rcParams
    rcParams[&#34;figure.figsize&#34;] = (4, 4)
    rcParams[&#34;figure.subplot.left&#34;] = 0.18
    rcParams[&#34;figure.subplot.right&#34;] = 0.96
    rcParams[&#34;figure.subplot.bottom&#34;] = 0.15
    rcParams[&#34;figure.subplot.top&#34;] = 0.91

    if plot_type == &#34;heatmap&#34;:
        myplot = sc.pl.heatmap(
            adata,
            markers,
            dendrogram=dendrogram,
            groupby=groupby,
            swap_axes=True,
            show_gene_labels=True,
            layer=&#34;arcsinh&#34;,
            standard_scale=&#34;var&#34;,
            var_group_rotation=0,
            cmap=cmap,
            show=False,
            **kwargs,
        )
        myplot[&#34;heatmap_ax&#34;].set_yticklabels(
            myplot[&#34;heatmap_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
        )
    if plot_type == &#34;dotplot&#34;:
        myplot = sc.pl.dotplot(
            adata,
            markers,
            dendrogram=dendrogram,
            groupby=groupby,
            layer=&#34;arcsinh&#34;,
            standard_scale=&#34;var&#34;,
            swap_axes=True,
            var_group_rotation=90,
            show=False,
            return_fig=True,
            **kwargs,
        )
        myplot.style(cmap=cmap, color_on=&#34;square&#34;, dot_edge_color=None, dot_edge_lw=1)
        myplot.get_axes()[&#34;mainplot_ax&#34;].set_yticklabels(
            myplot.get_axes()[&#34;mainplot_ax&#34;].get_yticklabels(), fontstyle=&#34;italic&#34;
        )
    if plot_type == &#34;matrixplot&#34;:
        myplot = sc.pl.matrixplot(
            adata,
            markers,
            dendrogram=dendrogram,
            groupby=groupby,
            layer=&#34;arcsinh&#34;,
            standard_scale=&#34;var&#34;,
            var_group_rotation=0,
            cmap=cmap,
            show=False,
            return_fig=True,
            **kwargs,
        )
        myplot.get_axes()[&#34;mainplot_ax&#34;].set_xticklabels(
            myplot.get_axes()[&#34;mainplot_ax&#34;].get_xticklabels(), fontstyle=&#34;italic&#34;
        )

    if save_to is not None:
        plt.savefig(save_to, bbox_inches=&#34;tight&#34;)
    else:
        return myplot</code></pre>
</details>
</dd>
<dt id="kitchen.ingredients.rank_genes_cnmf"><code class="name flex">
<span>def <span class="ident">rank_genes_cnmf</span></span>(<span>adata, attr='varm', keys='cnmf_spectra', indices=None, labels=None, titles=None, color='black', n_points=20, ncols=5, log=False, show=None, figsize=(5, 5))</span>
</code></dt>
<dd>
<div class="desc"><p>Plots rankings. [Adapted from <code>scanpy.plotting._anndata.ranking</code>]</p>
<p>See, for example, how this is used in <code>pl.pca_ranking</code>.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>the data</dd>
<dt><strong><code>attr</code></strong> :&ensp;<code>str {'var', 'obs', 'uns', 'varm', 'obsm'}</code></dt>
<dd>the attribute of adata that contains the score</dd>
<dt><strong><code>keys</code></strong> :&ensp;<code>str</code> or <code>list</code> of <code>str</code>, optional <code>(default="cnmf_spectra")</code></dt>
<dd>scores to look up an array from the attribute of adata</dd>
<dt><strong><code>indices</code></strong> :&ensp;<code>list</code> of <code>int</code>, optional <code>(default=None)</code></dt>
<dd>the column indices of keys for which to plot (e.g. [0,1,2] for first three
keys)</dd>
<dt><strong><code>labels</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=None)</code></dt>
<dd>Labels to use for features displayed as plt.txt objects on the axes</dd>
<dt><strong><code>titles</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=None)</code></dt>
<dd>Labels for titles of each plot panel, in order</dd>
<dt><strong><code>ncols</code></strong> :&ensp;<code>int</code>, optional <code>(default=5)</code></dt>
<dd>number of columns in gridspec</dd>
<dt><strong><code>show</code></strong> :&ensp;<code>bool</code>, optional <code>(default=None)</code></dt>
<dd>show figure or just return axes</dd>
<dt><strong><code>figsize</code></strong> :&ensp;<code>tuple</code> of <code>float</code>, optional <code>(default=(5,5))</code></dt>
<dd>size of matplotlib figure</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>matplotlib gridspec with access to the axes</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def rank_genes_cnmf(
    adata,
    attr=&#34;varm&#34;,
    keys=&#34;cnmf_spectra&#34;,
    indices=None,
    labels=None,
    titles=None,
    color=&#34;black&#34;,
    n_points=20,
    ncols=5,
    log=False,
    show=None,
    figsize=(5, 5),
):
    &#34;&#34;&#34;
    Plots rankings. [Adapted from `scanpy.plotting._anndata.ranking`]

    See, for example, how this is used in `pl.pca_ranking`.

    Parameters
    ----------

    adata : anndata.AnnData
        the data
    attr : str {&#39;var&#39;, &#39;obs&#39;, &#39;uns&#39;, &#39;varm&#39;, &#39;obsm&#39;}
        the attribute of adata that contains the score
    keys : str or list of str, optional (default=&#34;cnmf_spectra&#34;)
        scores to look up an array from the attribute of adata
    indices : list of int, optional (default=None)
        the column indices of keys for which to plot (e.g. [0,1,2] for first three
        keys)
    labels : list of str, optional (default=None)
        Labels to use for features displayed as plt.txt objects on the axes
    titles : list of str, optional (default=None)
        Labels for titles of each plot panel, in order
    ncols : int, optional (default=5)
        number of columns in gridspec
    show : bool, optional (default=None)
        show figure or just return axes
    figsize : tuple of float, optional (default=(5,5))
        size of matplotlib figure

    Returns
    -------

    matplotlib gridspec with access to the axes
    &#34;&#34;&#34;
    # default to all usages
    if indices is None:
        indices = [x for x in range(getattr(adata, attr)[keys].shape[1])]
    # get scores for each usage
    if isinstance(keys, str) and indices is not None:
        scores = np.array(getattr(adata, attr)[keys])[:, indices]
        keys = [&#34;{}_{}&#34;.format(keys, i + 1) for i in indices]
    n_panels = len(indices) if isinstance(indices, list) else 1
    if n_panels == 1:
        scores, keys = scores[:, None], [keys]
    if log:
        scores = np.log(scores)
    if labels is None:
        labels = (
            adata.var_names
            if attr in {&#34;var&#34;, &#34;varm&#34;}
            else np.arange(scores.shape[0]).astype(str)
        )
    if titles is not None:
        assert len(titles) == n_panels, &#34;Must provide {} titles&#34;.format(n_panels)
    if isinstance(labels, str):
        labels = [labels + str(i + 1) for i in range(scores.shape[0])]
    if n_panels &lt;= ncols:
        n_rows, n_cols = 1, n_panels
    else:
        n_rows, n_cols = ceil(n_panels / ncols), ncols
    fig = plt.figure(figsize=(n_cols * figsize[0], n_rows * figsize[1]))
    left, bottom = 0.1 / n_cols, 0.1 / n_rows
    gs = gridspec.GridSpec(
        nrows=n_rows,
        ncols=n_cols,
        wspace=0.1,
        left=left,
        bottom=bottom,
        right=1 - (n_cols - 1) * left - 0.01 / n_cols,
        top=1 - (n_rows - 1) * bottom - 0.1 / n_rows,
    )
    for iscore, score in enumerate(scores.T):
        plt.subplot(gs[iscore])
        indices = np.argsort(score)[::-1][: n_points + 1]
        for ig, g in enumerate(indices[::-1]):
            plt.text(
                x=score[g],
                y=ig,
                s=labels[g],
                color=color,
                verticalalignment=&#34;center&#34;,
                horizontalalignment=&#34;right&#34;,
                fontsize=&#34;medium&#34;,
                fontstyle=&#34;italic&#34;,
            )
        if titles is not None:
            plt.title(titles[iscore], fontsize=&#34;x-large&#34;)
        else:
            plt.title(keys[iscore].replace(&#34;_&#34;, &#34; &#34;), fontsize=&#34;x-large&#34;)
        plt.ylim(-0.9, ig + 0.9)
        score_min, score_max = np.min(score[indices]), np.max(score[indices])
        plt.xlim(
            (0.95 if score_min &gt; 0 else 1.05) * score_min,
            (1.05 if score_max &gt; 0 else 0.95) * score_max,
        )
        plt.xticks(rotation=45)
        plt.tick_params(labelsize=&#34;medium&#34;)
        plt.tick_params(
            axis=&#34;y&#34;,  # changes apply to the y-axis
            which=&#34;both&#34;,  # both major and minor ticks are affected
            left=False,
            right=False,
            labelleft=False,
        )
        plt.grid(False)
    gs.tight_layout(fig)
    if show == False:
        return gs</code></pre>
</details>
</dd>
<dt id="kitchen.ingredients.subset_adata"><code class="name flex">
<span>def <span class="ident">subset_adata</span></span>(<span>adata, subset, verbose=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Subsets AnnData object on one or more .obs columns</p>
<p>Columns should contain 0/False for cells to throw out, and 1/True for cells to
keep. Keeps union of all labels provided in subset.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>the data</dd>
<dt><strong><code>subset</code></strong> :&ensp;<code>str</code> or <code>list</code> of <code>str</code></dt>
<dd>adata.obs labels to use for subsetting. Labels must be binary (0, "0", False,
"False" to toss - 1, "1", True, "True" to keep). Multiple labels will keep
intersection.</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>print updates to console</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>new anndata object as subset of <code>adata</code></dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def subset_adata(adata, subset, verbose=True):
    &#34;&#34;&#34;
    Subsets AnnData object on one or more .obs columns

    Columns should contain 0/False for cells to throw out, and 1/True for cells to
    keep. Keeps union of all labels provided in subset.

    Parameters
    ----------

    adata : anndata.AnnData
        the data
    subset : str or list of str
        adata.obs labels to use for subsetting. Labels must be binary (0, &#34;0&#34;, False,
        &#34;False&#34; to toss - 1, &#34;1&#34;, True, &#34;True&#34; to keep). Multiple labels will keep
        intersection.
    verbose : bool, optional (default=True)
        print updates to console

    Returns
    -------

    adata : anndata.AnnData
        new anndata object as subset of `adata`
    &#34;&#34;&#34;
    if verbose:
        print(&#34;Subsetting AnnData on {}&#34;.format(subset), end=&#34;&#34;)
    if isinstance(subset, str):
        subset = [subset]
    # initialize .obs column for choosing cells
    adata.obs[&#34;adata_subset_combined&#34;] = 0
    # create label as union of given subset args
    for i in range(len(subset)):
        adata.obs.loc[
            adata.obs[subset[i]].isin([&#34;True&#34;, True, 1.0, 1]), &#34;adata_subset_combined&#34;
        ] = 1
    adata = adata[adata.obs[&#34;adata_subset_combined&#34;] == 1, :].copy()
    adata.obs.drop(columns=&#34;adata_subset_combined&#34;, inplace=True)
    if verbose:
        print(&#34; - now {} cells and {} genes&#34;.format(adata.n_obs, adata.n_vars))
    return adata</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="kitchen" href="index.html">kitchen</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="two-column">
<li><code><a title="kitchen.ingredients.cc_score" href="#kitchen.ingredients.cc_score">cc_score</a></code></li>
<li><code><a title="kitchen.ingredients.cellranger2" href="#kitchen.ingredients.cellranger2">cellranger2</a></code></li>
<li><code><a title="kitchen.ingredients.cellranger3" href="#kitchen.ingredients.cellranger3">cellranger3</a></code></li>
<li><code><a title="kitchen.ingredients.check_dir_exists" href="#kitchen.ingredients.check_dir_exists">check_dir_exists</a></code></li>
<li><code><a title="kitchen.ingredients.cluster_pie" href="#kitchen.ingredients.cluster_pie">cluster_pie</a></code></li>
<li><code><a title="kitchen.ingredients.dim_reduce" href="#kitchen.ingredients.dim_reduce">dim_reduce</a></code></li>
<li><code><a title="kitchen.ingredients.list_union" href="#kitchen.ingredients.list_union">list_union</a></code></li>
<li><code><a title="kitchen.ingredients.plot_embedding" href="#kitchen.ingredients.plot_embedding">plot_embedding</a></code></li>
<li><code><a title="kitchen.ingredients.plot_genes" href="#kitchen.ingredients.plot_genes">plot_genes</a></code></li>
<li><code><a title="kitchen.ingredients.plot_genes_cnmf" href="#kitchen.ingredients.plot_genes_cnmf">plot_genes_cnmf</a></code></li>
<li><code><a title="kitchen.ingredients.rank_genes_cnmf" href="#kitchen.ingredients.rank_genes_cnmf">rank_genes_cnmf</a></code></li>
<li><code><a title="kitchen.ingredients.subset_adata" href="#kitchen.ingredients.subset_adata">subset_adata</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.2</a>.</p>
</footer>
</body>
</html>