<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>kitchen.kitchen API documentation</title>
<meta name="description" content="Manipulate .h5ad files and cook scRNA-seq data from command line" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>kitchen.kitchen</code></h1>
</header>
<section id="section-intro">
<p>Manipulate .h5ad files and cook scRNA-seq data from command line</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python"># -*- coding: utf-8 -*-
&#34;&#34;&#34;
Manipulate .h5ad files and cook scRNA-seq data from command line
&#34;&#34;&#34;
import argparse, os
import numpy as np
import pandas as pd
import scanpy as sc
import matplotlib.pyplot as plt
from scipy import sparse
from dropkick import recipe_dropkick

from .ingredients import (
    check_dir_exists,
    cellranger2,
    cellranger3,
    subset_adata,
    cc_score,
    dim_reduce,
    plot_embedding,
    plot_genes,
    plot_genes_cnmf,
    rank_genes_cnmf,
    cluster_pie,
)
from ._version import get_versions


def info(args):
    &#34;&#34;&#34;Prints information about .h5ad file to console&#34;&#34;&#34;
    print(&#34;Reading {}\n&#34;.format(args.file))
    adata = sc.read(args.file)
    print(adata, &#34;\n&#34;)
    print(&#34;.X: {} with {}\n&#34;.format(type(adata.X), adata.X.dtype))
    print(&#34;obs_names: {}&#34;.format(adata.obs_names))
    print(&#34;var_names: {}&#34;.format(adata.var_names))


def to_h5ad(args):
    &#34;&#34;&#34;Converts counts matrix from flat file (.txt, .csv) to .h5ad&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = os.path.splitext(os.path.basename(args.file))[0]
    # check to make sure it&#39;s an .h5ad file
    if os.path.splitext(args.file)[1] == &#34;.h5ad&#34;:
        raise ValueError(&#34;Input file already in .h5ad format&#34;)
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        # print information about counts, including names of cells and genes
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
        print(&#34;obs_names: {}&#34;.format(a.obs_names))
        print(&#34;var_names: {}&#34;.format(a.var_names))
    # sparsify counts slot
    if args.verbose:
        print(&#34;sparsifying counts...&#34;)
    a.X = sparse.csr_matrix(a.X, dtype=int)
    # save file as .h5ad
    if args.verbose:
        print(&#34;Writing counts to {}/{}.h5ad&#34;.format(args.outdir, name))
    check_dir_exists(args.outdir)
    a.write(&#34;{}/{}.h5ad&#34;.format(args.outdir, name), compression=&#34;gzip&#34;)
    if args.rm_flat_file:
        # remove original, noncompressed flat file
        if args.verbose:
            print(&#34;Removing {}&#34;.format(args.file))
        os.remove(args.file)


def mtx_to_h5ad(args):
    &#34;&#34;&#34;Converts .mtx files from 10x CellRanger or DropEst to .h5ad&#34;&#34;&#34;
    if args.dropest:
        if args.verbose:
            print(&#34;Reading DropEst .mtx files from {}&#34;.format(args.dir))
        mtx_file = [x for x in os.listdir(args.dir) if x.endswith(&#34;_counts.mtx&#34;)][
            0
        ]  # get name of .mtx
        # define filenames
        mtx = args.dir + &#34;/&#34; + mtx_file
        genes = args.dir + &#34;/&#34; + mtx_file.split(&#34;_counts&#34;)[0] + &#34;_features.txt&#34;
        barcodes = args.dir + &#34;/&#34; + mtx_file.split(&#34;_counts&#34;)[0] + &#34;_barcodes.txt&#34;
        # read files
        a = sc.read_mtx(mtx)  # read matrix
        a = a.T  # transpose DropEst matrix to cells x genes
        g = pd.read_csv(genes, delimiter=&#34;\t&#34;, header=None)  # read genes
        b = pd.read_csv(barcodes, delimiter=&#34;\t&#34;, header=None)  # read barcodes
        # add gene and barcode names
        a.obs_names = b[0].values
        a.var_names = g[0].values
        if args.verbose:
            print(
                &#34;Writing counts to {}/{}.h5ad - {} cells and {} genes&#34;.format(
                    args.outdir, mtx_file.split(&#34;_counts&#34;)[0], a.shape[0], a.shape[1]
                )
            )
        a.write(
            &#34;{}/{}.h5ad&#34;.format(args.outdir, mtx_file.split(&#34;_counts&#34;)[0]),
            compression=&#34;gzip&#34;,
        )
    else:
        if args.verbose:
            print(&#34;Reading 10x CellRanger .mtx files from {}&#34;.format(args.dir))
        a = sc.read_10x_mtx(args.dir)
        name = args.dir.split(&#34;_gene_bc_matrices&#34;)[
            0
        ]  # extract name from 10x directory name
        if args.verbose:
            print(
                &#34;Writing counts to {}/{}.h5ad - {} cells and {} genes&#34;.format(
                    args.outdir, name, a.shape[0], a.shape[1]
                )
            )
        a.write(&#34;{}/{}.h5ad&#34;.format(args.outdir, name), compression=&#34;gzip&#34;)


def h5ad_to_csv(args):
    &#34;&#34;&#34;Converts counts matrix from .h5ad to flat file (.txt, .csv)&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = os.path.splitext(os.path.basename(args.file))[0]
    # check to make sure it&#39;s an .h5ad file
    if os.path.splitext(args.file)[1] != &#34;.h5ad&#34;:
        raise ValueError(&#34;Input file must be in .h5ad format&#34;)
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        # print information about counts, including names of cells and genes
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    # swap to desired layer
    if args.layer is not None:
        if args.verbose:
            print(&#34;Using .layers[&#39;{}&#39;]&#34;.format(args.layer))
        a.X = a.layers[args.layer].copy()
    # check for/create output directory
    check_dir_exists(args.outdir)
    if args.separate_indices:
        if args.verbose:
            print(&#34;Writing counts to {}/{}_X.csv&#34;.format(args.outdir, name))
        if isinstance(a.X, sparse.csr.csr_matrix):
            df = pd.DataFrame.sparse.from_spmatrix(a.X)
            df.to_csv(
                &#34;{}/{}_X.csv&#34;.format(args.outdir, name),
                sep=&#34;,&#34;,
                header=False,
                index=False,
            )
        elif isinstance(a.X, np.ndarray):
            np.savetxt(&#34;{}/{}_X.csv&#34;.format(args.outdir, name), a.X, delimiter=&#34;,&#34;)
        if args.verbose:
            print(&#34;Writing obs names to {}/{}_obs.csv&#34;.format(args.outdir, name))
        pd.DataFrame(a.obs_names).to_csv(
            &#34;{}/{}_obs.csv&#34;.format(args.outdir, name), header=False, index=False
        )
        if args.verbose:
            print(&#34;Writing var names to {}/{}_var.csv&#34;.format(args.outdir, name))
        pd.DataFrame(a.var_names).to_csv(
            &#34;{}/{}_var.csv&#34;.format(args.outdir, name), header=False, index=False
        )
    else:
        if isinstance(a.X, sparse.csr.csr_matrix):
            df = pd.DataFrame.sparse.from_spmatrix(
                a.X, index=a.obs_names, columns=a.var_names
            )
        else:
            df = pd.DataFrame(a.X, index=a.obs_names, columns=a.var_names)
        # save file as .csv
        if args.verbose:
            print(&#34;Writing counts to {}/{}.csv&#34;.format(args.outdir, name))
        df.to_csv(
            &#34;{}/{}.csv&#34;.format(args.outdir, name), sep=&#34;,&#34;, header=True, index=True
        )
    if args.rm_h5ad_file:
        # remove original, h5ad file
        if args.verbose:
            print(&#34;Removing {}&#34;.format(args.file))
        os.remove(args.file)


def transpose(args):
    &#34;&#34;&#34;Transposes anndata object, replacing obs with var, and overwrites .h5ad file&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    a = sc.read(args.file)
    if args.verbose:
        print(a)
    # transpose file
    if args.verbose:
        print(&#34;transposing file and saving...&#34;)
    a = a.T
    # save file as .h5ad
    a.write(args.file, compression=&#34;gzip&#34;)


def to_sparse(args):
    &#34;&#34;&#34;Converts .X slot of anndata object to scipy CSR format, overwrites .h5ad file&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    a = sc.read(args.file)
    if args.verbose:
        print(a)
    if isinstance(a.X, sparse.csr.csr_matrix):
        print(&#34;{} already in sparse format&#34;.format(args.file))
        return
    # sparsify counts slot
    if args.verbose:
        print(&#34;sparsifying counts...&#34;)
    a.X = sparse.csr_matrix(a.X, dtype=int)
    # save file as .h5ad
    a.write(args.file, compression=&#34;gzip&#34;)


def to_dense(args):
    &#34;&#34;&#34;Converts .X slot of anndata object to numpy.matrix format, overwrites .h5ad file&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    a = sc.read(args.file)
    if args.verbose:
        print(a)
    if isinstance(a.X, np.matrix) or isinstance(a.X, np.ndarray):
        print(&#34;{} already in dense format&#34;.format(args.file))
        return
    # densify counts slot
    if args.verbose:
        print(&#34;densifying counts...&#34;)
    a.X = a.X.todense()
    a.X = a.X.astype(int)
    # save file as .h5ad
    a.write(args.file, compression=&#34;gzip&#34;)


def to_X(args):
    &#34;&#34;&#34;Swaps a matrix from .layers to .X slot of anndata object, overwrites .h5ad file&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    a = sc.read(args.file)
    if args.verbose:
        print(a)
    # swap layers
    if args.verbose:
        print(&#34;Putting .layers[&#39;{}&#39;] in .X and saving&#34;.format(args.layer))
    a.X = a.layers[args.layer].copy()
    # save file as .h5ad
    a.write(args.file, compression=&#34;gzip&#34;)


def rename_obs(args):
    &#34;&#34;&#34;Renames .obs columns in anndata object, and overwrites .h5ad file&#34;&#34;&#34;
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    adata = sc.read(args.file)
    if args.verbose:
        print(&#34;Renaming columns {} to {}&#34;.format(args.old_names, args.new_names))
    adata.obs.rename(columns=dict(zip(args.old_names, args.new_names)), inplace=True)
    adata.write(args.file, compression=&#34;gzip&#34;)


def label_info(args):
    &#34;&#34;&#34;Prints value counts for .obs labels to console&#34;&#34;&#34;
    print(&#34;Reading {}\n&#34;.format(args.file))
    adata = sc.read(args.file)
    print(adata, &#34;\n&#34;)
    for l in args.labels:
        print(&#34;{}\n{}\n&#34;.format(l, adata.obs[l].value_counts()))


def obs_to_categorical(args):
    &#34;&#34;&#34;Makes .obs label categorical dtype&#34;&#34;&#34;
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    adata = sc.read(args.file)
    for label in args.labels:
        if args.verbose:
            print(&#34;Converting .obs[&#39;{}&#39;] to categorical&#34;.format(label))
        if args.to_bool:
            adata.obs[label] = adata.obs[label].astype(bool).astype(&#34;category&#34;)
        else:
            adata.obs[label] = adata.obs[label].astype(&#34;category&#34;)
    adata.write(args.file, compression=&#34;gzip&#34;)


def add_label(args):
    &#34;&#34;&#34;
    Uses .obs_names from filtered counts matrix to add binary label to a reference
    anndata object, &#34;True&#34; = present in filt, &#34;False&#34; = not present.
    Overwrites reference .h5ad file.
    &#34;&#34;&#34;
    # read reference file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.ref_file))
    a = sc.read(args.ref_file)
    if args.verbose:
        print(&#34;\t&#34;, a)
    # read query file into anndata obj
    if args.verbose:
        print(&#34;\nReading {}&#34;.format(args.filt_file))
    b = sc.read(args.filt_file)
    if args.verbose:
        print(&#34;\t&#34;, b)
    # add .obs column to ref_file
    a.obs[args.label] = &#34;False&#34;
    a.obs.loc[b.obs_names, args.label] = &#34;True&#34;
    if args.verbose:
        print(
            &#34;\nTransferring labels to {}:\n{}&#34;.format(
                args.ref_file, a.obs[args.label].value_counts()
            )
        )
    # save file as .h5ad
    if args.verbose:
        print(&#34;\nWriting counts to {}&#34;.format(args.ref_file))
    a.write(args.ref_file, compression=&#34;gzip&#34;)
    if args.rm_orig_file:
        # remove filtered file
        if args.verbose:
            print(&#34;\nRemoving {}&#34;.format(args.filt_file))
        os.remove(args.filt_file)


def knee_point(args):
    &#34;&#34;&#34;Labels cells using &#34;knee point&#34; method from CellRanger 2.1&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    # add knee_point label to anndata
    cellranger2(
        a,
        expected=args.expected,
        upper_quant=args.upper_quant,
        lower_prop=args.lower_prop,
        label=args.label,
        verbose=args.verbose,
    )
    # save file as .h5ad
    print(&#34;Writing counts to {}&#34;.format(args.file))
    a.write(args.file, compression=&#34;gzip&#34;)


def emptydrops(args):
    &#34;&#34;&#34;Labels cells using &#34;emptydrops&#34; method from CellRanger 3.0&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    # add EmptyDrops label to anndata
    cellranger3(
        a,
        init_counts=args.init_counts,
        min_umi_frac_of_median=args.min_umi_frac,
        min_umis_nonambient=args.min_umi,
        max_adj_pvalue=args.max_adj_pval,
    )
    # save file as .h5ad
    if args.verbose:
        print(&#34;Writing counts to {}&#34;.format(args.file))
    a.write(args.file, compression=&#34;gzip&#34;)


def subset(args):
    &#34;&#34;&#34;Subsets anndata object on binary .obs label(s), saves to new .h5ad file&#34;&#34;&#34;
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    a = subset_adata(a, subset=args.subset, verbose=args.verbose)
    if args.verbose:
        print(&#34;Writing subsetted counts to {}&#34;.format(args.out))
    a.write(args.out, compression=&#34;gzip&#34;)


def concatenate(args):
    &#34;&#34;&#34;Concatenates list of anndata objects in .h5ad format, keeping union of genes&#34;&#34;&#34;
    # read first file
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.files[0]))
    adata_0 = sc.read(args.files[0])
    # read the rest of the files into list
    adatas = []
    for f in args.files[1:]:
        # read file into anndata obj
        if args.verbose:
            print(&#34;Reading {}&#34;.format(f))
        adatas.append(sc.read(f))
    # concatenate all files
    if args.verbose:
        print(&#34;Concatenating files...&#34;)
    concat = adata_0.concatenate(
        adatas,
        join=&#34;outer&#34;,
        batch_categories=[os.path.splitext(os.path.basename(x))[0] for x in args.files],
        fill_value=0,
    )
    if args.verbose:
        print(
            &#34;Final shape: {} cells and {} genes&#34;.format(
                concat.shape[0], concat.shape[1]
            )
        )
    # save file as .h5ad
    if args.verbose:
        print(&#34;Writing counts to {}&#34;.format(args.out))
    concat.write(args.out, compression=&#34;gzip&#34;)


def recipe(args):
    &#34;&#34;&#34;Full automated processing of scRNA-seq data&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = [os.path.splitext(os.path.basename(args.file))[0]]
    if args.subset is not None:
        name.append(&#34;_&#34;.join(args.subset))
    if args.layer is not None:
        name.append(args.layer)
    if args.use_rep is not None:
        name.append(args.use_rep)
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    # subset anndata on .obs column if desired
    if args.subset is not None:
        a = subset_adata(a, subset=args.subset)
    if args.process:
        # switch to proper layer
        if args.layer is not None:
            if args.verbose:
                print(&#34;Using layer {} to reduce dimensions&#34;.format(args.layer))
            a.X = a.layers[args.layer].copy()
        # preprocess with dropkick recipe
        a = recipe_dropkick(
            a,
            X_final=&#34;arcsinh_norm&#34;,
            verbose=args.verbose,
            filter=True,
            min_genes=args.min_genes,
        )
        # reduce dimensions
        dim_reduce(
            a,
            use_rep=args.use_rep,
            clust_resolution=args.resolution,
            paga=args.paga,
            verbose=args.verbose,
            seed=args.seed,
        )
    # run cell cycle inference
    if args.cell_cycle:
        cc_score(a, verbose=args.verbose)
        args.colors = [&#34;phase&#34;] + args.colors
    # make sure output dir exists before saving plots
    check_dir_exists(args.outdir)
    # if there&#39;s DE to do, plot genes
    if args.diff_expr is not None:
        if isinstance(args.diff_expr, str):
            args.diff_expr = [args.diff_expr]
        for de in args.diff_expr:
            plot_genes(
                a,
                de_method=args.de_method,
                plot_type=de,
                groupby=&#34;leiden&#34;,
                n_genes=5,
                cmap=args.cmap,
                save_to=&#34;{}/{}_{}.png&#34;.format(args.outdir, de, &#34;_&#34;.join(name)),
                verbose=args.verbose,
            )
        # if there&#39;s cnmf results, plot those on a heatmap/matrix/dotplot too
        if &#34;cnmf_spectra&#34; in a.varm:
            for de in args.diff_expr:
                plot_genes_cnmf(
                    a,
                    plot_type=de,
                    groupby=&#34;leiden&#34;,
                    attr=&#34;varm&#34;,
                    keys=&#34;cnmf_spectra&#34;,
                    indices=None,
                    n_genes=5,
                    cmap=args.cmap,
                    save_to=&#34;{}/{}_cnmf_{}.png&#34;.format(args.outdir, de, &#34;_&#34;.join(name)),
                )
    # if there&#39;s a cnmf flag, try to plot loadings
    if args.cnmf:
        # check for cnmf results in anndata object
        if &#34;cnmf_spectra&#34; in a.varm:
            _ = rank_genes_cnmf(a, show=False)
            if args.verbose:
                print(
                    &#34;Saving cNMF loadings to {}/{}_cnmfspectra.png&#34;.format(
                        args.outdir, &#34;_&#34;.join(name)
                    )
                )
            plt.savefig(&#34;{}/{}_cnmfspectra.png&#34;.format(args.outdir, &#34;_&#34;.join(name)))
            if args.verbose:
                print(
                    &#34;Saving embeddings to {}/{}_embedding.png&#34;.format(
                        args.outdir, &#34;_&#34;.join(name)
                    )
                )
            # save embedding plot with cNMF loadings
            if args.colors is None:
                args.colors = []
            plot_embedding(
                a,
                colors=args.colors
                + a.obs.columns[a.obs.columns.str.startswith(&#34;usage_&#34;)].tolist(),
                show_clustering=True,
                n_cnmf_markers=args.n_cnmf_markers,
                cmap=args.cmap,
                seed=args.seed,
                save_to=&#34;{}/{}_embedding.png&#34;.format(args.outdir, &#34;_&#34;.join(name)),
                verbose=args.verbose,
                size=args.point_size,
            )
        else:
            print(
                &#34;cNMF results not detected in {}. Skipping cNMF overlay for embedding.&#34;.format(
                    args.file
                )
            )
            # save embedding plot without cNMF loadings
            if args.verbose:
                print(
                    &#34;Saving embeddings to {}/{}_embedding.png&#34;.format(
                        args.outdir, &#34;_&#34;.join(name)
                    )
                )
            plot_embedding(
                a,
                colors=args.colors,
                show_clustering=True,
                cmap=args.cmap,
                seed=args.seed,
                save_to=&#34;{}/{}_embedding.png&#34;.format(args.outdir, &#34;_&#34;.join(name)),
                verbose=args.verbose,
                size=args.point_size,
            )
    else:
        # save embedding plot
        if args.verbose:
            print(
                &#34;Saving embeddings to {}/{}_embedding.png&#34;.format(
                    args.outdir, &#34;_&#34;.join(name)
                )
            )
        plot_embedding(
            a,
            colors=args.colors,
            show_clustering=True,
            cmap=args.cmap,
            seed=args.seed,
            save_to=&#34;{}/{}_embedding.png&#34;.format(args.outdir, &#34;_&#34;.join(name)),
            verbose=args.verbose,
            size=args.point_size,
        )
    # save file as .h5ad
    if args.save_adata:
        if args.verbose:
            print(
                &#34;Saving AnnData object to to {}/{}_processed.h5ad&#34;.format(
                    args.outdir, &#34;_&#34;.join(name)
                )
            )
        a.write(
            &#34;{}/{}_processed.h5ad&#34;.format(args.outdir, &#34;_&#34;.join(name)),
            compression=&#34;gzip&#34;,
        )


def de(args):
    &#34;&#34;&#34;Performs differential expression analysis on a processed .h5ad file and plots results&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = [os.path.splitext(os.path.basename(args.file))[0]]
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    if isinstance(args.plot_type, str):
        args.plot_type = [args.plot_type]
    # perform DE analysis and plot genes
    for plot in args.plot_type:
        plot_genes(
            a,
            de_method=args.de_method,
            plot_type=plot,
            groupby=args.groupby,
            n_genes=args.n_genes,
            dendrogram=args.dendrogram,
            ambient=args.ambient,
            cmap=args.cmap,
            save_to=&#34;{}/{}_{}.png&#34;.format(args.outdir, plot, &#34;_&#34;.join(name)),
            verbose=args.verbose,
        )


def cnmf_markers(args):
    &#34;&#34;&#34;Plots heatmap/matrix/dotplot of cNMF loadings for desired groups&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = [os.path.splitext(os.path.basename(args.file))[0]]
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    if isinstance(args.plot_type, str):
        args.plot_type = [args.plot_type]
    # plot cNMF marker genes
    for plot in args.plot_type:
        plot_genes_cnmf(
            a,
            plot_type=plot,
            groupby=args.groupby,
            n_genes=args.n_genes,
            dendrogram=args.dendrogram,
            cmap=args.cmap,
            save_to=&#34;{}/{}_cnmf_{}.png&#34;.format(args.outdir, plot, &#34;_&#34;.join(name)),
        )


def pie(args):
    &#34;&#34;&#34;Plots populational pie charts for desired groups&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = [os.path.splitext(os.path.basename(args.file))[0]]
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    # generate cluster_pie plot
    os.chdir(args.outdir)  # set output directory for scanpy figures
    _ = cluster_pie(a, pie_by=args.pieby, groupby=args.groupby)
    if args.verbose:
        print(
            &#34;Saving cluster pie charts to {}/{}_pie.png&#34;.format(
                args.outdir, &#34;_&#34;.join(name)
            )
        )
    plt.savefig(&#34;{}/{}_pie.png&#34;.format(args.outdir, &#34;_&#34;.join(name)))


def main():
    parser = argparse.ArgumentParser(prog=&#34;kitchen&#34;)
    parser.add_argument(
        &#34;-V&#34;,
        &#34;--version&#34;,
        action=&#34;version&#34;,
        version=get_versions()[&#34;version&#34;],
    )
    subparsers = parser.add_subparsers()

    info_parser = subparsers.add_parser(
        &#34;info&#34;,
        help=&#34;Show information about .h5ad file&#34;,
    )
    info_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    info_parser.set_defaults(func=info)

    to_h5ad_parser = subparsers.add_parser(
        &#34;to_h5ad&#34;,
        help=&#34;Convert counts matrix to .h5ad format&#34;,
    )
    to_h5ad_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as comma or tab delimited text file&#34;,
    )
    to_h5ad_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for writing h5ad. Default &#39;./&#39;&#34;,
    )
    to_h5ad_parser.add_argument(
        &#34;-rm&#34;,
        &#34;--rm-flat-file&#34;,
        help=&#34;Remove original flat file. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    to_h5ad_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_h5ad_parser.set_defaults(func=to_h5ad)

    mtx_to_h5ad_parser = subparsers.add_parser(
        &#34;mtx_to_h5ad&#34;,
        help=&#34;Convert .mtx files from DropEst or 10x CellRanger to .h5ad format&#34;,
    )
    mtx_to_h5ad_parser.add_argument(
        &#34;dir&#34;,
        type=str,
        help=&#34;Directory containing counts matrix as .mtx and text files&#34;,
    )
    mtx_to_h5ad_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for writing h5ad. Default &#39;./&#39;&#34;,
    )
    mtx_to_h5ad_parser.add_argument(
        &#34;--dropest&#34;,
        help=&#34;Directory contains DropEst outputs, rather than 10x CellRanger outputs&#34;,
        action=&#34;store_true&#34;,
    )
    mtx_to_h5ad_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    mtx_to_h5ad_parser.set_defaults(func=mtx_to_h5ad)

    to_csv_parser = subparsers.add_parser(
        &#34;to_csv&#34;,
        help=&#34;Save .h5ad counts to .csv file(s)&#34;,
    )
    to_csv_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    to_csv_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for writing csv file(s). Default &#39;./&#39;&#34;,
    )
    to_csv_parser.add_argument(
        &#34;-l&#34;,
        &#34;--layer&#34;,
        type=str,
        default=None,
        help=&#34;Key from .layers to save. Default &#39;.X&#39;.&#34;,
    )
    to_csv_parser.add_argument(
        &#34;-s&#34;,
        &#34;--separate-indices&#34;,
        help=&#34;Save indices (.obs and .var names) to separate files&#34;,
        action=&#34;store_true&#34;,
    )
    to_csv_parser.add_argument(
        &#34;-rm&#34;,
        &#34;--rm-h5ad-file&#34;,
        help=&#34;Remove original h5ad file. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    to_csv_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_csv_parser.set_defaults(func=h5ad_to_csv)

    transpose_parser = subparsers.add_parser(
        &#34;transpose&#34;,
        help=&#34;Transpose counts matrix and save as .h5ad&#34;,
    )
    transpose_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Input counts matrix as .h5ad or tab delimited text file&#34;,
    )
    transpose_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    transpose_parser.set_defaults(func=transpose)

    to_sparse_parser = subparsers.add_parser(
        &#34;to_sparse&#34;,
        help=&#34;Convert .X slot of anndata object to scipy CSR format, overwrite .h5ad file&#34;,
    )
    to_sparse_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Input counts matrix as .h5ad or tab delimited text file&#34;,
    )
    to_sparse_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_sparse_parser.set_defaults(func=to_sparse)

    to_dense_parser = subparsers.add_parser(
        &#34;to_dense&#34;,
        help=&#34;Convert .X slot of anndata object to numpy.matrix format, overwrite .h5ad file&#34;,
    )
    to_dense_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Input counts matrix as .h5ad or tab delimited text file&#34;,
    )
    to_dense_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_dense_parser.set_defaults(func=to_dense)

    to_X_parser = subparsers.add_parser(
        &#34;to_X&#34;,
        help=&#34;Swap a matrix from .layers to .X slot of anndata object&#34;,
    )
    to_X_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    to_X_parser.add_argument(
        &#34;-l&#34;,
        &#34;--layer&#34;,
        type=str,
        required=True,
        help=&#34;Key from .layers to replace .X with&#34;,
    )
    to_X_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_X_parser.set_defaults(func=to_X)

    rename_obs_parser = subparsers.add_parser(
        &#34;rename_obs&#34;,
        help=&#34;Rename .obs columns in .h5ad file&#34;,
    )
    rename_obs_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    rename_obs_parser.add_argument(
        &#34;-o&#34;,
        &#34;--old-names&#34;,
        type=str,
        nargs=&#34;+&#34;,
        required=True,
        help=&#34;List of existing .obs column names to change&#34;,
    )
    rename_obs_parser.add_argument(
        &#34;-n&#34;,
        &#34;--new-names&#34;,
        type=str,
        nargs=&#34;+&#34;,
        required=True,
        help=&#34;List of new .obs column names corresponding to args.old_names&#34;,
    )
    rename_obs_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    rename_obs_parser.set_defaults(func=rename_obs)

    label_info_parser = subparsers.add_parser(
        &#34;label_info&#34;,
        help=&#34;Print value counts for .obs labels to console&#34;,
    )
    label_info_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    label_info_parser.add_argument(
        &#34;-l&#34;,
        &#34;--labels&#34;,
        type=str,
        nargs=&#34;+&#34;,
        required=True,
        help=&#34;List of .obs column names to print value counts for&#34;,
    )
    label_info_parser.set_defaults(func=label_info)

    to_categorical_parser = subparsers.add_parser(
        &#34;to_categorical&#34;,
        help=&#34;Make .obs label categorical dtype&#34;,
    )
    to_categorical_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    to_categorical_parser.add_argument(
        &#34;-l&#34;,
        &#34;--labels&#34;,
        type=str,
        nargs=&#34;+&#34;,
        required=True,
        help=&#34;List of .obs column names to convert to categorical dtype&#34;,
    )
    to_categorical_parser.add_argument(
        &#34;-b&#34;,
        &#34;--to-bool&#34;,
        required=False,
        help=&#34;Convert to boolean first (for columns with values of {0,1})&#34;,
        action=&#34;store_true&#34;,
    )
    to_categorical_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_categorical_parser.set_defaults(func=obs_to_categorical)

    add_label_parser = subparsers.add_parser(
        &#34;add_label&#34;,
        help=&#34;Add label to .obs of .h5ad reference file using cells in a filtered file&#34;,
    )
    add_label_parser.add_argument(
        &#34;ref_file&#34;,
        type=str,
        help=&#34;Reference counts matrix to add label to, as .h5ad file&#34;,
    )
    add_label_parser.add_argument(
        &#34;filt_file&#34;,
        type=str,
        help=&#34;Filtered counts matrix providing positive labels, as .h5ad file&#34;,
    )
    add_label_parser.add_argument(
        &#34;-l&#34;,
        &#34;--label&#34;,
        type=str,
        required=True,
        help=&#34;.obs column name to place final labels in&#34;,
    )
    add_label_parser.add_argument(
        &#34;-rm&#34;,
        &#34;--rm-orig-file&#34;,
        help=&#34;Remove filtered file. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    add_label_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    add_label_parser.set_defaults(func=add_label)

    cellranger2_parser = subparsers.add_parser(
        &#34;cellranger2&#34;,
        help=&#34;Label cells using &#39;knee point&#39; method from CellRanger 2.1&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;-e&#34;,
        &#34;--expected&#34;,
        type=int,
        nargs=&#34;?&#34;,
        default=3000,
        help=&#34;Number of expected cells in the dataset; default 3000&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;-u&#34;,
        &#34;--upper-quant&#34;,
        type=float,
        default=0.99,
        help=&#34;Upper quantile of expected cells for knee point; default 0.99&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;-lp&#34;,
        &#34;--lower-prop&#34;,
        type=float,
        default=0.1,
        help=&#34;Lower proportion of cells to set threshold at; default 0.1&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;-l&#34;,
        &#34;--label&#34;,
        type=str,
        default=&#34;CellRanger_2&#34;,
        help=&#34;.obs column name to place final labels in&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    cellranger2_parser.set_defaults(func=knee_point)

    cellranger3_parser = subparsers.add_parser(
        &#34;cellranger3&#34;,
        help=&#34;Label cells using &#39;emptydrops&#39; method from CellRanger 3.0&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;--init-counts&#34;,
        type=int,
        nargs=&#34;?&#34;,
        default=15000,
        help=&#34;Initial total counts threshold for calling cells; default 15000&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;--min-umi-frac&#34;,
        type=float,
        default=0.01,
        help=&#34;Minimum total counts for testing barcodes as fraction of median counts for initially labeled cells; default 0.01&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;--min-umi&#34;,
        type=int,
        default=500,
        help=&#34;Minimum total counts for testing barcodes; default 500&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;--max-adj-pval&#34;,
        type=float,
        default=0.01,
        help=&#34;Maximum p-value for cell calling after B-H correction; default 0.01&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    cellranger3_parser.set_defaults(func=emptydrops)

    subset_parser = subparsers.add_parser(
        &#34;subset&#34;,
        help=&#34;Subset AnnData object on one or more .obs columns&#34;,
    )
    subset_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    subset_parser.add_argument(
        &#34;-s&#34;,
        &#34;--subset&#34;,
        default=None,
        nargs=&#34;*&#34;,
        help=&#34;.obs column(s) to subset cells on&#34;,
    )
    subset_parser.add_argument(
        &#34;-o&#34;,
        &#34;--out&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;./subset.h5ad&#34;,
        help=&#34;Path to output .h5ad file. Default &#39;./subset.h5ad&#39;&#34;,
    )
    subset_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    subset_parser.set_defaults(func=subset)

    concatenate_parser = subparsers.add_parser(
        &#34;concatenate&#34;,
        help=&#34;Combine multiple .h5ad files&#34;,
    )
    concatenate_parser.add_argument(
        &#34;files&#34;, type=str, nargs=&#34;*&#34;, help=&#34;List of .h5ad files to concatenate&#34;
    )
    concatenate_parser.add_argument(
        &#34;-o&#34;,
        &#34;--out&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;./concat.h5ad&#34;,
        help=&#34;Path to output .h5ad file. Default &#39;./concat.h5ad&#39;&#34;,
    )
    concatenate_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    concatenate_parser.set_defaults(func=concatenate)

    recipe_parser = subparsers.add_parser(
        &#34;recipe&#34;,
        help=&#34;Full automated processing of scRNA-seq data from command line&#34;,
    )
    recipe_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts file as .h5ad or flat (.csv, .txt) in cells x genes format&#34;,
    )
    recipe_parser.add_argument(
        &#34;-p&#34;,
        &#34;--process&#34;,
        help=&#34;Process AnnData (PCA, PAGA, UMAP). Default False&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.add_argument(
        &#34;--min-genes&#34;,
        required=False,
        type=int,
        default=1,
        help=&#34;Minimum number of genes detected to keep cell. Default 1.&#34;,
    )
    recipe_parser.add_argument(
        &#34;-s&#34;,
        &#34;--subset&#34;,
        nargs=&#34;*&#34;,
        default=None,
        help=&#34;.obs column(s) to subset cells on before embedding&#34;,
    )
    recipe_parser.add_argument(
        &#34;-l&#34;,
        &#34;--layer&#34;,
        type=str,
        default=None,
        help=&#34;Key from .layers to use for embedding. Default &#39;.X&#39;.&#34;,
    )
    recipe_parser.add_argument(
        &#34;-ur&#34;,
        &#34;--use-rep&#34;,
        type=str,
        default=None,
        help=&#34;Key from .obsm to use for neighbors graph and embedding. Default run PCA.&#34;,
    )
    recipe_parser.add_argument(
        &#34;-r&#34;,
        &#34;--resolution&#34;,
        type=float,
        default=1.0,
        help=&#34;Resolution between 0.0 and 1.0 for Leiden clustering. Default 1.0.&#34;,
    )
    recipe_parser.add_argument(
        &#34;-cc&#34;,
        &#34;--cell-cycle&#34;,
        help=&#34;Calculate cell cycle scores. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.add_argument(
        &#34;--paga&#34;,
        help=&#34;Run PAGA to seed UMAP embedding. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.add_argument(
        &#34;-de&#34;,
        &#34;--diff_expr&#34;,
        type=str,
        nargs=&#34;*&#34;,
        default=None,
        help=&#34;Type(s) of DE gene expression plots [&#39;heatmap&#39;, &#39;dotplot&#39;, &#39;matrixplot&#39;]&#34;,
    )
    recipe_parser.add_argument(
        &#34;-m&#34;,
        &#34;--de-method&#34;,
        required=False,
        type=str,
        default=&#34;t-test_overestim_var&#34;,
        help=&#34;DE testing method if -de is not None [&#39;t-test_overestim_var&#39;, &#39;t-test&#39;, &#39;wilcoxon&#39;]&#34;,
    )
    recipe_parser.add_argument(
        &#34;-c&#34;,
        &#34;--colors&#34;,
        type=str,
        nargs=&#34;*&#34;,
        default=[],
        help=&#34;Colors to plot on embedding. Can be .obs columns or gene names.&#34;,
    )
    recipe_parser.add_argument(
        &#34;-cm&#34;,
        &#34;--cmap&#34;,
        required=False,
        type=str,
        default=&#34;viridis&#34;,
        help=&#34;Color map to use in embedding overlays and genes plot&#34;,
    )
    recipe_parser.add_argument(
        &#34;-ps&#34;,
        &#34;--point-size&#34;,
        required=False,
        type=int,
        default=None,
        help=&#34;Size of points in embedding. Default 120000/n_cells.&#34;,
    )
    recipe_parser.add_argument(
        &#34;--cnmf&#34;,
        help=&#34;Plot cNMF usages on embedding. Default False.&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.add_argument(
        &#34;--n-cnmf-markers&#34;,
        required=False,
        type=int,
        default=7,
        help=&#34;Number of top loaded genes to print on cNMF embeddings. Default 7.&#34;,
    )
    recipe_parser.add_argument(
        &#34;--seed&#34;,
        type=int,
        help=&#34;Random state for generating embeddings. Default 18&#34;,
        default=18,
    )
    recipe_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for saving plots. Default &#39;./&#39;&#34;,
    )
    recipe_parser.add_argument(
        &#34;-sa&#34;,
        &#34;--save-adata&#34;,
        help=&#34;Save updated AnnData to .h5ad file. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        help=&#34;Don&#39;t print updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.set_defaults(func=recipe)

    de_parser = subparsers.add_parser(
        &#34;de&#34;,
        help=&#34;Perform differential expression analysis on a processed .h5ad file and plot results&#34;,
    )
    de_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts file as .h5ad or flat (.csv, .txt) in cells x genes format&#34;,
    )
    de_parser.add_argument(
        &#34;-m&#34;,
        &#34;--de-method&#34;,
        required=False,
        type=str,
        default=&#34;t-test_overestim_var&#34;,
        help=&#34;DE testing method [&#39;t-test_overestim_var&#39;, &#39;t-test&#39;, &#39;wilcoxon&#39;]&#34;,
    )
    de_parser.add_argument(
        &#34;-p&#34;,
        &#34;--plot-type&#34;,
        type=str,
        nargs=&#34;*&#34;,
        default=None,
        help=&#34;Type(s) of DE gene expression plot [&#39;heatmap&#39;, &#39;dotplot&#39;, &#39;matrixplot&#39;]&#34;,
    )
    de_parser.add_argument(
        &#34;-g&#34;,
        &#34;--groupby&#34;,
        required=False,
        type=str,
        default=&#34;leiden&#34;,
        help=&#34;.obs variable to group cells by for DE analysis&#34;,
    )
    de_parser.add_argument(
        &#34;-n&#34;,
        &#34;--n-genes&#34;,
        type=int,
        default=5,
        help=&#34;Number of genes to plot per group. Default 5.&#34;,
    )
    de_parser.add_argument(
        &#34;-cm&#34;,
        &#34;--cmap&#34;,
        required=False,
        type=str,
        default=&#34;Reds&#34;,
        help=&#34;Color map to use in genes plot&#34;,
    )
    de_parser.add_argument(
        &#34;-a&#34;,
        &#34;--ambient&#34;,
        help=&#34;Include ambient genes&#34;,
        action=&#34;store_true&#34;,
    )
    de_parser.add_argument(
        &#34;-d&#34;,
        &#34;--dendrogram&#34;,
        help=&#34;Generate dendrogram of group similarities&#34;,
        action=&#34;store_true&#34;,
    )
    de_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        help=&#34;Output directory for saving plots. Default &#39;./&#39;&#34;,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
    )
    de_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        help=&#34;Don&#39;t print updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    de_parser.set_defaults(func=de)

    pie_parser = subparsers.add_parser(
        &#34;pie&#34;,
        help=&#34;Plot populational pie charts for desired groups&#34;,
    )
    pie_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts file as .h5ad or flat (.csv, .txt) in cells x genes format&#34;,
    )
    pie_parser.add_argument(
        &#34;-g&#34;,
        &#34;--groupby&#34;,
        required=False,
        type=str,
        default=&#34;leiden&#34;,
        help=&#34;.obs variable to group cells by for plotting pie charts&#34;,
    )
    pie_parser.add_argument(
        &#34;-p&#34;,
        &#34;--pieby&#34;,
        required=False,
        type=str,
        default=&#34;batch&#34;,
        help=&#34;.obs variable to group cells by for pie chart within each groupby category&#34;,
    )
    pie_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for saving plots. Default &#39;./&#39;&#34;,
    )
    pie_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        help=&#34;Don&#39;t print updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    pie_parser.set_defaults(func=pie)

    cnmf_markers_parser = subparsers.add_parser(
        &#34;cnmf_markers&#34;,
        help=&#34;Plot heatmap/matrix/dotplot of cNMF loadings for desired groups&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts file as .h5ad or flat (.csv, .txt) in cells x genes format&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-p&#34;,
        &#34;--plot-type&#34;,
        type=str,
        nargs=&#34;*&#34;,
        default=None,
        help=&#34;Type(s) of gene expression plot [&#39;heatmap&#39;, &#39;dotplot&#39;, &#39;matrixplot&#39;]&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-g&#34;,
        &#34;--groupby&#34;,
        required=False,
        type=str,
        default=&#34;leiden&#34;,
        help=&#34;.obs variable to group cells by for plotting cNMF genes&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-n&#34;,
        &#34;--n-genes&#34;,
        type=int,
        default=5,
        help=&#34;Number of genes to plot per group. Default 5.&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-cm&#34;,
        &#34;--cmap&#34;,
        required=False,
        type=str,
        default=&#34;Reds&#34;,
        help=&#34;Color map to use in genes plot&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-d&#34;,
        &#34;--dendrogram&#34;,
        help=&#34;Generate dendrogram of group similarities&#34;,
        action=&#34;store_true&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for saving plots. Default &#39;./&#39;&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        help=&#34;Don&#39;t print updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    cnmf_markers_parser.set_defaults(func=cnmf_markers)

    args = parser.parse_args()

    # if --quietly specified, reverse verbosity
    if hasattr(args, &#34;quietly&#34;):
        if args.quietly:
            args.verbose = False
            del args.quietly
        else:
            args.verbose = True
            del args.quietly

    args.func(args)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="kitchen.kitchen.add_label"><code class="name flex">
<span>def <span class="ident">add_label</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Uses .obs_names from filtered counts matrix to add binary label to a reference
anndata object, "True" = present in filt, "False" = not present.
Overwrites reference .h5ad file.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def add_label(args):
    &#34;&#34;&#34;
    Uses .obs_names from filtered counts matrix to add binary label to a reference
    anndata object, &#34;True&#34; = present in filt, &#34;False&#34; = not present.
    Overwrites reference .h5ad file.
    &#34;&#34;&#34;
    # read reference file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.ref_file))
    a = sc.read(args.ref_file)
    if args.verbose:
        print(&#34;\t&#34;, a)
    # read query file into anndata obj
    if args.verbose:
        print(&#34;\nReading {}&#34;.format(args.filt_file))
    b = sc.read(args.filt_file)
    if args.verbose:
        print(&#34;\t&#34;, b)
    # add .obs column to ref_file
    a.obs[args.label] = &#34;False&#34;
    a.obs.loc[b.obs_names, args.label] = &#34;True&#34;
    if args.verbose:
        print(
            &#34;\nTransferring labels to {}:\n{}&#34;.format(
                args.ref_file, a.obs[args.label].value_counts()
            )
        )
    # save file as .h5ad
    if args.verbose:
        print(&#34;\nWriting counts to {}&#34;.format(args.ref_file))
    a.write(args.ref_file, compression=&#34;gzip&#34;)
    if args.rm_orig_file:
        # remove filtered file
        if args.verbose:
            print(&#34;\nRemoving {}&#34;.format(args.filt_file))
        os.remove(args.filt_file)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.cnmf_markers"><code class="name flex">
<span>def <span class="ident">cnmf_markers</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Plots heatmap/matrix/dotplot of cNMF loadings for desired groups</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def cnmf_markers(args):
    &#34;&#34;&#34;Plots heatmap/matrix/dotplot of cNMF loadings for desired groups&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = [os.path.splitext(os.path.basename(args.file))[0]]
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    if isinstance(args.plot_type, str):
        args.plot_type = [args.plot_type]
    # plot cNMF marker genes
    for plot in args.plot_type:
        plot_genes_cnmf(
            a,
            plot_type=plot,
            groupby=args.groupby,
            n_genes=args.n_genes,
            dendrogram=args.dendrogram,
            cmap=args.cmap,
            save_to=&#34;{}/{}_cnmf_{}.png&#34;.format(args.outdir, plot, &#34;_&#34;.join(name)),
        )</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.concatenate"><code class="name flex">
<span>def <span class="ident">concatenate</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Concatenates list of anndata objects in .h5ad format, keeping union of genes</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def concatenate(args):
    &#34;&#34;&#34;Concatenates list of anndata objects in .h5ad format, keeping union of genes&#34;&#34;&#34;
    # read first file
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.files[0]))
    adata_0 = sc.read(args.files[0])
    # read the rest of the files into list
    adatas = []
    for f in args.files[1:]:
        # read file into anndata obj
        if args.verbose:
            print(&#34;Reading {}&#34;.format(f))
        adatas.append(sc.read(f))
    # concatenate all files
    if args.verbose:
        print(&#34;Concatenating files...&#34;)
    concat = adata_0.concatenate(
        adatas,
        join=&#34;outer&#34;,
        batch_categories=[os.path.splitext(os.path.basename(x))[0] for x in args.files],
        fill_value=0,
    )
    if args.verbose:
        print(
            &#34;Final shape: {} cells and {} genes&#34;.format(
                concat.shape[0], concat.shape[1]
            )
        )
    # save file as .h5ad
    if args.verbose:
        print(&#34;Writing counts to {}&#34;.format(args.out))
    concat.write(args.out, compression=&#34;gzip&#34;)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.de"><code class="name flex">
<span>def <span class="ident">de</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Performs differential expression analysis on a processed .h5ad file and plots results</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def de(args):
    &#34;&#34;&#34;Performs differential expression analysis on a processed .h5ad file and plots results&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = [os.path.splitext(os.path.basename(args.file))[0]]
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    if isinstance(args.plot_type, str):
        args.plot_type = [args.plot_type]
    # perform DE analysis and plot genes
    for plot in args.plot_type:
        plot_genes(
            a,
            de_method=args.de_method,
            plot_type=plot,
            groupby=args.groupby,
            n_genes=args.n_genes,
            dendrogram=args.dendrogram,
            ambient=args.ambient,
            cmap=args.cmap,
            save_to=&#34;{}/{}_{}.png&#34;.format(args.outdir, plot, &#34;_&#34;.join(name)),
            verbose=args.verbose,
        )</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.emptydrops"><code class="name flex">
<span>def <span class="ident">emptydrops</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Labels cells using "emptydrops" method from CellRanger 3.0</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def emptydrops(args):
    &#34;&#34;&#34;Labels cells using &#34;emptydrops&#34; method from CellRanger 3.0&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    # add EmptyDrops label to anndata
    cellranger3(
        a,
        init_counts=args.init_counts,
        min_umi_frac_of_median=args.min_umi_frac,
        min_umis_nonambient=args.min_umi,
        max_adj_pvalue=args.max_adj_pval,
    )
    # save file as .h5ad
    if args.verbose:
        print(&#34;Writing counts to {}&#34;.format(args.file))
    a.write(args.file, compression=&#34;gzip&#34;)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.h5ad_to_csv"><code class="name flex">
<span>def <span class="ident">h5ad_to_csv</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Converts counts matrix from .h5ad to flat file (.txt, .csv)</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def h5ad_to_csv(args):
    &#34;&#34;&#34;Converts counts matrix from .h5ad to flat file (.txt, .csv)&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = os.path.splitext(os.path.basename(args.file))[0]
    # check to make sure it&#39;s an .h5ad file
    if os.path.splitext(args.file)[1] != &#34;.h5ad&#34;:
        raise ValueError(&#34;Input file must be in .h5ad format&#34;)
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        # print information about counts, including names of cells and genes
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    # swap to desired layer
    if args.layer is not None:
        if args.verbose:
            print(&#34;Using .layers[&#39;{}&#39;]&#34;.format(args.layer))
        a.X = a.layers[args.layer].copy()
    # check for/create output directory
    check_dir_exists(args.outdir)
    if args.separate_indices:
        if args.verbose:
            print(&#34;Writing counts to {}/{}_X.csv&#34;.format(args.outdir, name))
        if isinstance(a.X, sparse.csr.csr_matrix):
            df = pd.DataFrame.sparse.from_spmatrix(a.X)
            df.to_csv(
                &#34;{}/{}_X.csv&#34;.format(args.outdir, name),
                sep=&#34;,&#34;,
                header=False,
                index=False,
            )
        elif isinstance(a.X, np.ndarray):
            np.savetxt(&#34;{}/{}_X.csv&#34;.format(args.outdir, name), a.X, delimiter=&#34;,&#34;)
        if args.verbose:
            print(&#34;Writing obs names to {}/{}_obs.csv&#34;.format(args.outdir, name))
        pd.DataFrame(a.obs_names).to_csv(
            &#34;{}/{}_obs.csv&#34;.format(args.outdir, name), header=False, index=False
        )
        if args.verbose:
            print(&#34;Writing var names to {}/{}_var.csv&#34;.format(args.outdir, name))
        pd.DataFrame(a.var_names).to_csv(
            &#34;{}/{}_var.csv&#34;.format(args.outdir, name), header=False, index=False
        )
    else:
        if isinstance(a.X, sparse.csr.csr_matrix):
            df = pd.DataFrame.sparse.from_spmatrix(
                a.X, index=a.obs_names, columns=a.var_names
            )
        else:
            df = pd.DataFrame(a.X, index=a.obs_names, columns=a.var_names)
        # save file as .csv
        if args.verbose:
            print(&#34;Writing counts to {}/{}.csv&#34;.format(args.outdir, name))
        df.to_csv(
            &#34;{}/{}.csv&#34;.format(args.outdir, name), sep=&#34;,&#34;, header=True, index=True
        )
    if args.rm_h5ad_file:
        # remove original, h5ad file
        if args.verbose:
            print(&#34;Removing {}&#34;.format(args.file))
        os.remove(args.file)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.info"><code class="name flex">
<span>def <span class="ident">info</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Prints information about .h5ad file to console</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def info(args):
    &#34;&#34;&#34;Prints information about .h5ad file to console&#34;&#34;&#34;
    print(&#34;Reading {}\n&#34;.format(args.file))
    adata = sc.read(args.file)
    print(adata, &#34;\n&#34;)
    print(&#34;.X: {} with {}\n&#34;.format(type(adata.X), adata.X.dtype))
    print(&#34;obs_names: {}&#34;.format(adata.obs_names))
    print(&#34;var_names: {}&#34;.format(adata.var_names))</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.knee_point"><code class="name flex">
<span>def <span class="ident">knee_point</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Labels cells using "knee point" method from CellRanger 2.1</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def knee_point(args):
    &#34;&#34;&#34;Labels cells using &#34;knee point&#34; method from CellRanger 2.1&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    # add knee_point label to anndata
    cellranger2(
        a,
        expected=args.expected,
        upper_quant=args.upper_quant,
        lower_prop=args.lower_prop,
        label=args.label,
        verbose=args.verbose,
    )
    # save file as .h5ad
    print(&#34;Writing counts to {}&#34;.format(args.file))
    a.write(args.file, compression=&#34;gzip&#34;)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.label_info"><code class="name flex">
<span>def <span class="ident">label_info</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Prints value counts for .obs labels to console</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def label_info(args):
    &#34;&#34;&#34;Prints value counts for .obs labels to console&#34;&#34;&#34;
    print(&#34;Reading {}\n&#34;.format(args.file))
    adata = sc.read(args.file)
    print(adata, &#34;\n&#34;)
    for l in args.labels:
        print(&#34;{}\n{}\n&#34;.format(l, adata.obs[l].value_counts()))</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.main"><code class="name flex">
<span>def <span class="ident">main</span></span>(<span>)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def main():
    parser = argparse.ArgumentParser(prog=&#34;kitchen&#34;)
    parser.add_argument(
        &#34;-V&#34;,
        &#34;--version&#34;,
        action=&#34;version&#34;,
        version=get_versions()[&#34;version&#34;],
    )
    subparsers = parser.add_subparsers()

    info_parser = subparsers.add_parser(
        &#34;info&#34;,
        help=&#34;Show information about .h5ad file&#34;,
    )
    info_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    info_parser.set_defaults(func=info)

    to_h5ad_parser = subparsers.add_parser(
        &#34;to_h5ad&#34;,
        help=&#34;Convert counts matrix to .h5ad format&#34;,
    )
    to_h5ad_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as comma or tab delimited text file&#34;,
    )
    to_h5ad_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for writing h5ad. Default &#39;./&#39;&#34;,
    )
    to_h5ad_parser.add_argument(
        &#34;-rm&#34;,
        &#34;--rm-flat-file&#34;,
        help=&#34;Remove original flat file. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    to_h5ad_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_h5ad_parser.set_defaults(func=to_h5ad)

    mtx_to_h5ad_parser = subparsers.add_parser(
        &#34;mtx_to_h5ad&#34;,
        help=&#34;Convert .mtx files from DropEst or 10x CellRanger to .h5ad format&#34;,
    )
    mtx_to_h5ad_parser.add_argument(
        &#34;dir&#34;,
        type=str,
        help=&#34;Directory containing counts matrix as .mtx and text files&#34;,
    )
    mtx_to_h5ad_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for writing h5ad. Default &#39;./&#39;&#34;,
    )
    mtx_to_h5ad_parser.add_argument(
        &#34;--dropest&#34;,
        help=&#34;Directory contains DropEst outputs, rather than 10x CellRanger outputs&#34;,
        action=&#34;store_true&#34;,
    )
    mtx_to_h5ad_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    mtx_to_h5ad_parser.set_defaults(func=mtx_to_h5ad)

    to_csv_parser = subparsers.add_parser(
        &#34;to_csv&#34;,
        help=&#34;Save .h5ad counts to .csv file(s)&#34;,
    )
    to_csv_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    to_csv_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for writing csv file(s). Default &#39;./&#39;&#34;,
    )
    to_csv_parser.add_argument(
        &#34;-l&#34;,
        &#34;--layer&#34;,
        type=str,
        default=None,
        help=&#34;Key from .layers to save. Default &#39;.X&#39;.&#34;,
    )
    to_csv_parser.add_argument(
        &#34;-s&#34;,
        &#34;--separate-indices&#34;,
        help=&#34;Save indices (.obs and .var names) to separate files&#34;,
        action=&#34;store_true&#34;,
    )
    to_csv_parser.add_argument(
        &#34;-rm&#34;,
        &#34;--rm-h5ad-file&#34;,
        help=&#34;Remove original h5ad file. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    to_csv_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_csv_parser.set_defaults(func=h5ad_to_csv)

    transpose_parser = subparsers.add_parser(
        &#34;transpose&#34;,
        help=&#34;Transpose counts matrix and save as .h5ad&#34;,
    )
    transpose_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Input counts matrix as .h5ad or tab delimited text file&#34;,
    )
    transpose_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    transpose_parser.set_defaults(func=transpose)

    to_sparse_parser = subparsers.add_parser(
        &#34;to_sparse&#34;,
        help=&#34;Convert .X slot of anndata object to scipy CSR format, overwrite .h5ad file&#34;,
    )
    to_sparse_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Input counts matrix as .h5ad or tab delimited text file&#34;,
    )
    to_sparse_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_sparse_parser.set_defaults(func=to_sparse)

    to_dense_parser = subparsers.add_parser(
        &#34;to_dense&#34;,
        help=&#34;Convert .X slot of anndata object to numpy.matrix format, overwrite .h5ad file&#34;,
    )
    to_dense_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Input counts matrix as .h5ad or tab delimited text file&#34;,
    )
    to_dense_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_dense_parser.set_defaults(func=to_dense)

    to_X_parser = subparsers.add_parser(
        &#34;to_X&#34;,
        help=&#34;Swap a matrix from .layers to .X slot of anndata object&#34;,
    )
    to_X_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    to_X_parser.add_argument(
        &#34;-l&#34;,
        &#34;--layer&#34;,
        type=str,
        required=True,
        help=&#34;Key from .layers to replace .X with&#34;,
    )
    to_X_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_X_parser.set_defaults(func=to_X)

    rename_obs_parser = subparsers.add_parser(
        &#34;rename_obs&#34;,
        help=&#34;Rename .obs columns in .h5ad file&#34;,
    )
    rename_obs_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    rename_obs_parser.add_argument(
        &#34;-o&#34;,
        &#34;--old-names&#34;,
        type=str,
        nargs=&#34;+&#34;,
        required=True,
        help=&#34;List of existing .obs column names to change&#34;,
    )
    rename_obs_parser.add_argument(
        &#34;-n&#34;,
        &#34;--new-names&#34;,
        type=str,
        nargs=&#34;+&#34;,
        required=True,
        help=&#34;List of new .obs column names corresponding to args.old_names&#34;,
    )
    rename_obs_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    rename_obs_parser.set_defaults(func=rename_obs)

    label_info_parser = subparsers.add_parser(
        &#34;label_info&#34;,
        help=&#34;Print value counts for .obs labels to console&#34;,
    )
    label_info_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    label_info_parser.add_argument(
        &#34;-l&#34;,
        &#34;--labels&#34;,
        type=str,
        nargs=&#34;+&#34;,
        required=True,
        help=&#34;List of .obs column names to print value counts for&#34;,
    )
    label_info_parser.set_defaults(func=label_info)

    to_categorical_parser = subparsers.add_parser(
        &#34;to_categorical&#34;,
        help=&#34;Make .obs label categorical dtype&#34;,
    )
    to_categorical_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    to_categorical_parser.add_argument(
        &#34;-l&#34;,
        &#34;--labels&#34;,
        type=str,
        nargs=&#34;+&#34;,
        required=True,
        help=&#34;List of .obs column names to convert to categorical dtype&#34;,
    )
    to_categorical_parser.add_argument(
        &#34;-b&#34;,
        &#34;--to-bool&#34;,
        required=False,
        help=&#34;Convert to boolean first (for columns with values of {0,1})&#34;,
        action=&#34;store_true&#34;,
    )
    to_categorical_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    to_categorical_parser.set_defaults(func=obs_to_categorical)

    add_label_parser = subparsers.add_parser(
        &#34;add_label&#34;,
        help=&#34;Add label to .obs of .h5ad reference file using cells in a filtered file&#34;,
    )
    add_label_parser.add_argument(
        &#34;ref_file&#34;,
        type=str,
        help=&#34;Reference counts matrix to add label to, as .h5ad file&#34;,
    )
    add_label_parser.add_argument(
        &#34;filt_file&#34;,
        type=str,
        help=&#34;Filtered counts matrix providing positive labels, as .h5ad file&#34;,
    )
    add_label_parser.add_argument(
        &#34;-l&#34;,
        &#34;--label&#34;,
        type=str,
        required=True,
        help=&#34;.obs column name to place final labels in&#34;,
    )
    add_label_parser.add_argument(
        &#34;-rm&#34;,
        &#34;--rm-orig-file&#34;,
        help=&#34;Remove filtered file. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    add_label_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    add_label_parser.set_defaults(func=add_label)

    cellranger2_parser = subparsers.add_parser(
        &#34;cellranger2&#34;,
        help=&#34;Label cells using &#39;knee point&#39; method from CellRanger 2.1&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;-e&#34;,
        &#34;--expected&#34;,
        type=int,
        nargs=&#34;?&#34;,
        default=3000,
        help=&#34;Number of expected cells in the dataset; default 3000&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;-u&#34;,
        &#34;--upper-quant&#34;,
        type=float,
        default=0.99,
        help=&#34;Upper quantile of expected cells for knee point; default 0.99&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;-lp&#34;,
        &#34;--lower-prop&#34;,
        type=float,
        default=0.1,
        help=&#34;Lower proportion of cells to set threshold at; default 0.1&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;-l&#34;,
        &#34;--label&#34;,
        type=str,
        default=&#34;CellRanger_2&#34;,
        help=&#34;.obs column name to place final labels in&#34;,
    )
    cellranger2_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    cellranger2_parser.set_defaults(func=knee_point)

    cellranger3_parser = subparsers.add_parser(
        &#34;cellranger3&#34;,
        help=&#34;Label cells using &#39;emptydrops&#39; method from CellRanger 3.0&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;--init-counts&#34;,
        type=int,
        nargs=&#34;?&#34;,
        default=15000,
        help=&#34;Initial total counts threshold for calling cells; default 15000&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;--min-umi-frac&#34;,
        type=float,
        default=0.01,
        help=&#34;Minimum total counts for testing barcodes as fraction of median counts for initially labeled cells; default 0.01&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;--min-umi&#34;,
        type=int,
        default=500,
        help=&#34;Minimum total counts for testing barcodes; default 500&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;--max-adj-pval&#34;,
        type=float,
        default=0.01,
        help=&#34;Maximum p-value for cell calling after B-H correction; default 0.01&#34;,
    )
    cellranger3_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    cellranger3_parser.set_defaults(func=emptydrops)

    subset_parser = subparsers.add_parser(
        &#34;subset&#34;,
        help=&#34;Subset AnnData object on one or more .obs columns&#34;,
    )
    subset_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts matrix as .h5ad file&#34;,
    )
    subset_parser.add_argument(
        &#34;-s&#34;,
        &#34;--subset&#34;,
        default=None,
        nargs=&#34;*&#34;,
        help=&#34;.obs column(s) to subset cells on&#34;,
    )
    subset_parser.add_argument(
        &#34;-o&#34;,
        &#34;--out&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;./subset.h5ad&#34;,
        help=&#34;Path to output .h5ad file. Default &#39;./subset.h5ad&#39;&#34;,
    )
    subset_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    subset_parser.set_defaults(func=subset)

    concatenate_parser = subparsers.add_parser(
        &#34;concatenate&#34;,
        help=&#34;Combine multiple .h5ad files&#34;,
    )
    concatenate_parser.add_argument(
        &#34;files&#34;, type=str, nargs=&#34;*&#34;, help=&#34;List of .h5ad files to concatenate&#34;
    )
    concatenate_parser.add_argument(
        &#34;-o&#34;,
        &#34;--out&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;./concat.h5ad&#34;,
        help=&#34;Path to output .h5ad file. Default &#39;./concat.h5ad&#39;&#34;,
    )
    concatenate_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        required=False,
        help=&#34;Run without printing processing updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    concatenate_parser.set_defaults(func=concatenate)

    recipe_parser = subparsers.add_parser(
        &#34;recipe&#34;,
        help=&#34;Full automated processing of scRNA-seq data from command line&#34;,
    )
    recipe_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts file as .h5ad or flat (.csv, .txt) in cells x genes format&#34;,
    )
    recipe_parser.add_argument(
        &#34;-p&#34;,
        &#34;--process&#34;,
        help=&#34;Process AnnData (PCA, PAGA, UMAP). Default False&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.add_argument(
        &#34;--min-genes&#34;,
        required=False,
        type=int,
        default=1,
        help=&#34;Minimum number of genes detected to keep cell. Default 1.&#34;,
    )
    recipe_parser.add_argument(
        &#34;-s&#34;,
        &#34;--subset&#34;,
        nargs=&#34;*&#34;,
        default=None,
        help=&#34;.obs column(s) to subset cells on before embedding&#34;,
    )
    recipe_parser.add_argument(
        &#34;-l&#34;,
        &#34;--layer&#34;,
        type=str,
        default=None,
        help=&#34;Key from .layers to use for embedding. Default &#39;.X&#39;.&#34;,
    )
    recipe_parser.add_argument(
        &#34;-ur&#34;,
        &#34;--use-rep&#34;,
        type=str,
        default=None,
        help=&#34;Key from .obsm to use for neighbors graph and embedding. Default run PCA.&#34;,
    )
    recipe_parser.add_argument(
        &#34;-r&#34;,
        &#34;--resolution&#34;,
        type=float,
        default=1.0,
        help=&#34;Resolution between 0.0 and 1.0 for Leiden clustering. Default 1.0.&#34;,
    )
    recipe_parser.add_argument(
        &#34;-cc&#34;,
        &#34;--cell-cycle&#34;,
        help=&#34;Calculate cell cycle scores. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.add_argument(
        &#34;--paga&#34;,
        help=&#34;Run PAGA to seed UMAP embedding. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.add_argument(
        &#34;-de&#34;,
        &#34;--diff_expr&#34;,
        type=str,
        nargs=&#34;*&#34;,
        default=None,
        help=&#34;Type(s) of DE gene expression plots [&#39;heatmap&#39;, &#39;dotplot&#39;, &#39;matrixplot&#39;]&#34;,
    )
    recipe_parser.add_argument(
        &#34;-m&#34;,
        &#34;--de-method&#34;,
        required=False,
        type=str,
        default=&#34;t-test_overestim_var&#34;,
        help=&#34;DE testing method if -de is not None [&#39;t-test_overestim_var&#39;, &#39;t-test&#39;, &#39;wilcoxon&#39;]&#34;,
    )
    recipe_parser.add_argument(
        &#34;-c&#34;,
        &#34;--colors&#34;,
        type=str,
        nargs=&#34;*&#34;,
        default=[],
        help=&#34;Colors to plot on embedding. Can be .obs columns or gene names.&#34;,
    )
    recipe_parser.add_argument(
        &#34;-cm&#34;,
        &#34;--cmap&#34;,
        required=False,
        type=str,
        default=&#34;viridis&#34;,
        help=&#34;Color map to use in embedding overlays and genes plot&#34;,
    )
    recipe_parser.add_argument(
        &#34;-ps&#34;,
        &#34;--point-size&#34;,
        required=False,
        type=int,
        default=None,
        help=&#34;Size of points in embedding. Default 120000/n_cells.&#34;,
    )
    recipe_parser.add_argument(
        &#34;--cnmf&#34;,
        help=&#34;Plot cNMF usages on embedding. Default False.&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.add_argument(
        &#34;--n-cnmf-markers&#34;,
        required=False,
        type=int,
        default=7,
        help=&#34;Number of top loaded genes to print on cNMF embeddings. Default 7.&#34;,
    )
    recipe_parser.add_argument(
        &#34;--seed&#34;,
        type=int,
        help=&#34;Random state for generating embeddings. Default 18&#34;,
        default=18,
    )
    recipe_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for saving plots. Default &#39;./&#39;&#34;,
    )
    recipe_parser.add_argument(
        &#34;-sa&#34;,
        &#34;--save-adata&#34;,
        help=&#34;Save updated AnnData to .h5ad file. Default False&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        help=&#34;Don&#39;t print updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    recipe_parser.set_defaults(func=recipe)

    de_parser = subparsers.add_parser(
        &#34;de&#34;,
        help=&#34;Perform differential expression analysis on a processed .h5ad file and plot results&#34;,
    )
    de_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts file as .h5ad or flat (.csv, .txt) in cells x genes format&#34;,
    )
    de_parser.add_argument(
        &#34;-m&#34;,
        &#34;--de-method&#34;,
        required=False,
        type=str,
        default=&#34;t-test_overestim_var&#34;,
        help=&#34;DE testing method [&#39;t-test_overestim_var&#39;, &#39;t-test&#39;, &#39;wilcoxon&#39;]&#34;,
    )
    de_parser.add_argument(
        &#34;-p&#34;,
        &#34;--plot-type&#34;,
        type=str,
        nargs=&#34;*&#34;,
        default=None,
        help=&#34;Type(s) of DE gene expression plot [&#39;heatmap&#39;, &#39;dotplot&#39;, &#39;matrixplot&#39;]&#34;,
    )
    de_parser.add_argument(
        &#34;-g&#34;,
        &#34;--groupby&#34;,
        required=False,
        type=str,
        default=&#34;leiden&#34;,
        help=&#34;.obs variable to group cells by for DE analysis&#34;,
    )
    de_parser.add_argument(
        &#34;-n&#34;,
        &#34;--n-genes&#34;,
        type=int,
        default=5,
        help=&#34;Number of genes to plot per group. Default 5.&#34;,
    )
    de_parser.add_argument(
        &#34;-cm&#34;,
        &#34;--cmap&#34;,
        required=False,
        type=str,
        default=&#34;Reds&#34;,
        help=&#34;Color map to use in genes plot&#34;,
    )
    de_parser.add_argument(
        &#34;-a&#34;,
        &#34;--ambient&#34;,
        help=&#34;Include ambient genes&#34;,
        action=&#34;store_true&#34;,
    )
    de_parser.add_argument(
        &#34;-d&#34;,
        &#34;--dendrogram&#34;,
        help=&#34;Generate dendrogram of group similarities&#34;,
        action=&#34;store_true&#34;,
    )
    de_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        help=&#34;Output directory for saving plots. Default &#39;./&#39;&#34;,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
    )
    de_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        help=&#34;Don&#39;t print updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    de_parser.set_defaults(func=de)

    pie_parser = subparsers.add_parser(
        &#34;pie&#34;,
        help=&#34;Plot populational pie charts for desired groups&#34;,
    )
    pie_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts file as .h5ad or flat (.csv, .txt) in cells x genes format&#34;,
    )
    pie_parser.add_argument(
        &#34;-g&#34;,
        &#34;--groupby&#34;,
        required=False,
        type=str,
        default=&#34;leiden&#34;,
        help=&#34;.obs variable to group cells by for plotting pie charts&#34;,
    )
    pie_parser.add_argument(
        &#34;-p&#34;,
        &#34;--pieby&#34;,
        required=False,
        type=str,
        default=&#34;batch&#34;,
        help=&#34;.obs variable to group cells by for pie chart within each groupby category&#34;,
    )
    pie_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for saving plots. Default &#39;./&#39;&#34;,
    )
    pie_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        help=&#34;Don&#39;t print updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    pie_parser.set_defaults(func=pie)

    cnmf_markers_parser = subparsers.add_parser(
        &#34;cnmf_markers&#34;,
        help=&#34;Plot heatmap/matrix/dotplot of cNMF loadings for desired groups&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;file&#34;,
        type=str,
        help=&#34;Counts file as .h5ad or flat (.csv, .txt) in cells x genes format&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-p&#34;,
        &#34;--plot-type&#34;,
        type=str,
        nargs=&#34;*&#34;,
        default=None,
        help=&#34;Type(s) of gene expression plot [&#39;heatmap&#39;, &#39;dotplot&#39;, &#39;matrixplot&#39;]&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-g&#34;,
        &#34;--groupby&#34;,
        required=False,
        type=str,
        default=&#34;leiden&#34;,
        help=&#34;.obs variable to group cells by for plotting cNMF genes&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-n&#34;,
        &#34;--n-genes&#34;,
        type=int,
        default=5,
        help=&#34;Number of genes to plot per group. Default 5.&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-cm&#34;,
        &#34;--cmap&#34;,
        required=False,
        type=str,
        default=&#34;Reds&#34;,
        help=&#34;Color map to use in genes plot&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-d&#34;,
        &#34;--dendrogram&#34;,
        help=&#34;Generate dendrogram of group similarities&#34;,
        action=&#34;store_true&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-o&#34;,
        &#34;--outdir&#34;,
        type=str,
        nargs=&#34;?&#34;,
        default=&#34;.&#34;,
        help=&#34;Output directory for saving plots. Default &#39;./&#39;&#34;,
    )
    cnmf_markers_parser.add_argument(
        &#34;-q&#34;,
        &#34;--quietly&#34;,
        help=&#34;Don&#39;t print updates to console&#34;,
        action=&#34;store_true&#34;,
    )
    cnmf_markers_parser.set_defaults(func=cnmf_markers)

    args = parser.parse_args()

    # if --quietly specified, reverse verbosity
    if hasattr(args, &#34;quietly&#34;):
        if args.quietly:
            args.verbose = False
            del args.quietly
        else:
            args.verbose = True
            del args.quietly

    args.func(args)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.mtx_to_h5ad"><code class="name flex">
<span>def <span class="ident">mtx_to_h5ad</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Converts .mtx files from 10x CellRanger or DropEst to .h5ad</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def mtx_to_h5ad(args):
    &#34;&#34;&#34;Converts .mtx files from 10x CellRanger or DropEst to .h5ad&#34;&#34;&#34;
    if args.dropest:
        if args.verbose:
            print(&#34;Reading DropEst .mtx files from {}&#34;.format(args.dir))
        mtx_file = [x for x in os.listdir(args.dir) if x.endswith(&#34;_counts.mtx&#34;)][
            0
        ]  # get name of .mtx
        # define filenames
        mtx = args.dir + &#34;/&#34; + mtx_file
        genes = args.dir + &#34;/&#34; + mtx_file.split(&#34;_counts&#34;)[0] + &#34;_features.txt&#34;
        barcodes = args.dir + &#34;/&#34; + mtx_file.split(&#34;_counts&#34;)[0] + &#34;_barcodes.txt&#34;
        # read files
        a = sc.read_mtx(mtx)  # read matrix
        a = a.T  # transpose DropEst matrix to cells x genes
        g = pd.read_csv(genes, delimiter=&#34;\t&#34;, header=None)  # read genes
        b = pd.read_csv(barcodes, delimiter=&#34;\t&#34;, header=None)  # read barcodes
        # add gene and barcode names
        a.obs_names = b[0].values
        a.var_names = g[0].values
        if args.verbose:
            print(
                &#34;Writing counts to {}/{}.h5ad - {} cells and {} genes&#34;.format(
                    args.outdir, mtx_file.split(&#34;_counts&#34;)[0], a.shape[0], a.shape[1]
                )
            )
        a.write(
            &#34;{}/{}.h5ad&#34;.format(args.outdir, mtx_file.split(&#34;_counts&#34;)[0]),
            compression=&#34;gzip&#34;,
        )
    else:
        if args.verbose:
            print(&#34;Reading 10x CellRanger .mtx files from {}&#34;.format(args.dir))
        a = sc.read_10x_mtx(args.dir)
        name = args.dir.split(&#34;_gene_bc_matrices&#34;)[
            0
        ]  # extract name from 10x directory name
        if args.verbose:
            print(
                &#34;Writing counts to {}/{}.h5ad - {} cells and {} genes&#34;.format(
                    args.outdir, name, a.shape[0], a.shape[1]
                )
            )
        a.write(&#34;{}/{}.h5ad&#34;.format(args.outdir, name), compression=&#34;gzip&#34;)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.obs_to_categorical"><code class="name flex">
<span>def <span class="ident">obs_to_categorical</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Makes .obs label categorical dtype</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def obs_to_categorical(args):
    &#34;&#34;&#34;Makes .obs label categorical dtype&#34;&#34;&#34;
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    adata = sc.read(args.file)
    for label in args.labels:
        if args.verbose:
            print(&#34;Converting .obs[&#39;{}&#39;] to categorical&#34;.format(label))
        if args.to_bool:
            adata.obs[label] = adata.obs[label].astype(bool).astype(&#34;category&#34;)
        else:
            adata.obs[label] = adata.obs[label].astype(&#34;category&#34;)
    adata.write(args.file, compression=&#34;gzip&#34;)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.pie"><code class="name flex">
<span>def <span class="ident">pie</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Plots populational pie charts for desired groups</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def pie(args):
    &#34;&#34;&#34;Plots populational pie charts for desired groups&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = [os.path.splitext(os.path.basename(args.file))[0]]
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    # generate cluster_pie plot
    os.chdir(args.outdir)  # set output directory for scanpy figures
    _ = cluster_pie(a, pie_by=args.pieby, groupby=args.groupby)
    if args.verbose:
        print(
            &#34;Saving cluster pie charts to {}/{}_pie.png&#34;.format(
                args.outdir, &#34;_&#34;.join(name)
            )
        )
    plt.savefig(&#34;{}/{}_pie.png&#34;.format(args.outdir, &#34;_&#34;.join(name)))</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.recipe"><code class="name flex">
<span>def <span class="ident">recipe</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Full automated processing of scRNA-seq data</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def recipe(args):
    &#34;&#34;&#34;Full automated processing of scRNA-seq data&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = [os.path.splitext(os.path.basename(args.file))[0]]
    if args.subset is not None:
        name.append(&#34;_&#34;.join(args.subset))
    if args.layer is not None:
        name.append(args.layer)
    if args.use_rep is not None:
        name.append(args.use_rep)
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    # subset anndata on .obs column if desired
    if args.subset is not None:
        a = subset_adata(a, subset=args.subset)
    if args.process:
        # switch to proper layer
        if args.layer is not None:
            if args.verbose:
                print(&#34;Using layer {} to reduce dimensions&#34;.format(args.layer))
            a.X = a.layers[args.layer].copy()
        # preprocess with dropkick recipe
        a = recipe_dropkick(
            a,
            X_final=&#34;arcsinh_norm&#34;,
            verbose=args.verbose,
            filter=True,
            min_genes=args.min_genes,
        )
        # reduce dimensions
        dim_reduce(
            a,
            use_rep=args.use_rep,
            clust_resolution=args.resolution,
            paga=args.paga,
            verbose=args.verbose,
            seed=args.seed,
        )
    # run cell cycle inference
    if args.cell_cycle:
        cc_score(a, verbose=args.verbose)
        args.colors = [&#34;phase&#34;] + args.colors
    # make sure output dir exists before saving plots
    check_dir_exists(args.outdir)
    # if there&#39;s DE to do, plot genes
    if args.diff_expr is not None:
        if isinstance(args.diff_expr, str):
            args.diff_expr = [args.diff_expr]
        for de in args.diff_expr:
            plot_genes(
                a,
                de_method=args.de_method,
                plot_type=de,
                groupby=&#34;leiden&#34;,
                n_genes=5,
                cmap=args.cmap,
                save_to=&#34;{}/{}_{}.png&#34;.format(args.outdir, de, &#34;_&#34;.join(name)),
                verbose=args.verbose,
            )
        # if there&#39;s cnmf results, plot those on a heatmap/matrix/dotplot too
        if &#34;cnmf_spectra&#34; in a.varm:
            for de in args.diff_expr:
                plot_genes_cnmf(
                    a,
                    plot_type=de,
                    groupby=&#34;leiden&#34;,
                    attr=&#34;varm&#34;,
                    keys=&#34;cnmf_spectra&#34;,
                    indices=None,
                    n_genes=5,
                    cmap=args.cmap,
                    save_to=&#34;{}/{}_cnmf_{}.png&#34;.format(args.outdir, de, &#34;_&#34;.join(name)),
                )
    # if there&#39;s a cnmf flag, try to plot loadings
    if args.cnmf:
        # check for cnmf results in anndata object
        if &#34;cnmf_spectra&#34; in a.varm:
            _ = rank_genes_cnmf(a, show=False)
            if args.verbose:
                print(
                    &#34;Saving cNMF loadings to {}/{}_cnmfspectra.png&#34;.format(
                        args.outdir, &#34;_&#34;.join(name)
                    )
                )
            plt.savefig(&#34;{}/{}_cnmfspectra.png&#34;.format(args.outdir, &#34;_&#34;.join(name)))
            if args.verbose:
                print(
                    &#34;Saving embeddings to {}/{}_embedding.png&#34;.format(
                        args.outdir, &#34;_&#34;.join(name)
                    )
                )
            # save embedding plot with cNMF loadings
            if args.colors is None:
                args.colors = []
            plot_embedding(
                a,
                colors=args.colors
                + a.obs.columns[a.obs.columns.str.startswith(&#34;usage_&#34;)].tolist(),
                show_clustering=True,
                n_cnmf_markers=args.n_cnmf_markers,
                cmap=args.cmap,
                seed=args.seed,
                save_to=&#34;{}/{}_embedding.png&#34;.format(args.outdir, &#34;_&#34;.join(name)),
                verbose=args.verbose,
                size=args.point_size,
            )
        else:
            print(
                &#34;cNMF results not detected in {}. Skipping cNMF overlay for embedding.&#34;.format(
                    args.file
                )
            )
            # save embedding plot without cNMF loadings
            if args.verbose:
                print(
                    &#34;Saving embeddings to {}/{}_embedding.png&#34;.format(
                        args.outdir, &#34;_&#34;.join(name)
                    )
                )
            plot_embedding(
                a,
                colors=args.colors,
                show_clustering=True,
                cmap=args.cmap,
                seed=args.seed,
                save_to=&#34;{}/{}_embedding.png&#34;.format(args.outdir, &#34;_&#34;.join(name)),
                verbose=args.verbose,
                size=args.point_size,
            )
    else:
        # save embedding plot
        if args.verbose:
            print(
                &#34;Saving embeddings to {}/{}_embedding.png&#34;.format(
                    args.outdir, &#34;_&#34;.join(name)
                )
            )
        plot_embedding(
            a,
            colors=args.colors,
            show_clustering=True,
            cmap=args.cmap,
            seed=args.seed,
            save_to=&#34;{}/{}_embedding.png&#34;.format(args.outdir, &#34;_&#34;.join(name)),
            verbose=args.verbose,
            size=args.point_size,
        )
    # save file as .h5ad
    if args.save_adata:
        if args.verbose:
            print(
                &#34;Saving AnnData object to to {}/{}_processed.h5ad&#34;.format(
                    args.outdir, &#34;_&#34;.join(name)
                )
            )
        a.write(
            &#34;{}/{}_processed.h5ad&#34;.format(args.outdir, &#34;_&#34;.join(name)),
            compression=&#34;gzip&#34;,
        )</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.rename_obs"><code class="name flex">
<span>def <span class="ident">rename_obs</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Renames .obs columns in anndata object, and overwrites .h5ad file</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def rename_obs(args):
    &#34;&#34;&#34;Renames .obs columns in anndata object, and overwrites .h5ad file&#34;&#34;&#34;
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    adata = sc.read(args.file)
    if args.verbose:
        print(&#34;Renaming columns {} to {}&#34;.format(args.old_names, args.new_names))
    adata.obs.rename(columns=dict(zip(args.old_names, args.new_names)), inplace=True)
    adata.write(args.file, compression=&#34;gzip&#34;)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.subset"><code class="name flex">
<span>def <span class="ident">subset</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Subsets anndata object on binary .obs label(s), saves to new .h5ad file</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def subset(args):
    &#34;&#34;&#34;Subsets anndata object on binary .obs label(s), saves to new .h5ad file&#34;&#34;&#34;
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
    a = subset_adata(a, subset=args.subset, verbose=args.verbose)
    if args.verbose:
        print(&#34;Writing subsetted counts to {}&#34;.format(args.out))
    a.write(args.out, compression=&#34;gzip&#34;)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.to_X"><code class="name flex">
<span>def <span class="ident">to_X</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Swaps a matrix from .layers to .X slot of anndata object, overwrites .h5ad file</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def to_X(args):
    &#34;&#34;&#34;Swaps a matrix from .layers to .X slot of anndata object, overwrites .h5ad file&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    a = sc.read(args.file)
    if args.verbose:
        print(a)
    # swap layers
    if args.verbose:
        print(&#34;Putting .layers[&#39;{}&#39;] in .X and saving&#34;.format(args.layer))
    a.X = a.layers[args.layer].copy()
    # save file as .h5ad
    a.write(args.file, compression=&#34;gzip&#34;)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.to_dense"><code class="name flex">
<span>def <span class="ident">to_dense</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Converts .X slot of anndata object to numpy.matrix format, overwrites .h5ad file</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def to_dense(args):
    &#34;&#34;&#34;Converts .X slot of anndata object to numpy.matrix format, overwrites .h5ad file&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    a = sc.read(args.file)
    if args.verbose:
        print(a)
    if isinstance(a.X, np.matrix) or isinstance(a.X, np.ndarray):
        print(&#34;{} already in dense format&#34;.format(args.file))
        return
    # densify counts slot
    if args.verbose:
        print(&#34;densifying counts...&#34;)
    a.X = a.X.todense()
    a.X = a.X.astype(int)
    # save file as .h5ad
    a.write(args.file, compression=&#34;gzip&#34;)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.to_h5ad"><code class="name flex">
<span>def <span class="ident">to_h5ad</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Converts counts matrix from flat file (.txt, .csv) to .h5ad</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def to_h5ad(args):
    &#34;&#34;&#34;Converts counts matrix from flat file (.txt, .csv) to .h5ad&#34;&#34;&#34;
    # get basename of file for writing outputs
    name = os.path.splitext(os.path.basename(args.file))[0]
    # check to make sure it&#39;s an .h5ad file
    if os.path.splitext(args.file)[1] == &#34;.h5ad&#34;:
        raise ValueError(&#34;Input file already in .h5ad format&#34;)
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file), end=&#34;&#34;)
    a = sc.read(args.file)
    if args.verbose:
        # print information about counts, including names of cells and genes
        print(&#34; - {} cells and {} genes&#34;.format(a.shape[0], a.shape[1]))
        print(&#34;obs_names: {}&#34;.format(a.obs_names))
        print(&#34;var_names: {}&#34;.format(a.var_names))
    # sparsify counts slot
    if args.verbose:
        print(&#34;sparsifying counts...&#34;)
    a.X = sparse.csr_matrix(a.X, dtype=int)
    # save file as .h5ad
    if args.verbose:
        print(&#34;Writing counts to {}/{}.h5ad&#34;.format(args.outdir, name))
    check_dir_exists(args.outdir)
    a.write(&#34;{}/{}.h5ad&#34;.format(args.outdir, name), compression=&#34;gzip&#34;)
    if args.rm_flat_file:
        # remove original, noncompressed flat file
        if args.verbose:
            print(&#34;Removing {}&#34;.format(args.file))
        os.remove(args.file)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.to_sparse"><code class="name flex">
<span>def <span class="ident">to_sparse</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Converts .X slot of anndata object to scipy CSR format, overwrites .h5ad file</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def to_sparse(args):
    &#34;&#34;&#34;Converts .X slot of anndata object to scipy CSR format, overwrites .h5ad file&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    a = sc.read(args.file)
    if args.verbose:
        print(a)
    if isinstance(a.X, sparse.csr.csr_matrix):
        print(&#34;{} already in sparse format&#34;.format(args.file))
        return
    # sparsify counts slot
    if args.verbose:
        print(&#34;sparsifying counts...&#34;)
    a.X = sparse.csr_matrix(a.X, dtype=int)
    # save file as .h5ad
    a.write(args.file, compression=&#34;gzip&#34;)</code></pre>
</details>
</dd>
<dt id="kitchen.kitchen.transpose"><code class="name flex">
<span>def <span class="ident">transpose</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Transposes anndata object, replacing obs with var, and overwrites .h5ad file</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def transpose(args):
    &#34;&#34;&#34;Transposes anndata object, replacing obs with var, and overwrites .h5ad file&#34;&#34;&#34;
    # read file into anndata obj
    if args.verbose:
        print(&#34;Reading {}&#34;.format(args.file))
    a = sc.read(args.file)
    if args.verbose:
        print(a)
    # transpose file
    if args.verbose:
        print(&#34;transposing file and saving...&#34;)
    a = a.T
    # save file as .h5ad
    a.write(args.file, compression=&#34;gzip&#34;)</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="kitchen" href="index.html">kitchen</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="two-column">
<li><code><a title="kitchen.kitchen.add_label" href="#kitchen.kitchen.add_label">add_label</a></code></li>
<li><code><a title="kitchen.kitchen.cnmf_markers" href="#kitchen.kitchen.cnmf_markers">cnmf_markers</a></code></li>
<li><code><a title="kitchen.kitchen.concatenate" href="#kitchen.kitchen.concatenate">concatenate</a></code></li>
<li><code><a title="kitchen.kitchen.de" href="#kitchen.kitchen.de">de</a></code></li>
<li><code><a title="kitchen.kitchen.emptydrops" href="#kitchen.kitchen.emptydrops">emptydrops</a></code></li>
<li><code><a title="kitchen.kitchen.h5ad_to_csv" href="#kitchen.kitchen.h5ad_to_csv">h5ad_to_csv</a></code></li>
<li><code><a title="kitchen.kitchen.info" href="#kitchen.kitchen.info">info</a></code></li>
<li><code><a title="kitchen.kitchen.knee_point" href="#kitchen.kitchen.knee_point">knee_point</a></code></li>
<li><code><a title="kitchen.kitchen.label_info" href="#kitchen.kitchen.label_info">label_info</a></code></li>
<li><code><a title="kitchen.kitchen.main" href="#kitchen.kitchen.main">main</a></code></li>
<li><code><a title="kitchen.kitchen.mtx_to_h5ad" href="#kitchen.kitchen.mtx_to_h5ad">mtx_to_h5ad</a></code></li>
<li><code><a title="kitchen.kitchen.obs_to_categorical" href="#kitchen.kitchen.obs_to_categorical">obs_to_categorical</a></code></li>
<li><code><a title="kitchen.kitchen.pie" href="#kitchen.kitchen.pie">pie</a></code></li>
<li><code><a title="kitchen.kitchen.recipe" href="#kitchen.kitchen.recipe">recipe</a></code></li>
<li><code><a title="kitchen.kitchen.rename_obs" href="#kitchen.kitchen.rename_obs">rename_obs</a></code></li>
<li><code><a title="kitchen.kitchen.subset" href="#kitchen.kitchen.subset">subset</a></code></li>
<li><code><a title="kitchen.kitchen.to_X" href="#kitchen.kitchen.to_X">to_X</a></code></li>
<li><code><a title="kitchen.kitchen.to_dense" href="#kitchen.kitchen.to_dense">to_dense</a></code></li>
<li><code><a title="kitchen.kitchen.to_h5ad" href="#kitchen.kitchen.to_h5ad">to_h5ad</a></code></li>
<li><code><a title="kitchen.kitchen.to_sparse" href="#kitchen.kitchen.to_sparse">to_sparse</a></code></li>
<li><code><a title="kitchen.kitchen.transpose" href="#kitchen.kitchen.transpose">transpose</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>