<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">
<meta name="generator" content="pdoc3 0.11.1">
<title>kitchen.recipes API documentation</title>
<meta name="description" content="Functions for processing and manipulating .h5ad objects and automated processing of
scRNA-seq data">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/13.0.0/sanitize.min.css" integrity="sha512-y1dtMcuvtTMJc1yPgEqF0ZjQbhnc/bFhyvIyVNb9Zk5mIGtqVaAB1Ttl28su8AvFMOY0EwRbAe+HCLqj6W7/KA==" crossorigin>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/13.0.0/typography.min.css" integrity="sha512-Y1DYSb995BAfxobCkKepB1BqJJTPrOp3zPL74AWFugHHmmdcvO+C48WLrUOlhGMc0QG7AE3f7gmvvcrmX2fDoA==" crossorigin>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/styles/default.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:1.5em;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:2em 0 .50em 0}h3{font-size:1.4em;margin:1.6em 0 .7em 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .2s ease-in-out}a:visited{color:#503}a:hover{color:#b62}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900;font-weight:bold}pre code{font-size:.8em;line-height:1.4em;padding:1em;display:block}code{background:#f3f3f3;font-family:"DejaVu Sans Mono",monospace;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em 1em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul ul{padding-left:1em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/highlight.min.js" integrity="sha512-D9gUyxqja7hBtkWpPWGt9wfbfaMGVt9gnyCvYa+jojwwPHLCzUm5i8rpk7vD7wNee9bA35eYIjobYPaQuKS1MQ==" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => {
hljs.configure({languages: ['bash', 'css', 'diff', 'graphql', 'ini', 'javascript', 'json', 'plaintext', 'python', 'python-repl', 'rust', 'shell', 'sql', 'typescript', 'xml', 'yaml']});
hljs.highlightAll();
})</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>kitchen.recipes</code></h1>
</header>
<section id="section-intro">
<p>Functions for processing and manipulating .h5ad objects and automated processing of
scRNA-seq data</p>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="kitchen.recipes.NMF_decoupler_ORA"><code class="name flex">
<span>def <span class="ident">NMF_decoupler_ORA</span></span>(<span>adata, net, top_n_genes=None, max_FDRpval=0.05, out_dir='./', save_prefix='', save_output=False, get_ora_df_kwargs={}, decoupler_dotplot_facet_kwargs={})</span>
</code></dt>
<dd>
<div class="desc"><p>Quickly process cNMF loadings through <code>decoupler</code> ORA and create plot for
initial look at pathways</p>
<p><code>decoupler</code> tools used:
* "ORA" for biological pathways</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>Object containing cNMF results in <code>adata.uns["cnmf_markers"]</code></dd>
<dt><strong><code>net</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>Network dataframe required to run ORA. e.g. <code>msigdb</code> where <code>msigdb</code> is a
<code>pd.DataFrame</code> from <code>dc.get_resource</code>.</dd>
<dt><strong><code>top_n_genes</code></strong> :&ensp;<code>int</code> or <code>None</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>If <code>None</code>, use entire <code>adata.uns["cnmf_markers"]</code> dataframe. If an integer,
select first <code>top_n_genes</code> rows of <code>adata.uns["cnmf_markers"]</code>.</dd>
<dt><strong><code>max_FDRpval</code></strong> :&ensp;<code>float</code> or <code>None</code>, optional <code>(default=0.05)</code></dt>
<dd>FDR p-value cutoff for plotting significant ORA pathways. If <code>None</code>, don't
filter and show top 20 terms by FDR p-value.</dd>
<dt><strong><code>out_dir</code></strong> :&ensp;<code>str</code>, optional <code>(default="./")</code></dt>
<dd>Path to directory to save plots to</dd>
<dt><strong><code>save_prefix</code></strong> :&ensp;<code>str</code>, optional <code>(default="")</code></dt>
<dd>String to prepend to output plots to make names unique</dd>
<dt><strong><code>save_output</code></strong> :&ensp;<code>bool</code>, optional <code>(default=</code>False<code>)</code></dt>
<dd>If <code>True</code>, save output dataframe to <code>out_dir/save_prefix_NMF_ORA.csv</code></dd>
<dt><strong><code>get_ora_df_kwargs</code></strong> :&ensp;<code>dict</code>, optional <code>(default={})</code></dt>
<dd>Keyword arguments to pass to <code>dc.get_ora_df</code>. e.g. <code>source</code>, <code>target</code>,
<code>n_background</code>.</dd>
<dt><strong><code>decoupler_dotplot_facet_kwargs</code></strong> :&ensp;<code>dict</code>, optional <code>(default={})</code></dt>
<dd>Keyword arguments to pass to <code>decoupler_dotplot_facet</code>. e.g. <code>top_n</code>, <code>cmap</code>,
<code>dpi</code></dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>enr_pvals</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>ORA output as dataframe</dd>
</dl>
<p>Saves <code>decoupler</code> ORA dotplots to <code>out_dir/</code>.</p></div>
</dd>
<dt id="kitchen.recipes.PCA_decoupler_GSEA"><code class="name flex">
<span>def <span class="ident">PCA_decoupler_GSEA</span></span>(<span>adata, net, n_pcs=None, max_FDRpval=0.05, out_dir='./', save_prefix='', save_output=False, plot=False, get_gsea_df_kwargs={}, decoupler_dotplot_facet_kwargs={})</span>
</code></dt>
<dd>
<div class="desc"><p>Quickly process PCA loadings through <code>decoupler</code> GSEA and create plot for
initial look at pathways</p>
<p><code>decoupler</code> tools used:
* "GSEA" for biological pathways</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>Object containing PCA results in <code>adata.varm["PCs"]</code></dd>
<dt><strong><code>net</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>Network dataframe required to run GSEA. e.g. <code>msigdb</code> where <code>msigdb</code> is a
<code>pd.DataFrame</code> from <code>dc.get_resource</code>.</dd>
<dt><strong><code>max_FDRpval</code></strong> :&ensp;<code>float</code> or <code>None</code>, optional <code>(default=0.05)</code></dt>
<dd>FDR p-value cutoff for plotting significant ORA pathways. If <code>None</code>, don't
filter and show top 20 terms by FDR p-value.</dd>
<dt><strong><code>out_dir</code></strong> :&ensp;<code>str</code>, optional <code>(default="./")</code></dt>
<dd>Path to directory to save plots to</dd>
<dt><strong><code>save_prefix</code></strong> :&ensp;<code>str</code>, optional <code>(default="")</code></dt>
<dd>String to prepend to output plots to make names unique</dd>
<dt><strong><code>save_output</code></strong> :&ensp;<code>bool</code>, optional <code>(default=</code>False<code>)</code></dt>
<dd>If <code>True</code>, save output dataframe to <code>out_dir/save_prefix_PCA_GSEA.csv</code></dd>
<dt><strong><code>get_gsea_df_kwargs</code></strong> :&ensp;<code>dict</code>, optional <code>(default={})</code></dt>
<dd>Keyword arguments to pass to <code>dc.get_gsea_df</code>. e.g. <code>source</code>, <code>target</code>,
<code>n_background</code>.</dd>
<dt><strong><code>decoupler_dotplot_facet_kwargs</code></strong> :&ensp;<code>dict</code>, optional <code>(default={})</code></dt>
<dd>Keyword arguments to pass to <code>decoupler_dotplot_facet</code>. e.g. <code>top_n</code>, <code>cmap</code>,
<code>dpi</code></dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>enr_pvals</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>GSEA output as dataframe</dd>
</dl>
<p>Saves <code>decoupler</code> GSEA dotplots to <code>out_dir/</code>.</p>
<h2 id="examples">Examples</h2>
<p>nets = fetch_decoupler_resources(
resources=["msigdb", "progeny", "collectri", "liana"],
genome="human",
)
enr_pvals = PCA_decoupler_GSEA(
adata=a,
n_pcs=3,
net=nets["msigdb"],
max_FDRpval=0.05,
out_dir="plots/",
save_prefix="",
save_output=False,
get_gsea_df_kwargs={
"source":"geneset",
"target":"genesymbol",
"times":500,
"min_n":5,
},
)</p></div>
</dd>
<dt id="kitchen.recipes.calculate_cell_proportions"><code class="name flex">
<span>def <span class="ident">calculate_cell_proportions</span></span>(<span>obs_df, celltype_column, groupby, celltype_subset=None)</span>
</code></dt>
<dd>
<div class="desc"><p>calculate proportions of each cell type in a column</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>obs_df</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>Dataframe containing single cells as rows and metadata columns</dd>
<dt><strong><code>celltype_column</code></strong> :&ensp;<code>str</code></dt>
<dd>Column of <code>obs_df</code> containing cell type labels for proportion calculation</dd>
<dt><strong><code>groupby</code></strong> :&ensp;<code>str</code></dt>
<dd>Column of <code>obs_df</code> to group by for proportions (i.e. sample or patient)</dd>
<dt><strong><code>celltype_subset</code></strong> :&ensp;<code>list</code> of <code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>List of cell types in <code>obs_df.celltype_column</code> to keep in output. If <code>None</code>,
return proportions of all celltypes</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>props_df</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>Dataframe with values of <code>obs_df.groupby</code> as rows and values of
<code>obs_df.celltype_column</code> (or <code>celltype_subset</code>) as columns</dd>
</dl></div>
</dd>
<dt id="kitchen.recipes.cc_score"><code class="name flex">
<span>def <span class="ident">cc_score</span></span>(<span>adata, layer=None, seed=18, verbose=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Calculates cell cycle scores and implied phase for each observation</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing transformed and normalized (arcsinh or log1p) counts in
'layer'.</dd>
<dt><strong><code>layer</code></strong> :&ensp;<code>str</code>, optional <code>(default=None)</code></dt>
<dd>key from adata.layers to use for cc phase calculation. Default None to
use .X</dd>
<dt><strong><code>seed</code></strong> :&ensp;<code>int</code>, optional <code>(default=18)</code></dt>
<dd>random state for PCA, neighbors graph and clustering</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>print updates to console</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>adata is edited in place to add 'G2M_score', 'S_score', and 'phase' to .obs</code></dt>
<dd>&nbsp;</dd>
</dl></div>
</dd>
<dt id="kitchen.recipes.cellranger2"><code class="name flex">
<span>def <span class="ident">cellranger2</span></span>(<span>adata, expected=1500, upper_quant=0.99, lower_prop=0.1, label='CellRanger_2', verbose=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Labels cells using "knee point" method from CellRanger 2.1</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing unfiltered counts</dd>
<dt><strong><code>expected</code></strong> :&ensp;<code>int</code>, optional <code>(default=1500)</code></dt>
<dd>estimated number of real cells expected in dataset</dd>
<dt><strong><code>upper_quant</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.99)</code></dt>
<dd>upper quantile of real cells to test</dd>
<dt><strong><code>lower_prop</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.1)</code></dt>
<dd>percentage of expected quantile to calculate total counts threshold for</dd>
<dt><strong><code>label</code></strong> :&ensp;<code>str</code>, optional <code>(default="CellRanger_2")</code></dt>
<dd>how to name .obs column containing output</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>print updates to console</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>adata edited in place to add .obs[label] binary label</code></dt>
<dd>&nbsp;</dd>
</dl></div>
</dd>
<dt id="kitchen.recipes.cellranger3"><code class="name flex">
<span>def <span class="ident">cellranger3</span></span>(<span>adata, init_counts=15000, min_umi_frac_of_median=0.01, min_umis_nonambient=500, max_adj_pvalue=0.01)</span>
</code></dt>
<dd>
<div class="desc"><p>Labels cells using "emptydrops" method from CellRanger 3.0</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing unfiltered counts</dd>
<dt><strong><code>init_counts</code></strong> :&ensp;<code>int</code>, optional <code>(default=15000)</code></dt>
<dd>initial total counts threshold for calling cells</dd>
<dt><strong><code>min_umi_frac_of_median</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.01)</code></dt>
<dd>minimum total counts for testing barcodes as fraction of median counts for
initially labeled cells</dd>
<dt><strong><code>min_umis_nonambient</code></strong> :&ensp;<code>float</code>, optional <code>(default=500)</code></dt>
<dd>minimum total counts for testing barcodes</dd>
<dt><strong><code>max_adj_pvalue</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.01)</code></dt>
<dd>maximum p-value for cell calling after B-H correction</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>adata edited in place to add .obs["CellRanger_3"] binary label</code></dt>
<dd>&nbsp;</dd>
<dt><code>and .obs["CellRanger_3_ll"] log-likelihoods for tested barcodes</code></dt>
<dd>&nbsp;</dd>
</dl></div>
</dd>
<dt id="kitchen.recipes.dim_reduce"><code class="name flex">
<span>def <span class="ident">dim_reduce</span></span>(<span>adata, layer=None, use_rep=None, clust_resolution=1.0, paga=True, seed=18, verbose=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Reduces dimensions of single-cell dataset using standard methods</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing preprocessed counts matrix</dd>
<dt><strong><code>layer</code></strong> :&ensp;<code>str</code>, optional <code>(default=None)</code></dt>
<dd>layer to use; default None for .X</dd>
<dt><strong><code>use_rep</code></strong> :&ensp;<code>str</code>, optional <code>(default=None)</code></dt>
<dd>.obsm key to use for neighbors graph instead of PCA;
default None, generate new PCA from layer</dd>
<dt><strong><code>clust_resolution</code></strong> :&ensp;<code>float</code>, optional <code>(default=1.0)</code></dt>
<dd>resolution as fraction on [0.0, 1.0] for leiden
clustering. default 1.0</dd>
<dt><strong><code>paga</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>run PAGA to seed UMAP embedding</dd>
<dt><strong><code>seed</code></strong> :&ensp;<code>int</code>, optional <code>(default=18)</code></dt>
<dd>random state for PCA, neighbors graph and clustering</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>print updates to console</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>adata is edited in place, adding PCA, neighbors graph, PAGA, and UMAP</code></dt>
<dd>&nbsp;</dd>
</dl></div>
</dd>
<dt id="kitchen.recipes.plot_genes"><code class="name flex">
<span>def <span class="ident">plot_genes</span></span>(<span>adata, de_method='t-test_overestim_var', layer='log1p_norm', groupby='leiden', key_added='rank_genes_groups', ambient=False, plot_type=None, n_genes=5, dendrogram=True, cmap='Greys', figsize_scale=1.0, save_to='de.png', verbose=True, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Calculates and plot <code>rank_genes_groups</code> results</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing preprocessed and dimension-reduced counts matrix</dd>
<dt><strong><code>de_method</code></strong> :&ensp;<code>str</code>, optional <code>(default="t-test_overestim_var")</code></dt>
<dd>one of "t-test", "t-test_overestim_var", "wilcoxon"</dd>
<dt><strong><code>layer</code></strong> :&ensp;<code>str</code>, optional <code>(default="log1p_norm")</code></dt>
<dd>one of <code>adata.layers</code> to use for DEG analysis. Recommended to use 'raw_counts'
if <code>de_method=='wilcoxon'</code> and 'log1p_norm' if <code>de_method=='t-test'</code>.</dd>
<dt><strong><code>groupby</code></strong> :&ensp;<code>str</code>, optional <code>(default="leiden")</code></dt>
<dd><code>adata.obs</code> key to group cells by</dd>
<dt><strong><code>key_added</code></strong> :&ensp;<code>str</code>, optional <code>(default="rank_genes_groups")</code></dt>
<dd><code>adata.uns</code> key to add DEG results to</dd>
<dt><strong><code>ambient</code></strong> :&ensp;<code>bool</code>, optional <code>(default=False)</code></dt>
<dd>include ambient genes as a group in the plot/output dictionary</dd>
<dt><strong><code>plot_type</code></strong> :&ensp;<code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>One of "dotplot", "matrixplot", "dotmatrix", "stacked_violin", or "heatmap".
If <code>None</code>, don't plot, just return DEGs as dictionary.</dd>
<dt><strong><code>n_genes</code></strong> :&ensp;<code>int</code>, optional <code>(default=5)</code></dt>
<dd>number of top genes per group to show</dd>
<dt><strong><code>dendrogram</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>show dendrogram of cluster similarity</dd>
<dt><strong><code>cmap</code></strong> :&ensp;<code>str</code>, optional <code>(default="Greys")</code></dt>
<dd>matplotlib colormap for dots</dd>
<dt><strong><code>figsize_scale</code></strong> :&ensp;<code>float</code>, optional <code>(default=1.0)</code></dt>
<dd>scale dimensions of the figure</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code>, optional <code>(default="de.png")</code></dt>
<dd>string to add to plot name using scanpy plot defaults</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>print updates to console</dd>
<dt><strong><code>**kwargs</code></strong> :&ensp;<code>optional</code></dt>
<dd>keyword args to add to <code><a title="kitchen.plotting.custom_heatmap" href="plotting.html#kitchen.plotting.custom_heatmap">custom_heatmap()</a></code></dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>markers</code></strong> :&ensp;<code>dict</code></dt>
<dd>dictionary of top <code>n_genes</code> DEGs per group</dd>
<dt><strong><code>myplot</code></strong> :&ensp;<code>matplotlib.Figure</code></dt>
<dd><code>custom_heatmap</code> object if <code>plot_type!=None</code></dd>
</dl></div>
</dd>
<dt id="kitchen.recipes.plot_genes_cnmf"><code class="name flex">
<span>def <span class="ident">plot_genes_cnmf</span></span>(<span>adata, plot_type='heatmap', groupby='leiden', attr='varm', keys='cnmf_spectra', indices=None, n_genes=5, dendrogram=True, figsize_scale=1.0, cmap='Greys', save_to='de_cnmf.png', **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Calculates and plots top cNMF gene loadings</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>object containing preprocessed and dimension-reduced counts matrix</dd>
<dt><strong><code>plot_type</code></strong> :&ensp;<code>str</code>, optional <code>(default=</code>None<code>)</code></dt>
<dd>One of "dotplot", "matrixplot", "dotmatrix", "stacked_violin", or "heatmap".</dd>
<dt><strong><code>groupby</code></strong> :&ensp;<code>str</code>, optional <code>(default="leiden")</code></dt>
<dd>.obs key to group cells by</dd>
<dt><strong><code>attr</code></strong> :&ensp;<code>str {"var", "obs", "uns", "varm", "obsm"}</code></dt>
<dd>attribute of adata that contains the score</dd>
<dt><strong><code>keys</code></strong> :&ensp;<code>str</code> or <code>list</code> of <code>str</code>, optional <code>(default="cnmf_spectra")</code></dt>
<dd>scores to look up an array from the attribute of adata</dd>
<dt><strong><code>indices</code></strong> :&ensp;<code>list</code> of <code>int</code>, optional <code>(default=None)</code></dt>
<dd>column indices of keys for which to plot (e.g. [0,1,2] for first three keys)</dd>
<dt><strong><code>n_genes</code></strong> :&ensp;<code>int</code>, optional <code>(default=5)</code></dt>
<dd>number of top genes per group to show</dd>
<dt><strong><code>dendrogram</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>show dendrogram of cluster similarity</dd>
<dt><strong><code>figsize_scale</code></strong> :&ensp;<code>float</code>, optional <code>(default=1.0)</code></dt>
<dd>scale dimensions of the figure</dd>
<dt><strong><code>cmap</code></strong> :&ensp;<code>str</code>, optional <code>(default="Greys")</code></dt>
<dd>valid color map for the plot</dd>
<dt><strong><code>save_to</code></strong> :&ensp;<code>str</code>, optional <code>(default="de.png")</code></dt>
<dd>string to add to plot name using scanpy plot defaults</dd>
<dt><strong><code>**kwargs</code></strong> :&ensp;<code>optional</code></dt>
<dd>keyword args to add to sc.pl.matrixplot, sc.pl.dotplot, or sc.pl.heatmap</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>markers</code></strong> :&ensp;<code>dict</code></dt>
<dd>dictionary of top <code>n_genes</code> features per NMF factor</dd>
<dt><strong><code>myplot</code></strong> :&ensp;<code>matplotlib.Figure</code></dt>
<dd><code>custom_heatmap</code> object</dd>
</dl></div>
</dd>
<dt id="kitchen.recipes.scDEG_decoupler_GSEA"><code class="name flex">
<span>def <span class="ident">scDEG_decoupler_GSEA</span></span>(<span>adata, uns_key, net, max_FDRpval=0.05, out_dir='./', save_prefix='', save_output=False, get_gsea_df_kwargs={}, decoupler_dotplot_facet_kwargs={})</span>
</code></dt>
<dd>
<div class="desc"><p>Quickly process DEGs from scRNA dataset through <code>decoupler</code> GSEA and create plot for
initial look at pathways</p>
<p><code>decoupler</code> tools used:
* "GSEA" for biological pathways</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>Object containing rank_genes_groups results in <code>adata.uns</code></dd>
<dt><strong><code>uns_key</code></strong> :&ensp;<code>str</code></dt>
<dd>Key to <code>adata.uns</code> containing rank_genes_groups results</dd>
<dt><strong><code>net</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>Network dataframe required to run ORA. e.g. <code>msigdb</code> where <code>msigdb</code> is a
<code>pd.DataFrame</code> from <code>dc.get_resource</code>.</dd>
<dt><strong><code>max_FDRpval</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.05)</code></dt>
<dd>FDR p-value cutoff for using genes and plotting significant GSEA pathways.</dd>
<dt><strong><code>out_dir</code></strong> :&ensp;<code>str</code>, optional <code>(default="./")</code></dt>
<dd>Path to directory to save plots to</dd>
<dt><strong><code>save_prefix</code></strong> :&ensp;<code>str</code>, optional <code>(default="")</code></dt>
<dd>String to prepend to output plots to make names unique</dd>
<dt><strong><code>save_output</code></strong> :&ensp;<code>bool</code>, optional <code>(default=</code>False<code>)</code></dt>
<dd>If <code>True</code>, save output dataframe to <code>out_dir/save_prefix_NMF_GSEA.csv</code></dd>
<dt><strong><code>get_gsea_df_kwargs</code></strong> :&ensp;<code>dict</code>, optional <code>(default={})</code></dt>
<dd>Keyword arguments to pass to <code>dc.get_ora_df</code>. e.g. <code>source</code>, <code>target</code>,
<code>n_background</code>.</dd>
<dt><strong><code>decoupler_dotplot_facet_kwargs</code></strong> :&ensp;<code>dict</code>, optional <code>(default={})</code></dt>
<dd>Keyword arguments to pass to <code>decoupler_dotplot_facet</code>. e.g. <code>top_n</code>, <code>cmap</code>,
<code>dpi</code></dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>enr_pvals</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>ORA output as dataframe</dd>
</dl>
<p>Saves <code>decoupler</code> ORA dotplots to <code>out_dir/</code>.</p>
<h2 id="examples">Examples</h2>
<p>nets = fetch_decoupler_resources(
resources=["msigdb", "progeny", "collectri", "liana"],
genome="human",
)
frame = scDEG_decoupler_GSEA(
a,
uns_key="rank_genes_groups_{}".format(key),
net=nets["msigdb"],
max_FDRpval=0.05,
out_dir="plots/",
save_prefix=f"HALLMARK_{key}_",
save_output=False,
get_gsea_df_kwargs={
"source":"geneset",
"target":"genesymbol",
"times":200,
"min_n":5,
"seed":8,
},
decoupler_dotplot_facet_kwargs={
"top_n":10,
"dpi":200,
"cmap":"Reds_r",
"figsize_scale":1.2,
},
)</p></div>
</dd>
<dt id="kitchen.recipes.scDEG_decoupler_ORA"><code class="name flex">
<span>def <span class="ident">scDEG_decoupler_ORA</span></span>(<span>adata, uns_key, net, max_FDRpval=0.05, min_logFC=0.5849625007211562, out_dir='./', save_prefix='', save_output=False, get_ora_df_kwargs={}, decoupler_dotplot_facet_kwargs={})</span>
</code></dt>
<dd>
<div class="desc"><p>Quickly process DEGs from scRNA dataset through <code>decoupler</code> ORA and create plot for
initial look at pathways</p>
<p><code>decoupler</code> tools used:
* "ORA" for biological pathways</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>Object containing rank_genes_groups results in <code>adata.uns</code></dd>
<dt><strong><code>uns_key</code></strong> :&ensp;<code>str</code></dt>
<dd>Key to <code>adata.uns</code> containing rank_genes_groups results</dd>
<dt><strong><code>net</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>Network dataframe required to run ORA. e.g. <code>msigdb</code> where <code>msigdb</code> is a
<code>pd.DataFrame</code> from <code>dc.get_resource</code>.</dd>
<dt><strong><code>max_FDRpval</code></strong> :&ensp;<code>float</code>, optional <code>(default=0.05)</code></dt>
<dd>FDR p-value cutoff for using genes and plotting significant ORA pathways.</dd>
<dt><strong><code>min_logFC</code></strong> :&ensp;<code>float</code>, optional <code>(default=1.5)</code></dt>
<dd>logFC cutoff for using genes and plotting significant ORA pathways.</dd>
<dt><strong><code>out_dir</code></strong> :&ensp;<code>str</code>, optional <code>(default="./")</code></dt>
<dd>Path to directory to save plots to</dd>
<dt><strong><code>save_prefix</code></strong> :&ensp;<code>str</code>, optional <code>(default="")</code></dt>
<dd>String to prepend to output plots to make names unique</dd>
<dt><strong><code>save_output</code></strong> :&ensp;<code>bool</code>, optional <code>(default=</code>False<code>)</code></dt>
<dd>If <code>True</code>, save output dataframe to <code>out_dir/save_prefix_NMF_ORA.csv</code></dd>
<dt><strong><code>get_ora_df_kwargs</code></strong> :&ensp;<code>dict</code>, optional <code>(default={})</code></dt>
<dd>Keyword arguments to pass to <code>dc.get_ora_df</code>. e.g. <code>source</code>, <code>target</code>,
<code>n_background</code>.</dd>
<dt><strong><code>decoupler_dotplot_facet_kwargs</code></strong> :&ensp;<code>dict</code>, optional <code>(default={})</code></dt>
<dd>Keyword arguments to pass to <code>decoupler_dotplot_facet</code>. e.g. <code>top_n</code>, <code>cmap</code>,
<code>dpi</code></dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>enr_pvals</code></strong> :&ensp;<code>pd.DataFrame</code></dt>
<dd>ORA output as dataframe</dd>
</dl>
<p>Saves <code>decoupler</code> ORA dotplots to <code>out_dir/</code>.</p></div>
</dd>
<dt id="kitchen.recipes.score_gene_signatures"><code class="name flex">
<span>def <span class="ident">score_gene_signatures</span></span>(<span>adata, signatures_dict, method='seurat', sig_subset=None, layer=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Score gene signatures in AnnData using <code>sc.tl.score_genes</code> or <code>dc.run_gsva</code></p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>Object containing gene expression data for scoring</dd>
<dt><strong><code>signatures_dict</code></strong> :&ensp;<code>dict</code></dt>
<dd>Dictionary of signature names (keys) and constituent genes (values), with gene
names matching <code>adata.var_names</code></dd>
<dt><strong><code>method</code></strong> :&ensp;<code>literal ['seurat','gsva']</code>, optional <code>(default='seurat')</code></dt>
<dd>Method to use for scoring. 'seurat':<code>sc.tl.score_genes</code>; 'gsva':<code>dc.run_gsva</code>.</dd>
<dt><strong><code>sig_subset</code></strong> :&ensp;<code>list, Optional (default=</code>None<code>)</code></dt>
<dd>Subset of keys from <code>signatures_dict</code> to attempt scoring. If <code>None</code>, score all
signatures from <code>signatures_dict</code></dd>
<dt><strong><code>layer</code></strong> :&ensp;<code>str, Optional (default=</code>None<code>)</code></dt>
<dd>Layer key from <code>adata.layers</code> to use for scoring. If <code>None</code>, use <code>adata.X</code>.</dd>
</dl>
<h2 id="returns">Returns</h2>
<p><code>adata</code> is edited with signature scores added to <code>.obs</code></p>
<dl>
<dt><strong><code>failed_sigs</code></strong> :&ensp;<code>list</code></dt>
<dd>List of signatures not scored properly</dd>
<dt><strong><code>scored_sigs</code></strong> :&ensp;<code>list</code></dt>
<dd>List of signatures scored successfully</dd>
</dl></div>
</dd>
<dt id="kitchen.recipes.subset_adata"><code class="name flex">
<span>def <span class="ident">subset_adata</span></span>(<span>adata, subset, verbose=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Subsets AnnData object on one or more .obs columns</p>
<p>Columns should contain 0/False for cells to throw out, and 1/True for cells to
keep. Keeps union of all labels provided in subset.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>the data</dd>
<dt><strong><code>subset</code></strong> :&ensp;<code>str</code> or <code>list</code> of <code>str</code></dt>
<dd>adata.obs labels to use for subsetting. Labels must be binary (0, "0", False,
"False" to toss - 1, "1", True, "True" to keep). Multiple labels will keep
intersection.</dd>
<dt><strong><code>verbose</code></strong> :&ensp;<code>bool</code>, optional <code>(default=True)</code></dt>
<dd>print updates to console</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>adata</code></strong> :&ensp;<code>anndata.AnnData</code></dt>
<dd>new anndata object as subset of <code>adata</code></dd>
</dl></div>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="kitchen" href="index.html">kitchen</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="kitchen.recipes.NMF_decoupler_ORA" href="#kitchen.recipes.NMF_decoupler_ORA">NMF_decoupler_ORA</a></code></li>
<li><code><a title="kitchen.recipes.PCA_decoupler_GSEA" href="#kitchen.recipes.PCA_decoupler_GSEA">PCA_decoupler_GSEA</a></code></li>
<li><code><a title="kitchen.recipes.calculate_cell_proportions" href="#kitchen.recipes.calculate_cell_proportions">calculate_cell_proportions</a></code></li>
<li><code><a title="kitchen.recipes.cc_score" href="#kitchen.recipes.cc_score">cc_score</a></code></li>
<li><code><a title="kitchen.recipes.cellranger2" href="#kitchen.recipes.cellranger2">cellranger2</a></code></li>
<li><code><a title="kitchen.recipes.cellranger3" href="#kitchen.recipes.cellranger3">cellranger3</a></code></li>
<li><code><a title="kitchen.recipes.dim_reduce" href="#kitchen.recipes.dim_reduce">dim_reduce</a></code></li>
<li><code><a title="kitchen.recipes.plot_genes" href="#kitchen.recipes.plot_genes">plot_genes</a></code></li>
<li><code><a title="kitchen.recipes.plot_genes_cnmf" href="#kitchen.recipes.plot_genes_cnmf">plot_genes_cnmf</a></code></li>
<li><code><a title="kitchen.recipes.scDEG_decoupler_GSEA" href="#kitchen.recipes.scDEG_decoupler_GSEA">scDEG_decoupler_GSEA</a></code></li>
<li><code><a title="kitchen.recipes.scDEG_decoupler_ORA" href="#kitchen.recipes.scDEG_decoupler_ORA">scDEG_decoupler_ORA</a></code></li>
<li><code><a title="kitchen.recipes.score_gene_signatures" href="#kitchen.recipes.score_gene_signatures">score_gene_signatures</a></code></li>
<li><code><a title="kitchen.recipes.subset_adata" href="#kitchen.recipes.subset_adata">subset_adata</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.11.1</a>.</p>
</footer>
</body>
</html>
